<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jennifer Blanc" />


<title>Calculate TGWAS</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Calculate TGWAS</h1>
<h4 class="author">Jennifer Blanc</h4>
<h4 class="date">4/14/2022</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-04-26
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201015code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201015)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201015code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201015)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
unnamed-chunk-5
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory <code>Calculate_TGWAS_cache</code> and re-run the analysis. To have workflowr automatically delete the cache directory prior to building the file, set <code>delete_cache = TRUE</code> when running <code>wflow_build()</code> or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree931f81811a70a5965859780201a24c6e17e2fab8targetblank931f818a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/931f81811a70a5965859780201a24c6e17e2fab8" target="_blank">931f818</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree931f81811a70a5965859780201a24c6e17e2fab8targetblank931f818a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/931f81811a70a5965859780201a24c6e17e2fab8" target="_blank">931f818</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/
    Ignored:    analysis/Calculate_TGWAS_cache/
    Ignored:    code/.DS_Store
    Ignored:    code/1kg/.DS_Store
    Ignored:    code/Calculate_Tm/.DS_Store
    Ignored:    code/Debug/.DS_Store
    Ignored:    code/PGA_test/.DS_Store
    Ignored:    code/PRS/.DS_Store
    Ignored:    code/Simulate_Genotypes/.DS_Store
    Ignored:    code/Simulate_Phenotypes/.DS_Store
    Ignored:    code/Snakefiles/.DS_Store
    Ignored:    code/Snakefiles/4PopSplit/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/1kg/
    Ignored:    data/cgd_example/
    Ignored:    data/projection_example/
    Ignored:    data/proof/
    Ignored:    output/.DS_Store
    Ignored:    output/Calculate_TGWAS/
    Ignored:    output/Calculate_Tm/4PopSplit/A10/
    Ignored:    output/Calculate_Tm/4PopSplit/A2/
    Ignored:    output/Calculate_Tm/4PopSplit/B1/
    Ignored:    output/Calculate_Tm/4PopSplit/E2/
    Ignored:    output/Calculate_Tm/4PopSplit/E20/
    Ignored:    output/Calculate_Tm/4PopSplit/E3/
    Ignored:    output/Calculate_Tm/4PopSplit/E4/
    Ignored:    output/Calculate_Tm/4PopSplit/E5/
    Ignored:    output/Calculate_Tm/4PopSplit/E6/
    Ignored:    output/Calculate_Tm/4PopSplit/E7/
    Ignored:    output/Calculate_Tm/4PopSplit/E8/
    Ignored:    output/Calculate_Tm/4PopSplit/E9/
    Ignored:    output/Calculate_Tm/SimpleGrid/A1/
    Ignored:    output/Calculate_Tm/SimpleGrid/B1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E2/
    Ignored:    output/Calculate_Tm/SimpleGrid/E3/
    Ignored:    output/Calculate_Tm/SimpleGrid/E4/
    Ignored:    output/Calculate_Tm/SimpleGrid/E5/
    Ignored:    output/PGA_test/
    Ignored:    output/PRS/
    Ignored:    output/Run_GWAS/
    Ignored:    output/Simulate_Genotypes/.DS_Store
    Ignored:    output/Simulate_Genotypes/4PopSplit/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/A1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E2/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E3/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E4/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E5/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E6/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E7/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E8/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/S1/
    Ignored:    output/Simulate_Phenotypes/

Untracked files:
    Untracked:  _workflowr.yml~
    Untracked:  analysis/Calculate_TGWAS.Rmd
    Untracked:  analysis/Run_GWAS.Rmd
    Untracked:  code/Snakefiles/SimpleGrid/North-South-Meta/
    Untracked:  plink2.acount
    Untracked:  plink2.fst.summary
    Untracked:  plink2.log
    Untracked:  temp.R

Unstaged changes:
    Modified:   _workflowr.yml
    Modified:   analysis/Midparent.Rmd
    Modified:   analysis/Projection_plink2.Rmd
    Modified:   analysis/Simulate_Genotypes.Rmd
    Modified:   analysis/Simulate_Phenotypes.Rmd
    Modified:   analysis/Simulation_Methods_Reproducible.Rmd
    Modified:   analysis/index.Rmd
    Modified:   code/Calculate_Tm/4PopSplit_make_tvec.R
    Modified:   code/Calculate_Tm/calc_Tm.R
    Modified:   code/Calculate_Tm/format_ID_covars.R
    Modified:   code/PRS/clump.R
    Modified:   code/Simulate_Genotypes/generate_genotypes_4PopSplit.py
    Modified:   code/Simulate_Genotypes/split_gwas-test_4PopSplit.R
    Modified:   code/Simulate_Phenotypes/draw_effect_sizes_4PopSplit.R
    Modified:   code/Simulate_Phenotypes/simulate_phenotypes_4PopSplit.R
    Deleted:    code/Snakefiles/4PopSplit/Signal/snakefile
    Modified:   snakefile

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="calculate-tgwas" class="section level2">
<h2>Calculate <span class="math inline">\(T^{GWAS}\)</span></h2>
<p>Here we will walk through our procedure for estimating <span class="math inline">\(T^{GWAS}\)</span>.</p>
<div id="make-test-vector" class="section level3">
<h3>Make Test Vector</h3>
<p>First we need to make the test vector along which will we test for selection. In our 4 population toy model this is simply population ID so we create a <span class="math inline">\(n \times1\)</span> vector, where <span class="math inline">\(n\)</span> is the number of individuals in the test panel, where individuals in the first test population are coded as <span class="math inline">\(1\)</span> and individuals in the second test population are <span class="math inline">\(0\)</span>. We then mean center the test vector.</p>
<pre class="python"><code># Snakemake rule 
rule make_test_vector:
    input:
        pops=&quot;output/Simulate_Genotypes/4PopSplit/{rep}/genos.pop&quot;,
        fam=&quot;output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-test_common.psam&quot;
    output:
        &quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tvec.txt&quot;
    shell:
        &quot;Rscript code/Calculate_Tm/4PopSplit_make_tvec.R {input.pops} {input.fam} {output}&quot;</code></pre>
<p>Example:</p>
<pre class="r"><code># Read in Fam file
fam &lt;- fread(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.psam&quot;)
colnames(fam) &lt;- c(&quot;FID&quot;, &quot;IID&quot;, &quot;SEX&quot;)

# Read in population information 
pops &lt;- fread(&quot;../output/Simulate_Genotypes/4PopSplit/B1/genos.pop&quot;, header = F)
colnames(pops)  &lt;- c(&quot;FID&quot;, &quot;IID&quot;, &quot;POP&quot;)

# Join tables
pop &lt;- dplyr::inner_join(pops, fam) %&gt;% dplyr::select(c(&quot;FID&quot;, &quot;IID&quot;, &quot;POP&quot;))
test_pops &lt;- unique(pop$POP)

# Make test vector as a new column
pop &lt;- pop %&gt;% mutate(Tvec = case_when(POP == test_pops[1] ~ 1, POP == test_pops[2] ~ 0)) %&gt;% mutate(Tvec = Tvec - mean(Tvec))

# Dispaly test vector 
table(pop$POP, pop$Tvec)</code></pre>
<pre><code>   
    -0.5  0.5
  B    0 1000
  D 1000    0</code></pre>
</div>
<div id="compute-covariate" class="section level3">
<h3>Compute covariate</h3>
<p>In the main text we derive a single covariate, <span class="math inline">\(T^{GWAS}\)</span>, to control for stratification specific to the test of interest.</p>
<p><span class="math display">\[\vec{T}^{GWAS} = GX^T(XX^T)^{-1}\vec{T}\]</span></p>
<p>Here <span class="math inline">\(G\)</span> is the GWAS genotype matrix while <span class="math inline">\(X\)</span> is the test genotype matrix. Due to the size of the genotype matrices we cannot simply multiply the matrices in R. Instead we use plink2 to project each test panel principal component into the GWAS panel and then weight them according to their covariance with the test vector.</p>
<p>First we can write the test panel genotype matrix in terms of its singular value decomposition, <span class="math display">\[\vec{T}^{GWAS} = GV_X\Lambda_XU_X^T(U_X\Lambda^{2}U_X^T)^{-1}\vec{T} \\ 
= GV_X\Lambda_XU_X^TU_X\Lambda^{-2}U_X^T\vec{T} \\ 
= GV_X\Lambda_X^{-1}U_X^T\vec{T}\]</span></p>
<p>We can then write our test vector as a sum of eigenvectors of the test panel, <span class="math inline">\(\vec{T} = \sum\limits_{n=1}^N\alpha_n\vec{U}_{X,n}\)</span></p>
<p><span class="math display">\[\vec{T}^{GWAS} = \sum\limits_{n=1}^N \frac{\alpha_n}{\lambda_n} G \vec{V}_{X, n}\]</span></p>
<p>Using the above express we can see that one way to compute <span class="math inline">\(\vec{T}^{GWAS}\)</span> is to:</p>
<ol style="list-style-type: decimal">
<li>Use plink2 <code>--pca allele-wts</code> on the test panel genotype matrix to get the eigenvectors <span class="math inline">\(U_X\)</span>, the SNP loadings <span class="math inline">\(V_X\)</span>, and the eigenvalues <span class="math inline">\(\Lambda_X\)</span>.<br />
</li>
<li>Use R to calculate the weight of each test panel eigenvector on the test vector: <span class="math inline">\(\alpha_k = \vec{U}_{X,k}^T \vec{T}\)</span><br />
</li>
<li>Use plink2 <code>--sccore</code> to multiply the rows of <span class="math inline">\(G\)</span> by all the columns of <span class="math inline">\(V_X\)</span>. This is equivalent to projecting each of the test panel eigenvectors in <span class="math inline">\(G\)</span>.<br />
</li>
<li>Use R to weight each projected eigenvalue by <span class="math inline">\(\frac{\alpha_k}{\lambda_k}\)</span> and sum across <span class="math inline">\(N\)</span>.</li>
</ol>
<pre class="python"><code># Snakemake rules
rule proj_T:
    input:
        &quot;output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-test_common.psam&quot;,
        &quot;output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-gwas_common.psam&quot;,
        &quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tvec.txt&quot;
    params:
        n_minus_1 = int(SIZE)-1,
        col_start = 6,
        col_end = int(SIZE) + 4
    output:
        &quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/pca.eigenvec&quot;,
        &quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/pca.eigenval&quot;,
        &quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/pca.eigenvec.allele&quot;,
        &quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/projection.sscore&quot;
    shell:
        &quot;&quot;&quot;
        plink2 \
        --pfile output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/{wildcards.config}/genos-test_common \
        --pca allele-wts {params.n_minus_1} \
        --out output/Calculate_Tm/4PopSplit/{wildcards.rep}/{wildcards.config}/pca

        plink2 \
        --pfile output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/{wildcards.config}/genos-gwas_common \
        --score output/Calculate_Tm/4PopSplit/{wildcards.rep}/{wildcards.config}/pca.eigenvec.allele 2 5 header-read no-mean-imputation variance-standardize \
        --score-col-nums {params.col_start}-{params.col_end} \
        --out output/Calculate_Tm/4PopSplit/{wildcards.rep}/{wildcards.config}/projection
        &quot;&quot;&quot;

rule calc_Tm:
    input:
        vecs=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/pca.eigenvec&quot;,
        vals=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/pca.eigenval&quot;,
        proj=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/projection.sscore&quot;,
        tvec=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tvec.txt&quot;,
        allele=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/pca.eigenvec.allele&quot;
    output:
        Tm=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tm.txt&quot;,
        weights=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Test_weights.txt&quot;
    shell:
        &quot;&quot;&quot;
        Rscript code/Calculate_Tm/calc_Tm.R {input.vecs} {input.vals} {input.proj} {input.tvec} {output.Tm} {output.weights}
        rm {input.allele}
        &quot;&quot;&quot;</code></pre>
<p>Next we walk through our 4 population example. In this example we are working in the configuration where there is shared structure between the GWAS and test panels (i.e Figure 1A). Here our test vector is separating populations on opposite sides of the deepest split. Because there is overlapping structure and shared drift between panels, we expect that <span class="math inline">\(\vec{T}^{GWAS}\)</span> will also separate the two population that comprise the GWAS panel.</p>
<p>Let’s read in both the test and GWAS genotype matrices.</p>
<pre class="r"><code># Read GWAS Matrix 
pvar &lt;- NewPvar(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pvar&quot;)
d1 &lt;- NewPgen(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pgen&quot;)
G &lt;- ReadList(d1,seq(1,10905), meanimpute=F)
G &lt;- scale(G)

# Read in Test Matrix 
pvar &lt;- NewPvar(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.pvar&quot;)
d1 &lt;- NewPgen(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.pgen&quot;)
X &lt;- ReadList(d1,seq(1, 10905), meanimpute=F)
X &lt;- scale(X)</code></pre>
<p>First, let’s calculate <span class="math inline">\(\vec{T}^{GWAS}\)</span> just by doing matrix multiplication.</p>
<pre class="r"><code>Tvec &lt;- pop$Tvec
TGWAS_mat &lt;-  G %*% t(X) %*% pinv(X %*% t(X)) %*% Tvec
plot(TGWAS_mat)</code></pre>
<p><img src="figure/Calculate_TGWAS.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /> We can see that <span class="math inline">\(\vec{T}^{GWAS}\)</span> is recapitulating the population structure we expect.</p>
<p>Now let’s check that the our plink procedure get’s us the same answer.</p>
<pre class="r"><code># Load test eigen vecs
vecs &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/pca.eigenvec&quot;)
vecs &lt;- vecs[,3:ncol(vecs)]
vecs &lt;- apply(vecs, 2, as.numeric)

# Get the weights of each eigenvector
B &lt;- t(vecs) %*% Tvec
Br = abs(B) / sum(abs(B))

# Load Projected eigenvectors
proj_vecs &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/projection.sscore&quot;)
proj_vecs &lt;- proj_vecs[,5:ncol(proj_vecs)]
proj_vecs &lt;- apply(proj_vecs, 2, as.numeric)

# Load Eigenvalues
vals &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/pca.eigenval&quot;)
vals &lt;- vals$V1

# Calculate Tm (for some reason need to scale by -2)
Tm &lt;- proj_vecs %*% diag(1/sqrt(vals)) %*% B
TGWAS &lt;- -2 * Tm

plot(TGWAS, TGWAS_mat)
abline(0,1, col = &quot;red&quot;)</code></pre>
<p><img src="figure/Calculate_TGWAS.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Using plink2 to compute <span class="math inline">\(\vec{T}^{GWAS}\)</span> allows us to efficiently calculate our covariate of interest while avoiding any larger scale matrix multiplication.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] pracma_2.3.3      dplyr_1.0.7       data.table_1.14.2 pgenlibr_0.3.1   

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       highr_0.9        pillar_1.6.4     compiler_3.6.2  
 [5] bslib_0.3.0      later_1.3.0      jquerylib_0.1.4  git2r_0.28.0    
 [9] workflowr_1.6.2  tools_3.6.2      digest_0.6.29    lattice_0.20-44 
[13] jsonlite_1.7.2   evaluate_0.14    lifecycle_1.0.1  tibble_3.1.6    
[17] png_0.1-7        pkgconfig_2.0.3  rlang_0.4.12     Matrix_1.3-4    
[21] DBI_1.1.1        yaml_2.2.1       xfun_0.29        fastmap_1.1.0   
[25] stringr_1.4.0    knitr_1.37       generics_0.1.0   fs_1.5.2        
[29] vctrs_0.3.8      sass_0.4.0       grid_3.6.2       tidyselect_1.1.1
[33] rprojroot_2.0.2  reticulate_1.20  glue_1.6.0       R6_2.5.1        
[37] fansi_0.5.0      rmarkdown_2.10   purrr_0.3.4      magrittr_2.0.1  
[41] promises_1.2.0.1 ellipsis_0.3.2   htmltools_0.5.2  assertthat_0.2.1
[45] httpuv_1.6.2     utf8_1.2.2       stringi_1.7.6    crayon_1.4.2    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
