<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Home</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Home</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-05-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 2 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtreeb6e4c90096db0893b4de50953255fb6c89aebaf8targetblankb6e4c90a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/b6e4c90096db0893b4de50953255fb6c89aebaf8" target="_blank">b6e4c90</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtreeb6e4c90096db0893b4de50953255fb6c89aebaf8targetblankb6e4c90a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/b6e4c90096db0893b4de50953255fb6c89aebaf8" target="_blank">b6e4c90</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/
    Ignored:    analysis/Calculate_TGWAS_cache/
    Ignored:    analysis/Polygenic_Selection_Test_cache/
    Ignored:    code/.DS_Store
    Ignored:    code/1kg/.DS_Store
    Ignored:    code/Calculate_Tm/.DS_Store
    Ignored:    code/Debug/.DS_Store
    Ignored:    code/PGA_test/.DS_Store
    Ignored:    code/PRS/.DS_Store
    Ignored:    code/Simulate_Genotypes/.DS_Store
    Ignored:    code/Simulate_Phenotypes/.DS_Store
    Ignored:    code/Snakefiles/.DS_Store
    Ignored:    code/Snakefiles/4PopSplit/.DS_Store
    Ignored:    code/Snakefiles/4PopSplit/Signal/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/conda-archive/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/conda/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/locks/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/metadata/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/shadow/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/singularity/
    Ignored:    data/.DS_Store
    Ignored:    data/1kg/
    Ignored:    data/cgd_example/
    Ignored:    data/projection_example/
    Ignored:    data/proof/
    Ignored:    output/.DS_Store
    Ignored:    output/Calculate_TGWAS/
    Ignored:    output/Calculate_Tm/4PopSplit/
    Ignored:    output/Calculate_Tm/SimpleGrid/A1/
    Ignored:    output/Calculate_Tm/SimpleGrid/B1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E2/
    Ignored:    output/Calculate_Tm/SimpleGrid/E3/
    Ignored:    output/Calculate_Tm/SimpleGrid/E4/
    Ignored:    output/Calculate_Tm/SimpleGrid/E5/
    Ignored:    output/PGA_test/
    Ignored:    output/PRS/
    Ignored:    output/Run_GWAS/
    Ignored:    output/Simulate_Genotypes/.DS_Store
    Ignored:    output/Simulate_Genotypes/4PopSplit/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/A1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E2/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E3/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E4/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E5/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E6/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E7/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E8/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/S1/
    Ignored:    output/Simulate_Phenotypes/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/index.Rmd</code>) and HTML (<code>docs/index.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/51065b39b2903aa1b80e64ca9c56a19bb0356ea5/analysis/index.Rmd" target="_blank">51065b3</a>
</td>
<td>
jgblanc
</td>
<td>
2022-04-29
</td>
<td>
ready for final 4 pop run
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/51065b39b2903aa1b80e64ca9c56a19bb0356ea5/docs/index.html" target="_blank">51065b3</a>
</td>
<td>
jgblanc
</td>
<td>
2022-04-29
</td>
<td>
ready for final 4 pop run
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/61a28b20f1219773940c9d5f15c03b3ea1ef3d30/docs/index.html" target="_blank">61a28b2</a>
</td>
<td>
jgblanc
</td>
<td>
2022-02-07
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/deb2706feab2347cd1851cab086ad39b9f6c6020/analysis/index.Rmd" target="_blank">deb2706</a>
</td>
<td>
jgblanc
</td>
<td>
2022-02-07
</td>
<td>
wflow_publish(“analysis/index.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/db79e2ed4f64a80e540c758f89ecac59b34e9fd0/docs/index.html" target="_blank">db79e2e</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/1e10dec627b5ea23e6a139befadaf18de7b4dfbc/analysis/index.Rmd" target="_blank">1e10dec</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
<td>
wflow_publish(c(“analysis/Simulation_Methods_Reproducible.Rmd”,
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/3c774bfa756e4900af963112d14deb7ed000f4e8/docs/index.html" target="_blank">3c774bf</a>
</td>
<td>
jgblanc
</td>
<td>
2021-10-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/cba0a794d92676b31dc4ac30147503c02840d20d/analysis/index.Rmd" target="_blank">cba0a79</a>
</td>
<td>
jgblanc
</td>
<td>
2021-10-05
</td>
<td>
Publish website
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/d395ec753b671275347ed497cdb3d84aed17816f/analysis/index.Rmd" target="_blank">d395ec7</a>
</td>
<td>
jgblanc
</td>
<td>
2021-10-04
</td>
<td>
transfer true signal up
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/d395ec753b671275347ed497cdb3d84aed17816f/docs/index.html" target="_blank">d395ec7</a>
</td>
<td>
jgblanc
</td>
<td>
2021-10-04
</td>
<td>
transfer true signal up
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/c811553addf687a880dbfc099fc4b74459348796/docs/index.html" target="_blank">c811553</a>
</td>
<td>
jgblanc
</td>
<td>
2021-10-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/b47db7e65401219f3095b0c4d0e71e5afe55cf20/analysis/index.Rmd" target="_blank">b47db7e</a>
</td>
<td>
jgblanc
</td>
<td>
2021-10-04
</td>
<td>
trying git wflow
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/4a478051738fc7e6fb2d027da74d1a0b02c9722c/analysis/index.Rmd" target="_blank">4a47805</a>
</td>
<td>
jgblanc
</td>
<td>
2021-09-29
</td>
<td>
organizing website
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/4a478051738fc7e6fb2d027da74d1a0b02c9722c/docs/index.html" target="_blank">4a47805</a>
</td>
<td>
jgblanc
</td>
<td>
2021-09-29
</td>
<td>
organizing website
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/4da4d1ddf7ed4417adc6f911e23e6e6ac552da30/docs/index.html" target="_blank">4da4d1d</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/de741dbbaf6452eed618be884b1839c4cb253a52/analysis/index.Rmd" target="_blank">de741db</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-18
</td>
<td>
testing wflowr
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/552d55057f8bec9a558a259677122d854ee84f88/docs/index.html" target="_blank">552d550</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-18
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/db67bf87403dbf414c8c82b5a8d462b62632ba80/docs/index.html" target="_blank">db67bf8</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-17
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/ba6e2e083d37c95adc026f7ff7184e00771a852c/docs/index.html" target="_blank">ba6e2e0</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-17
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/011a710681ef21d98e57d073ec4d9334518b15b4/analysis/index.Rmd" target="_blank">011a710</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-17
</td>
<td>
Meeting notes
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/35227cccc214f2dcdf39fa20bde86dbbab86e954/docs/index.html" target="_blank">35227cc</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/3f17ad7e08262766b5ec7df99202ce429477e959/analysis/index.Rmd" target="_blank">3f17ad7</a>
</td>
<td>
jgblanc
</td>
<td>
2021-08-02
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/Simulations_methods.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/d9097240c6da5ad034c144b1e1cc9a9740394d9d/docs/index.html" target="_blank">d909724</a>
</td>
<td>
jgblanc
</td>
<td>
2021-06-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/64a812dcced2ff5b9aff69c250cecf27d719312f/analysis/index.Rmd" target="_blank">64a812d</a>
</td>
<td>
jgblanc
</td>
<td>
2021-06-15
</td>
<td>
wflow_publish(c(“analysis/Plot_4PopSplit.Rmd”, “analysis/Plot_SimpleGrid.Rmd”,
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/1b63da0512a6ca0644c7f36a67529f759688fe56/docs/index.html" target="_blank">1b63da0</a>
</td>
<td>
jgblanc
</td>
<td>
2021-06-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/46a0ddd147cf0dcf35fce52284ada596d752c366/analysis/index.Rmd" target="_blank">46a0ddd</a>
</td>
<td>
jgblanc
</td>
<td>
2021-06-15
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/e1445422fbc438ca45004c039cb4992ca300e576/docs/index.html" target="_blank">e144542</a>
</td>
<td>
jgblanc
</td>
<td>
2021-05-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/a83c1a34853e12d7ade74b2c6114286de9af4cf9/analysis/index.Rmd" target="_blank">a83c1a3</a>
</td>
<td>
jgblanc
</td>
<td>
2021-05-04
</td>
<td>
calc qx page
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/d3bef4b79b2cf08b7ed0842a998c10bf55b6aa2e/docs/index.html" target="_blank">d3bef4b</a>
</td>
<td>
jgblanc
</td>
<td>
2021-03-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/b2c89c77e94c6c91507be6d9de05607707dfd766/analysis/index.Rmd" target="_blank">b2c89c7</a>
</td>
<td>
jgblanc
</td>
<td>
2021-03-15
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/Projection_plink2.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/44151c058d1a60c2de19ec65ec3b38ab65c962e4/docs/index.html" target="_blank">44151c0</a>
</td>
<td>
jgblanc
</td>
<td>
2021-03-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/2d14caf606eada4a7f878c269e62793673b7dd5b/analysis/index.Rmd" target="_blank">2d14caf</a>
</td>
<td>
jgblanc
</td>
<td>
2021-03-04
</td>
<td>
projection example
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/be1f7f87a2462914f1fc7bcf770d313a5748b80f/docs/index.html" target="_blank">be1f7f8</a>
</td>
<td>
jgblanc
</td>
<td>
2021-01-31
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/645d3aa8666de71d46e9e090bc4bc90f1911693b/analysis/index.Rmd" target="_blank">645d3aa</a>
</td>
<td>
jgblanc
</td>
<td>
2021-01-31
</td>
<td>
wflow_publish(c(“analysis/Plot_4Popsplit.Rmd”, “analysis/index.Rmd”))
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/25150cd2fa70005c08724bb459d853b0d4465440/docs/index.html" target="_blank">25150cd</a>
</td>
<td>
jgblanc
</td>
<td>
2020-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/675a6155cd1eae8dac8272c65f88fbe6326334de/analysis/index.Rmd" target="_blank">675a615</a>
</td>
<td>
jgblanc
</td>
<td>
2020-12-14
</td>
<td>
Lab meeting
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/911a4621242dedac83b0a17283134accbd45ff61/docs/index.html" target="_blank">911a462</a>
</td>
<td>
jgblanc
</td>
<td>
2020-11-15
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/05d143af52e3ba9319e8e546367b5f27399183cf/analysis/index.Rmd" target="_blank">05d143a</a>
</td>
<td>
jgblanc
</td>
<td>
2020-11-15
</td>
<td>
help
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/dba2c6fd247f284867bfa05ea2a160bd71975896/docs/index.html" target="_blank">dba2c6f</a>
</td>
<td>
jgblanc
</td>
<td>
2020-11-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/fc12aa99cb8dd574a5d74f94bcc370fbe355d2ea/analysis/index.Rmd" target="_blank">fc12aa9</a>
</td>
<td>
jgblanc
</td>
<td>
2020-11-03
</td>
<td>
draft
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/8b1fa832db5d8cf52623306978b8414777f70722/docs/index.html" target="_blank">8b1fa83</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/401d8dd52aefb13d571e5fc0109732544b24bf6b/docs/index.html" target="_blank">401d8dd</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/5b17f8385a152534402ce3b765358a751cdc5668/analysis/index.Rmd" target="_blank">5b17f83</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Theory
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/854d4f7080c4700bbd1f0e700d7372a0c10e8628/docs/index.html" target="_blank">854d4f7</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/f10649af9ea0c80e583d8c9f590ac45a0863ce45/analysis/index.Rmd" target="_blank">f10649a</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Theory
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/e644370277c72670300dc91aa786948370ee88c0/analysis/index.Rmd" target="_blank">e644370</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-15
</td>
<td>
Start workflowr project.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="guaranteeing-unbiasedness-in-selection-tests-based-on-polygenic-scores" class="section level1">
<h1>Guaranteeing unbiasedness in selection tests based on polygenic scores</h1>
<p>Link to manuscript: <a href="https://www.overleaf.com/project/5ff4eec3e5325e0db841f6eb" class="uri">https://www.overleaf.com/project/5ff4eec3e5325e0db841f6eb</a></p>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This website contains all the code used for the analysis.</p>
</div>
<div id="simulation-pipeline" class="section level2">
<h2>Simulation Pipeline</h2>
<p>This pipline is written in snakemake. To run the entire simulation pipeline you simply need to swap in the correct snakefile.</p>
<ul>
<li><code>code/Snakefiles/4PopSplit/snakefile</code> is the correct file to run the 4 population toy model from Figure 1.<br />
</li>
<li><code>code/Snakefiles/SimpleGrid/snakefile</code> runs the stepping stone grid simulations and has options to simulate the confounders and do the polygenic adaptation tests from Figure 2.</li>
</ul>
<p>The detailed methods and results for each simulation can be found <a href="Simulations_methods.html">here</a>.</p>
<p>The pages below explain the 5 major steps of the pipeline:</p>
<ol style="list-style-type: decimal">
<li><a href="Simulate_Genotypes.html">Simulate Genotypes</a></li>
<li><a href="Simulate_Phenotypes.html">Simulate Phenotypes</a></li>
<li><a href="Calculate_TGWAS.html">Project Test Vector</a></li>
<li><a href="Run_GWAS.html">Run GWAS and Ascertain SNPs</a></li>
<li><a href="Polygenic_Selection_Test.html">Polygenic Adaptation Test</a></li>
</ol>
<p><br /> <br /> <br /> <br /> <br /> <br /> <br /> <br /></p>
<p>WIP links<br />
Reproducible figures <a href="Simulation_Methods_Reproducible.html">link</a> Theory <a href="Theory.html">link</a><br />
4popslit <a href="4Popsplit.html">link</a><br />
scratch <a href="scratch.html">link</a><br />
test_vec <a href="Test_vec.html">link</a><br />
Plot_4PopSplit <a href="Plot_4PopSplit.html">link</a><br />
projection_example <a href="projection_example.html">link</a><br />
plink2 projection <a href="Projection_plink2.html">link</a><br />
calc Qx <a href="Calc_Qx.html">link</a><br />
Plot_SimpleGrid <a href="Plot_SimpleGrid.html">link</a><br />
methods <a href="Simulations_methods.html">link</a><br />
Meeting Notes <a href="Meetings_Notes.html">link</a><br />
Proof <a href="Proof.html">link</a></p>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
