<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jennifer Blanc" />


<title>Polygenic_Selection_Test</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Polygenic_Selection_Test</h1>
<h4 class="author">Jennifer Blanc</h4>
<h4 class="date">4/27/2022</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-04-29
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 5 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 2
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201015code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201015)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201015code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201015)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
unnamed-chunk-4
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory <code>Polygenic_Selection_Test_cache</code> and re-run the analysis. To have workflowr automatically delete the cache directory prior to building the file, set <code>delete_cache = TRUE</code> when running <code>wflow_build()</code> or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree931f81811a70a5965859780201a24c6e17e2fab8targetblank931f818a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/931f81811a70a5965859780201a24c6e17e2fab8" target="_blank">931f818</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree931f81811a70a5965859780201a24c6e17e2fab8targetblank931f818a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/931f81811a70a5965859780201a24c6e17e2fab8" target="_blank">931f818</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/
    Ignored:    analysis/Calculate_TGWAS_cache/
    Ignored:    analysis/Polygenic_Selection_Test_cache/
    Ignored:    code/.DS_Store
    Ignored:    code/1kg/.DS_Store
    Ignored:    code/Calculate_Tm/.DS_Store
    Ignored:    code/Debug/.DS_Store
    Ignored:    code/PGA_test/.DS_Store
    Ignored:    code/PRS/.DS_Store
    Ignored:    code/Simulate_Genotypes/.DS_Store
    Ignored:    code/Simulate_Phenotypes/.DS_Store
    Ignored:    code/Snakefiles/.DS_Store
    Ignored:    code/Snakefiles/4PopSplit/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/1kg/
    Ignored:    data/cgd_example/
    Ignored:    data/projection_example/
    Ignored:    data/proof/
    Ignored:    output/.DS_Store
    Ignored:    output/Calculate_TGWAS/
    Ignored:    output/Calculate_Tm/4PopSplit/A10/
    Ignored:    output/Calculate_Tm/4PopSplit/A2/
    Ignored:    output/Calculate_Tm/4PopSplit/B1/
    Ignored:    output/Calculate_Tm/4PopSplit/E2/
    Ignored:    output/Calculate_Tm/4PopSplit/E20/
    Ignored:    output/Calculate_Tm/4PopSplit/E3/
    Ignored:    output/Calculate_Tm/4PopSplit/E4/
    Ignored:    output/Calculate_Tm/4PopSplit/E5/
    Ignored:    output/Calculate_Tm/4PopSplit/E6/
    Ignored:    output/Calculate_Tm/4PopSplit/E7/
    Ignored:    output/Calculate_Tm/4PopSplit/E8/
    Ignored:    output/Calculate_Tm/4PopSplit/E9/
    Ignored:    output/Calculate_Tm/SimpleGrid/A1/
    Ignored:    output/Calculate_Tm/SimpleGrid/B1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E2/
    Ignored:    output/Calculate_Tm/SimpleGrid/E3/
    Ignored:    output/Calculate_Tm/SimpleGrid/E4/
    Ignored:    output/Calculate_Tm/SimpleGrid/E5/
    Ignored:    output/PGA_test/
    Ignored:    output/PRS/
    Ignored:    output/Run_GWAS/
    Ignored:    output/Simulate_Genotypes/.DS_Store
    Ignored:    output/Simulate_Genotypes/4PopSplit/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/A1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E2/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E3/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E4/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E5/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E6/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E7/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E8/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/S1/
    Ignored:    output/Simulate_Phenotypes/

Untracked files:
    Untracked:  _workflowr.yml~
    Untracked:  analysis/Calculate_TGWAS.Rmd
    Untracked:  analysis/Polygenic_Selection_Test.Rmd
    Untracked:  analysis/Run_GWAS.Rmd
    Untracked:  code/Snakefiles/SimpleGrid/.snakemake/
    Untracked:  code/Snakefiles/SimpleGrid/North-South-Meta/
    Untracked:  temp.R

Unstaged changes:
    Modified:   _workflowr.yml
    Modified:   analysis/Midparent.Rmd
    Modified:   analysis/Projection_plink2.Rmd
    Modified:   analysis/Simulate_Genotypes.Rmd
    Modified:   analysis/Simulate_Phenotypes.Rmd
    Modified:   analysis/Simulation_Methods_Reproducible.Rmd
    Modified:   analysis/index.Rmd
    Modified:   code/Calculate_Tm/4PopSplit_make_tvec.R
    Modified:   code/Calculate_Tm/calc_Tm.R
    Modified:   code/Calculate_Tm/format_ID_covars.R
    Modified:   code/PGA_test/4PopSplit_calc_lambdaT.R
    Modified:   code/PGA_test/calc_Qx_4PopSplit.R
    Modified:   code/PRS/clump.R
    Modified:   code/Simulate_Genotypes/generate_genotypes_4PopSplit.py
    Modified:   code/Simulate_Genotypes/split_gwas-test_4PopSplit.R
    Modified:   code/Simulate_Phenotypes/draw_effect_sizes_4PopSplit.R
    Modified:   code/Simulate_Phenotypes/simulate_phenotypes_4PopSplit.R
    Modified:   code/Simulate_Phenotypes/simulate_phenotypes_SimpleGrid.R
    Deleted:    code/Snakefiles/4PopSplit/Signal/snakefile
    Modified:   snakefile

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="polygenic-selection-test" class="section level2">
<h2>Polygenic Selection Test</h2>
<p>Here we will go through the steps to conduct the polygenic selection tests shown in the main text. We will outline how to compute both <span class="math inline">\(Q\)</span> and <span class="math inline">\(Q_x\)</span> as well as our procedure for checking the significance of the statistics.</p>
<div id="calculate-q" class="section level3">
<h3>Calculate Q</h3>
<p>Our statistic of interest is,</p>
<p><span class="math display">\[\hat{Q} := \vec{T}^TX\hat{\vec{\beta}}\]</span></p>
<p>where <span class="math inline">\(\hat{\vec{\beta}}\)</span> is a vector that contains the GWAS effect sizes for the sites ascertained in the previous step and 0 for non-ascertained sites. <span class="math inline">\(X\)</span> is the mean centered genotype matrix of the test panel individuals and <span class="math inline">\(\vec{T}\)</span> is our pre-determined mean-centered test vector. For the 4 Population example presented here we assume the test vector is population ID and will therfore have 0.5 entries for the first population and -0.5 for the second.</p>
<p>First we read in a set of ascertained effect sizes, the genotype matrix at those corresponding sites, and the test vector. Then we calculate <span class="math inline">\(Q\)</span> as defined above. Note that we only include ascertained in <span class="math inline">\(X\)</span> as non-ascertained sites have <span class="math inline">\(\hat{\beta}_{\ell} = 0\)</span>.</p>
<pre class="r"><code># Read in betas 
betas &lt;- fread(&quot;../output/PRS/4PopSplit/B1/C1/h2-0.3/p-0.50/env_0.0/genos-gwas_common-Tm.nc.betas&quot;)

# Read in genotype matrix at ascertained sites and mean center
Xcounts &lt;- read_genos(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common&quot;, betas = betas)
X &lt;- scale(Xcounts, scale = F)

# Read in test vector 
df_Tvec &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/Tvec.txt&quot;)
Tvec &lt;- df_Tvec$Tvec

# Calculate Q 
Q &lt;- Tvec %*% X %*% betas$BETA_Strat</code></pre>
</div>
<div id="calculate-neutral-distribution-of-q" class="section level3">
<h3>Calculate neutral distribution of Q</h3>
<p>To derive the null distribution of <span class="math inline">\(Q\)</span> first consider that the joint distibution of polygenic scores <span class="math inline">\(X\hat{\vec{\beta}}\)</span> is given by,</p>
<p><span class="math display">\[X\hat{\vec{\beta}} \sim MVN(0, \hat{V_a}F)\]</span></p>
<p>where <span class="math inline">\(F\)</span> is an <span class="math inline">\(N \times N\)</span> matrix descibing the correlation structure of allele frequencies relative to the mean frequency and <span class="math inline">\(\hat{V_a} = 2 \sum_{\ell} \hat{\beta}_{\ell} p_{\ell}(1 - p_{\ell})\)</span> is the additive genetic variance of associated sites.</p>
<p>We then multiply the vector of polygenic scores <span class="math inline">\(X\hat{\vec{\beta}}\)</span> by the mean centered test vector <span class="math inline">\(\vec{T}\)</span> to get,</p>
<p><span class="math display">\[\vec{T}^TX\hat{\vec{\beta}} \sim N(0, \hat{V_a}\vec{T}^TF\vec{T})\]</span></p>
<p>where <span class="math inline">\(\vec{T}^TF\vec{T}\)</span> is the amount of variance in <span class="math inline">\(F\)</span> explained by the test vector. Finally we can calculate <span class="math inline">\(F\)</span> by defining <span class="math inline">\(\tilde{X} = XS\)</span> where <span class="math inline">\(S\)</span> is a diagonal matrix such that <span class="math inline">\(s_{\ell \ell} = \frac{1}{\sqrt{2p_{\ell}(1 - p_{\ell})}}\)</span> and computing <span class="math inline">\(F = \frac{1}{L-1}\tilde{X}\tilde{X}^T\)</span>. Therefore, under neutrality, the distibution of <span class="math inline">\(Q\)</span> is</p>
<p><span class="math display">\[\hat{Q} \sim N(0, \frac{\hat{V_a}}{L-1}\vec{T}^T\tilde{X}\tilde{X}^T\vec{T})\]</span></p>
<p>To actually caclulate this additive genetic variance from associated sites, we will first compute <span class="math inline">\(\hat{V_a} = 2 \sum_{\ell} \hat{\beta}_{\ell} p_{\ell}(1 - p_{\ell})\)</span>.</p>
<pre class="r"><code># Calculate allele frequency 
p &lt;- colMeans(Xcounts)/2

# Compute Va
Va &lt;- 2 * sum(betas$BETA_Strat^2 * p * (1 - p))</code></pre>
<p>Next we compute <span class="math inline">\(\vec{T}^T\frac{\tilde{X}\tilde{X}^T}{L-1}\vec{T}\)</span>. Note that we must read in the genotype matrix for all <span class="math inline">\(L\)</span> sites, not just the ascertained ones.</p>
<pre class="r"><code># Read in entire test panel genotype matrix 
pvar &lt;- NewPvar(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.pvar&quot;)
d1 &lt;- NewPgen(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.pgen&quot;)
Xall &lt;- ReadList(d1,seq(1, 10905), meanimpute=F)

# Calculate standardized GRM 
n &lt;- nrow(Xall)
het &lt;- 1 / (2 * colMeans(Xall/2) * (1 - colMeans(Xall/2)))
Tmat &lt;- matrix(-1/n, ncol = n, nrow = n)
diag(Tmat) &lt;- (n-1)/n
cov_mat &lt;- (1/(ncol(Xall))-1) * Tmat %*% t(t(Xall) * het) %*% t(Xall) %*% t(Tmat) 

# Calculate variance explained by test vector
lambdaT &lt;- t(Tvec) %*% cov_mat %*% Tvec</code></pre>
<p>Here we have directly computed the covariance matrix in R. In the simulation pipeline we use the eigen decomposition of the covariance matrix <span class="math inline">\(\vec{T}^TU_X \Lambda_XU_X^T\vec{T}\)</span> from <code>--pca</code> in plink2 to compute the variance explained.</p>
<pre class="r"><code># Load Eigen vectors 
vecs &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/pca.eigenvec&quot;)
vecs &lt;- vecs[,3:ncol(vecs)]
vecs &lt;- apply(vecs, 2, as.numeric)

# Load Eigenvalues
vals &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/pca.eigenval&quot;)
vals &lt;- vals$V1

# Calculate Lambda T
lambda_T &lt;- t(Tvec) %*% vecs %*% diag(vals) %*% t(vecs) %*% Tvec</code></pre>
<p>Now we can put these pieces together to get the variance of <span class="math inline">\(Q\)</span>.</p>
<pre class="r"><code>varQ &lt;- Va * lambda_T</code></pre>
</div>
<div id="assessing-the-significance-of-q" class="section level3">
<h3>Assessing the significance of Q</h3>
<p>Now that we have <span class="math inline">\(Q\)</span> and the variance of <span class="math inline">\(Q\)</span> under neutrality, we can compute a p-value for our selection test. Inutitively, if the value of <span class="math inline">\(Q\)</span> falls into the tails of the distribution, then selection has caused frequencies to diverge at associated sites more than we would expect under neutral drift. First we do a simple 2-tailed test and get the p-value using the variance we derived above.</p>
<pre class="r"><code>pvalNorm &lt;- 2 * (1 - pnorm(abs(Q), mean = 0, sd = sqrt(varQ)))
paste0(&quot;The p-value is &quot;, round(pvalNorm,3))</code></pre>
<pre><code>[1] &quot;The p-value is 0.85&quot;</code></pre>
<p>To avoid relying on the assumption that <span class="math inline">\(X\hat{\vec{\beta}}\)</span> follows a multivariate distribution we also computed p-values using an sign-flipping empirical null. We obtain pseudo-samples by randomizing the sign of the effect sizes and recomputing <span class="math inline">\(Q\)</span>. We then obtain a new p-value,</p>
<p><span class="math display">\[P = 2 * \frac{\sum_i^S I(Q_i &gt; |Q_{True}|)}{S}\]</span></p>
<p>where <span class="math inline">\(S\)</span> is the number of pseudo samples, here set to 1,000, and <span class="math inline">\(I()\)</span> is an indicator function.</p>
<pre class="r"><code># Function to flip betas
flip &lt;- function(betas) {
  new_betas &lt;- sample(c(-1,1), length(betas),  replace = T) * betas
  return(new_betas)
}

# Generate pseudo random samples
qs &lt;- rep(0, 1000)
for (i in 1:1000) {
  b &lt;- flip(betas$BETA_Strat)
  qs[i] &lt;- Tvec %*% X %*% b
}

# Get pvalue
pEN &lt;- 2 * sum(qs &gt; abs(Q[1,1])) / length(qs)
paste0(&quot;That p-value from the empirical null is &quot;, round(pEN, 2))</code></pre>
<pre><code>[1] &quot;That p-value from the empirical null is 0.91&quot;</code></pre>
</div>
<div id="relationship-to-qx" class="section level3">
<h3>Relationship to Qx</h3>
<p>As introduced in <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004412#abstract0">Berg and Coop 2014</a>,</p>
<p><span class="math display">\[Q_x =\frac{(\vec{Z} - \mu)^TF^{-1}(\vec{Z} - \mu)}{2V_a}\]</span></p>
<p>where <span class="math inline">\(\vec{Z}\)</span> is a vector of average population polygenic scores (<span class="math inline">\(Z_i = 2 \sum\limits_{\ell = 1}^L \hat{\beta}_{\ell}p_{\ell}\)</span>). Under neutrality <span class="math inline">\(Q_X\)</span> is expected to follow a chi-squared distribution with <span class="math inline">\(M-1\)</span> degrees of freedom where <span class="math inline">\(M\)</span> is the number of populations. Above we have derived <span class="math inline">\(Q\)</span> in terms of individual level polygenic scores and a single test vector. The relationship between these statistics is</p>
<p><span class="math display">\[\frac{Q^2}{\frac{\hat{V_a}}{L-1}\vec{T}^T\tilde{X}\tilde{X}^T\vec{T}} = Q_X\]</span></p>
<pre class="r"><code># Calculate Qx
Qx &lt;- Q^2 / varQ

# Calculate Chi-square p-value
pCHI &lt;- pchisq(Qx, df=1, lower.tail=FALSE)
paste(&quot;The Chi-squared p-value is &quot;, round(pCHI,  3))</code></pre>
<pre><code>[1] &quot;The Chi-squared p-value is  0.85&quot;</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] pgenlibr_0.3.1    dplyr_1.0.7       data.table_1.14.2

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       pillar_1.6.4     compiler_3.6.2   bslib_0.3.0     
 [5] later_1.3.0      jquerylib_0.1.4  git2r_0.28.0     workflowr_1.6.2 
 [9] tools_3.6.2      digest_0.6.29    jsonlite_1.7.2   evaluate_0.14   
[13] lifecycle_1.0.1  tibble_3.1.6     pkgconfig_2.0.3  rlang_0.4.12    
[17] DBI_1.1.1        yaml_2.2.1       xfun_0.29        fastmap_1.1.0   
[21] stringr_1.4.0    knitr_1.37       generics_0.1.0   fs_1.5.2        
[25] vctrs_0.3.8      sass_0.4.0       rprojroot_2.0.2  tidyselect_1.1.1
[29] glue_1.6.0       R6_2.5.1         fansi_0.5.0      rmarkdown_2.10  
[33] purrr_0.3.4      magrittr_2.0.1   promises_1.2.0.1 ellipsis_0.3.2  
[37] htmltools_0.5.2  assertthat_0.2.1 httpuv_1.6.2     utf8_1.2.2      
[41] stringi_1.7.6    crayon_1.4.2    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
