<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jennifer Blanc" />


<title>Projection_plink2</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fas fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Projection_plink2</h1>
<h4 class="author">Jennifer Blanc</h4>
<h4 class="date">3/10/2021</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-03-15
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201015code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201015)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201015code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201015)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtreeb2c89c77e94c6c91507be6d9de05607707dfd766targetblankb2c89c7a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/b2c89c77e94c6c91507be6d9de05607707dfd766" target="_blank">b2c89c7</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtreeb2c89c77e94c6c91507be6d9de05607707dfd766targetblankb2c89c7a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/b2c89c77e94c6c91507be6d9de05607707dfd766" target="_blank">b2c89c7</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/
    Ignored:    analysis/.DS_Store
    Ignored:    code/.DS_Store
    Ignored:    code/Calculate_Tm/.ipynb_checkpoints/
    Ignored:    data/projection_example/
    Ignored:    output/Calculate_Tm/
    Ignored:    output/PRS/4PopSplit/S1/C1/genos-test_common.c.log
    Ignored:    output/PRS/4PopSplit/S1/C1/genos-test_common.c.p.log
    Ignored:    output/PRS/4PopSplit/S1/C1/genos-test_common.nc.log
    Ignored:    output/PRS/4PopSplit/S1/C2/genos-test_common.c.log
    Ignored:    output/PRS/4PopSplit/S1/C2/genos-test_common.c.p.log
    Ignored:    output/PRS/4PopSplit/S1/C2/genos-test_common.nc.log
    Ignored:    output/PRS/4PopSplit/S2/C1/genos-test_common.c.log
    Ignored:    output/PRS/4PopSplit/S2/C1/genos-test_common.c.p.log
    Ignored:    output/PRS/4PopSplit/S2/C1/genos-test_common.nc.log
    Ignored:    output/PRS/4PopSplit/S2/C2/genos-test_common.c.log
    Ignored:    output/PRS/4PopSplit/S2/C2/genos-test_common.c.p.log
    Ignored:    output/PRS/4PopSplit/S2/C2/genos-test_common.nc.log
    Ignored:    output/PRS/4PopSplit/S3/
    Ignored:    output/PRS/4PopSplit/S4/
    Ignored:    output/PRS/4PopSplit/S5/
    Ignored:    output/PRS/4PopSplit/S6/
    Ignored:    output/PRS/4PopSplit/S7/
    Ignored:    output/PRS/4PopSplit/S8/
    Ignored:    output/PRS/4PopSplit/S9/
    Ignored:    output/Run_GWAS/
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/common_snp_ids.txt
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test-big.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/common_snp_ids.txt
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-gwas.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-gwas_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-gwas_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-gwas_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-gwas_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-gwas_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-test-big.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-test.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-test_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-test_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-test_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-test_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/C2/genos-test_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/ff.txt
    Ignored:    output/Simulate_Genotypes/4PopSplit/S1/genos.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/common_snp_ids.txt
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test-big.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/common_snp_ids.txt
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-gwas.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-gwas_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-gwas_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-gwas_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-gwas_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-gwas_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-test-big.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-test.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-test_common.afreq
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-test_common.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-test_common.pgen
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-test_common.psam
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/C2/genos-test_common.pvar
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/ff.txt
    Ignored:    output/Simulate_Genotypes/4PopSplit/S2/genos.log
    Ignored:    output/Simulate_Genotypes/4PopSplit/S3/
    Ignored:    output/Simulate_Genotypes/4PopSplit/S4/
    Ignored:    output/Simulate_Genotypes/4PopSplit/S5/
    Ignored:    output/Simulate_Genotypes/4PopSplit/S6/
    Ignored:    output/Simulate_Genotypes/4PopSplit/S7/
    Ignored:    output/Simulate_Genotypes/4PopSplit/S8/
    Ignored:    output/Simulate_Genotypes/4PopSplit/S9/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S1/C1/genos-gwas_common.gvalue.log
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S1/C2/genos-gwas_common.gvalue.log
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S2/C1/genos-gwas_common.gvalue.log
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S2/C2/genos-gwas_common.gvalue.log
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S3/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S4/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S5/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S6/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S7/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S8/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/S9/

Unstaged changes:
    Modified:   analysis/projection_example.Rmd
    Deleted:    output/PRS/4PopSplit/S1/C1/genos-gwas_common-Tm.c.betas
    Deleted:    output/PRS/4PopSplit/S1/C1/genos-gwas_common-Tm.c.p.betas
    Deleted:    output/PRS/4PopSplit/S1/C1/genos-gwas_common-Tm.nc.betas
    Modified:   output/PRS/4PopSplit/S1/C1/genos-gwas_common.c.betas
    Modified:   output/PRS/4PopSplit/S1/C1/genos-gwas_common.c.p.betas
    Modified:   output/PRS/4PopSplit/S1/C1/genos-gwas_common.nc.betas
    Deleted:    output/PRS/4PopSplit/S1/C1/genos-test_common-Tm.c.p.sscore
    Deleted:    output/PRS/4PopSplit/S1/C1/genos-test_common-Tm.c.sscore
    Deleted:    output/PRS/4PopSplit/S1/C1/genos-test_common-Tm.nc.sscore
    Modified:   output/PRS/4PopSplit/S1/C1/genos-test_common.c.p.sscore
    Modified:   output/PRS/4PopSplit/S1/C1/genos-test_common.c.sscore
    Modified:   output/PRS/4PopSplit/S1/C1/genos-test_common.nc.sscore
    Deleted:    output/PRS/4PopSplit/S1/C1/genos-test_common.true.sscore
    Deleted:    output/PRS/4PopSplit/S1/C2/genos-gwas_common-Tm.c.betas
    Deleted:    output/PRS/4PopSplit/S1/C2/genos-gwas_common-Tm.c.p.betas
    Deleted:    output/PRS/4PopSplit/S1/C2/genos-gwas_common-Tm.nc.betas
    Modified:   output/PRS/4PopSplit/S1/C2/genos-gwas_common.c.betas
    Modified:   output/PRS/4PopSplit/S1/C2/genos-gwas_common.c.p.betas
    Modified:   output/PRS/4PopSplit/S1/C2/genos-gwas_common.nc.betas
    Deleted:    output/PRS/4PopSplit/S1/C2/genos-test_common-Tm.c.p.sscore
    Deleted:    output/PRS/4PopSplit/S1/C2/genos-test_common-Tm.c.sscore
    Deleted:    output/PRS/4PopSplit/S1/C2/genos-test_common-Tm.nc.sscore
    Modified:   output/PRS/4PopSplit/S1/C2/genos-test_common.c.p.sscore
    Modified:   output/PRS/4PopSplit/S1/C2/genos-test_common.c.sscore
    Modified:   output/PRS/4PopSplit/S1/C2/genos-test_common.nc.sscore
    Deleted:    output/PRS/4PopSplit/S1/C2/genos-test_common.true.sscore
    Modified:   output/PRS/4PopSplit/S2/C1/genos-gwas_common.c.betas
    Modified:   output/PRS/4PopSplit/S2/C1/genos-gwas_common.c.p.betas
    Modified:   output/PRS/4PopSplit/S2/C1/genos-gwas_common.nc.betas
    Modified:   output/PRS/4PopSplit/S2/C1/genos-test_common.c.p.sscore
    Modified:   output/PRS/4PopSplit/S2/C1/genos-test_common.c.sscore
    Modified:   output/PRS/4PopSplit/S2/C1/genos-test_common.nc.sscore
    Modified:   output/PRS/4PopSplit/S2/C2/genos-gwas_common.c.betas
    Modified:   output/PRS/4PopSplit/S2/C2/genos-gwas_common.c.p.betas
    Modified:   output/PRS/4PopSplit/S2/C2/genos-gwas_common.nc.betas
    Modified:   output/PRS/4PopSplit/S2/C2/genos-test_common.c.p.sscore
    Modified:   output/PRS/4PopSplit/S2/C2/genos-test_common.c.sscore
    Modified:   output/PRS/4PopSplit/S2/C2/genos-test_common.nc.sscore
    Modified:   output/Simulate_Phenotypes/4PopSplit/S1/C1/genos-gwas_common.effects.txt
    Modified:   output/Simulate_Phenotypes/4PopSplit/S1/C1/genos-gwas_common.gvalue.sscore
    Modified:   output/Simulate_Phenotypes/4PopSplit/S1/C1/genos-gwas_common.phenos.txt
    Modified:   output/Simulate_Phenotypes/4PopSplit/S1/C2/genos-gwas_common.effects.txt
    Modified:   output/Simulate_Phenotypes/4PopSplit/S1/C2/genos-gwas_common.gvalue.sscore
    Modified:   output/Simulate_Phenotypes/4PopSplit/S1/C2/genos-gwas_common.phenos.txt
    Modified:   output/Simulate_Phenotypes/4PopSplit/S2/C1/genos-gwas_common.effects.txt
    Modified:   output/Simulate_Phenotypes/4PopSplit/S2/C1/genos-gwas_common.gvalue.sscore
    Modified:   output/Simulate_Phenotypes/4PopSplit/S2/C1/genos-gwas_common.phenos.txt
    Modified:   output/Simulate_Phenotypes/4PopSplit/S2/C2/genos-gwas_common.effects.txt
    Modified:   output/Simulate_Phenotypes/4PopSplit/S2/C2/genos-gwas_common.gvalue.sscore
    Modified:   output/Simulate_Phenotypes/4PopSplit/S2/C2/genos-gwas_common.phenos.txt
    Modified:   snakefile

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Projection_plink2.Rmd</code>) and HTML (<code>docs/Projection_plink2.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/b2c89c77e94c6c91507be6d9de05607707dfd766/analysis/Projection_plink2.Rmd" target="_blank">b2c89c7</a>
</td>
<td>
jgblanc
</td>
<td>
2021-03-15
</td>
<td>
wflow_publish(c(“analysis/index.Rmd”, “analysis/Projection_plink2.Rmd”))
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="projecting-the-test-vector-using-plink2" class="section level2">
<h2>Projecting the Test vector using Plink2</h2>
<p><span class="math display">\[T_m = \frac{M X^T (XX^T)^{-1}T}{L^2}\]</span></p>
<pre class="r"><code># Read GWAS Matrix 
pvar &lt;- NewPvar(&quot;../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.pvar&quot;)
d1 &lt;- NewPgen(&quot;../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.pgen&quot;)
M &lt;- ReadList(d1,seq(1,95974), meanimpute=F)
M &lt;- scale(M)

# Read in Test Matrix 
pvar &lt;- NewPvar(&quot;../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.pvar&quot;)
d1 &lt;- NewPgen(&quot;../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.pgen&quot;)
X &lt;- ReadList(d1,seq(1,95974), meanimpute=F)
X &lt;- scale(X)

# Read in Test Vector 
#Tvec &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/S1/C1/Tvec.txt&quot;)$V1
L &lt;- ncol(X)

tvec &lt;- runif(200)
ctvec &lt;- (tvec-mean(tvec))
Tvec &lt;- ctvec/sqrt(sum(ctvec^2))
head(Tvec)</code></pre>
<pre><code>[1] -0.109849080 -0.046266741  0.116218351 -0.104822657  0.037839696
[6] -0.003246035</code></pre>
<p>Calculate Tm using SVD</p>
<pre class="r"><code># Do SVD
s &lt;- svd(X)
u &lt;- s$u
d &lt;- s$d
v &lt;- s$v

# Square singular values and divide by L
svd_vals &lt;- (d^2 / L)
n &lt;- nrow(X)

# Calculate Tm
K = (M %*% t(X)) / (L)
Tm = K %*% u[,1:(n-1)] %*% diag(1/svd_vals[1:(n-1)]) %*% t(u[,1:(n-1)]) %*% Tvec</code></pre>
<p>Calculate Tm via projecting each eigenvector individually and weighting by the Betas</p>
<pre class="r"><code>betas &lt;- t(u) %*% Tvec

tmp &lt;- M %*% t(X)%*% u[,1:(n-1)] %*% diag(1/svd_vals[1:(n-1)]) %*% t(u[,1:(n-1)]) %*% u[,1:(n-1)]
Tm_weighted &lt;- tmp %*% betas[1:(n-1)] / L

plot(Tm, Tm_weighted)
abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<p><img src="figure/Projection_plink2.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Calculate Tm via projecting each eigenvector individually using right singular values (loadings) and weighting by the Betas</p>
<pre class="r"><code>tmp &lt;- M %*% v[,1:(n-1)] %*% diag(1/d[1:(n-1)])
Tm_weighted_v &lt;- tmp %*% betas[1:(n-1)] 

plot(Tm, Tm_weighted_v)
abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<p><img src="figure/Projection_plink2.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Do PCA in plink (including loadings)</p>
<pre class="bash"><code>~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common \
       --pca biallelic-var-wts 199 \
       --out ../data/projection_example/S2_PC</code></pre>
<pre><code>PLINK v2.00a3 AVX2 (25 Feb 2020)               www.cog-genomics.org/plink/2.0/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to ../data/projection_example/S2_PC.log.
Options in effect:
  --out ../data/projection_example/S2_PC
  --pca biallelic-var-wts 199
  --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common

Start time: Mon Mar 15 12:21:20 2021
8192 MiB RAM detected; reserving 4096 MiB for main workspace.
Using up to 8 compute threads.
200 samples (0 females, 0 males, 200 ambiguous; 200 founders) loaded from
../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.psam.
95974 variants loaded from
../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.pvar.
Note: No phenotype data present.
Calculating allele frequencies... 0%68%done.
Constructing GRM: 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.
Extracting eigenvalues and eigenvectors... done.
--pca: Variant weights written to ../data/projection_example/S2_PC.eigenvec.var
.
--pca: Eigenvectors written to ../data/projection_example/S2_PC.eigenvec , and
eigenvalues written to ../data/projection_example/S2_PC.eigenval .
End time: Mon Mar 15 12:21:21 2021</code></pre>
<p>Calculating Tm, creating pseudoinverse with plink vecs</p>
<pre class="r"><code>plink_vecs &lt;-fread(&quot;../data/projection_example/S2_PC.eigenvec&quot;)[,3:201]
plink_vecs &lt;- apply(plink_vecs, 2, as.numeric)
plink_vals &lt;- fread(&quot;../data/projection_example/S2_PC.eigenval&quot;)$V1

betas_plink &lt;- t(plink_vecs) %*% Tvec
tmp &lt;- M %*% t(X)%*% plink_vecs[,1:(n-1)] %*% diag(1/plink_vals[1:(n-1)]) %*% t(plink_vecs[,1:(n-1)]) %*% plink_vecs[,1:(n-1)]
Tm_plink &lt;- tmp %*% betas_plink[1:(n-1)] / L

plot(Tm, Tm_plink)
abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<p><img src="figure/Projection_plink2.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Calculate Tm, using loadings from plink</p>
<pre class="r"><code>plink_loadings &lt;- fread(&quot;../data/projection_example/S2_PC.eigenvec.var&quot;)[,5:203]
plink_loadings &lt;- apply(plink_loadings, 2, as.numeric)

tmp &lt;- M %*% plink_loadings %*% diag(1/sqrt(plink_vals[1:(n-1)])) 
Tm_plink &lt;- tmp %*% diag(1/sqrt(plink_vals)) %*% betas_plink[1:(n-1)] 

plot(Tm, Tm_plink)
abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<p><img src="figure/Projection_plink2.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Calculate using projection step from plink2</p>
<pre class="bash"><code>~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common \
       --pca allele-wts 199 \
       --out ../data/projection_example/S2_PC_proj

~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common \
       --score ../data/projection_example/S2_PC_proj.eigenvec.allele 2 5 header-read no-mean-imputation \
               variance-standardize \
       --score-col-nums 6-204 \
       --out ../data/projection_example/S2_projection</code></pre>
<pre><code>PLINK v2.00a3 AVX2 (25 Feb 2020)               www.cog-genomics.org/plink/2.0/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to ../data/projection_example/S2_PC_proj.log.
Options in effect:
  --out ../data/projection_example/S2_PC_proj
  --pca allele-wts 199
  --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common

Start time: Mon Mar 15 12:21:30 2021
8192 MiB RAM detected; reserving 4096 MiB for main workspace.
Using up to 8 compute threads.
200 samples (0 females, 0 males, 200 ambiguous; 200 founders) loaded from
../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.psam.
95974 variants loaded from
../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common.pvar.
Note: No phenotype data present.
Calculating allele frequencies... 0%68%done.
Constructing GRM: 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.
Extracting eigenvalues and eigenvectors... done.
--pca: Allele weights written to
../data/projection_example/S2_PC_proj.eigenvec.allele .
--pca: Eigenvectors written to ../data/projection_example/S2_PC_proj.eigenvec ,
and eigenvalues written to ../data/projection_example/S2_PC_proj.eigenval .
End time: Mon Mar 15 12:21:31 2021
PLINK v2.00a3 AVX2 (25 Feb 2020)               www.cog-genomics.org/plink/2.0/
(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to ../data/projection_example/S2_projection.log.
Options in effect:
  --out ../data/projection_example/S2_projection
  --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common
  --score ../data/projection_example/S2_PC_proj.eigenvec.allele 2 5 header-read no-mean-imputation variance-standardize
  --score-col-nums 6-204

Start time: Mon Mar 15 12:21:31 2021
8192 MiB RAM detected; reserving 4096 MiB for main workspace.
Using up to 8 compute threads.
200 samples (0 females, 0 males, 200 ambiguous; 200 founders) loaded from
../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.psam.
95974 variants loaded from
../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common.pvar.
Note: No phenotype data present.
Calculating allele frequencies... 0%68%done.

--score: 10k variants loaded.
--score: 20k variants loaded.
--score: 30k variants loaded.
--score: 40k variants loaded.
--score: 50k variants loaded.
--score: 60k variants loaded.
--score: 70k variants loaded.
--score: 80k variants loaded.
--score: 90k variants loaded.
--score: 95974 variants processed.
--score: Results written to ../data/projection_example/S2_projection.sscore .
End time: Mon Mar 15 12:21:32 2021</code></pre>
<pre class="r"><code>proj_vecs_plink &lt;- fread(&quot;../data/projection_example/S2_projection.sscore&quot;)[,5:203]
proj_vecs_plink &lt;- apply(proj_vecs_plink, 2, as.numeric)

Tm_plink &lt;- proj_vecs_plink %*% diag(1/sqrt(plink_vals)) %*% betas_plink[1:199] 

plot(Tm, Tm_plink)
abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<p><img src="figure/Projection_plink2.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Calculate Tm via plink2</p>
<pre class="bash"><code>#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
#       --pca allele-wts 199 \
#       --out ../data/projection_example/S1_PC</code></pre>
<pre class="bash"><code>#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
#       --score ../data/projection_example/S1_PC.eigenvec.allele 2 5 header-read no-mean-imputation \
#               variance-standardize \
#       --score-col-nums 6-204 \
#       --out ../data/projection_example/S1_projection</code></pre>
<pre class="bash"><code>#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common \
#       --score ../data/projection_example/S1_PC.eigenvec.allele 2 5 header-read no-mean-imputation \
#               variance-standardize \
#       --score-col-nums 6-204 \
#       --out ../data/projection_example/S1_projection</code></pre>
<pre class="r"><code>#plink_vecs &lt;- fread(&quot;../data/projection_example/S1_PC.eigenvec&quot;)[,3:201]
#plink_vecs &lt;- apply(plink_vecs, 2, as.numeric)

#betas &lt;- t(plink_vecs) %*% Tvec

#proj_vecs_plink &lt;- fread(&quot;../data/projection_example/S1_projection.sscore&quot;)[,5:203]
#proj_vecs_plink &lt;- apply(proj_vecs_plink, 2, as.numeric)
#plink_vals &lt;- fread(&quot;../data/projection_example/S1_PC.eigenval&quot;)
#tst &lt;- proj_vecs_plink %*%  diag(1/sqrt(plink_vals$V1))

#Tm_plink &lt;- tst %*% betas[1:199] 

#plot(Tm, -1.8 * Tm_plink)
#abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<pre class="r"><code># plink_loadings &lt;- fread(&quot;../data/projection_example/S1_PC.eigenvec.var&quot;)[,5:203]
# plink_loadings &lt;- apply(plink_loadings, 2, as.numeric)
# W &lt;- scale(M) %*% plink_loadings 
# betas &lt;- t(plink_vecs) %*% Tvec
# Tm_man &lt;- W %*% betas[1:199] / (L)
# 
# plot(Tm_plink, Tm_man)
# abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<pre class="bash"><code>#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
#       --score ../data/projection_example/S1_PC.eigenvec.allele 2 5 header-read no-mean-imputation \
#               variance-standardize \
#       --score-col-nums 6-204 \
#       --out ../data/projection_example/S1_projection_test</code></pre>
<pre class="r"><code># plink_vecs &lt;- fread(&quot;../data/projection_example/S1_PC.eigenvec&quot;)[,3:201]
# plink_vecs &lt;- apply(plink_vecs, 2, as.numeric)
# 
# betas &lt;- t(plink_vecs) %*% Tvec
# 
# proj_vecs_plink &lt;- fread(&quot;../data/projection_example/S1_projection_test.sscore&quot;)[,5:203]
# proj_vecs_plink &lt;- apply(proj_vecs_plink, 2, as.numeric)
# plink_vals &lt;- fread(&quot;../data/projection_example/S1_PC.eigenval&quot;)
# tst &lt;- proj_vecs_plink %*%  diag(1/sqrt(plink_vals$V1))
# 
# Tm_plink_t &lt;- tst %*% betas[1:199] 
# 
# plot(-2 *Tm_plink_t, Tvec)
# abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<pre class="bash"><code># Make GRM
#~/Desktop/gcta64 --bfile ../data/projection_example/genos-test_common --make-grm --out ../data/projection_example/GRM_S1

# Do PCA
#~/Desktop/gcta64 --grm ../data/projection_example/GRM_S1 --pca 199 --out ../data/projection_example/GRM_S1

# Get Loadings for n-1 PCs in test panel  
#~/Desktop/gcta64 --bfile ../data/projection_example/genos-test_common --pc-loading ../data/projection_example/GRM_S1 --out ../data/projection_example/GCTA_loadings</code></pre>
<pre class="bash"><code># Project n-1 PCs into GWAS panel
#~/Desktop/gcta64 --bfile ../data/projection_example/genos-test_common --project-loading  #../data/projection_example/GCTA_loadings 199 --out ../data/projection_example/projection_gcta</code></pre>
<pre class="r"><code># gcta_vecs &lt;- fread(&quot;../data/projection_example/GRM_S1.eigenvec&quot;)[,3:201]
# gcta_vecs &lt;- apply(gcta_vecs, 2, as.numeric)
# 
# betas &lt;- t(gcta_vecs) %*% Tvec
# 
# proj_vecs &lt;- fread(&quot;../data/projection_example/projection_gcta.proj.eigenvec&quot;)[,3:201]
# proj_vecs &lt;- apply(proj_vecs, 2, as.numeric)
# 
# Tm_gcta &lt;- proj_vecs %*% betas 
# 
# plot(Tvec, Tm_gcta)
# abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<pre class="r"><code># plot(Tm, -2 * Tm_plink)
# abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<pre class="r"><code># allele &lt;- fread(&quot;../data/projection_example/S2_PC.eigenvec.allele&quot;)
# allele_random &lt;- allele %&gt;% group_by(ID) %&gt;% slice(1) %&gt;%ungroup
# allele_random &lt;- allele_random[,6:204]
# allele_random &lt;- apply(allele_random, 2, as.numeric)
# 
# tmp &lt;- X %*% allele_random %*% diag(1/sqrt(plink_vals[1:199]))
# Tm_plink &lt;- tmp %*% betas_plink[1:199] 
# 
# plot(Tvec, Tm_plink)
# abline(a = 0, b =1 , col = &quot;red&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.14.0 pgenlibr_0.3.1    workflowr_1.6.2  

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6        whisker_0.4       knitr_1.31        magrittr_2.0.1   
 [5] R6_2.5.0          rlang_0.4.10      fansi_0.4.2       highr_0.8        
 [9] stringr_1.4.0     tools_3.6.2       xfun_0.21         utf8_1.1.4       
[13] git2r_0.28.0      htmltools_0.5.1.1 ellipsis_0.3.1    rprojroot_2.0.2  
[17] yaml_2.2.1        digest_0.6.27     tibble_3.1.0      lifecycle_1.0.0  
[21] crayon_1.4.1      later_1.1.0.1     vctrs_0.3.6       promises_1.2.0.1 
[25] fs_1.5.0          glue_1.4.2        evaluate_0.14     rmarkdown_2.7    
[29] stringi_1.5.3     compiler_3.6.2    pillar_1.5.0      httpuv_1.5.5     
[33] pkgconfig_2.0.3  </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
