<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jennifer Blanc" />


<title>Run_GWAS</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Run_GWAS</h1>
<h4 class="author">Jennifer Blanc</h4>
<h4 class="date">4/26/2022</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-04-26
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201015code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201015)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201015code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201015)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree931f81811a70a5965859780201a24c6e17e2fab8targetblank931f818a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/931f81811a70a5965859780201a24c6e17e2fab8" target="_blank">931f818</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree931f81811a70a5965859780201a24c6e17e2fab8targetblank931f818a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/931f81811a70a5965859780201a24c6e17e2fab8" target="_blank">931f818</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/
    Ignored:    analysis/Calculate_TGWAS_cache/
    Ignored:    analysis/figure/
    Ignored:    code/.DS_Store
    Ignored:    code/1kg/.DS_Store
    Ignored:    code/Calculate_Tm/.DS_Store
    Ignored:    code/Debug/.DS_Store
    Ignored:    code/PGA_test/.DS_Store
    Ignored:    code/PRS/.DS_Store
    Ignored:    code/Simulate_Genotypes/.DS_Store
    Ignored:    code/Simulate_Phenotypes/.DS_Store
    Ignored:    code/Snakefiles/.DS_Store
    Ignored:    code/Snakefiles/4PopSplit/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/1kg/
    Ignored:    data/cgd_example/
    Ignored:    data/projection_example/
    Ignored:    data/proof/
    Ignored:    output/.DS_Store
    Ignored:    output/Calculate_TGWAS/
    Ignored:    output/Calculate_Tm/4PopSplit/A10/
    Ignored:    output/Calculate_Tm/4PopSplit/A2/
    Ignored:    output/Calculate_Tm/4PopSplit/B1/
    Ignored:    output/Calculate_Tm/4PopSplit/E2/
    Ignored:    output/Calculate_Tm/4PopSplit/E20/
    Ignored:    output/Calculate_Tm/4PopSplit/E3/
    Ignored:    output/Calculate_Tm/4PopSplit/E4/
    Ignored:    output/Calculate_Tm/4PopSplit/E5/
    Ignored:    output/Calculate_Tm/4PopSplit/E6/
    Ignored:    output/Calculate_Tm/4PopSplit/E7/
    Ignored:    output/Calculate_Tm/4PopSplit/E8/
    Ignored:    output/Calculate_Tm/4PopSplit/E9/
    Ignored:    output/Calculate_Tm/SimpleGrid/A1/
    Ignored:    output/Calculate_Tm/SimpleGrid/B1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E2/
    Ignored:    output/Calculate_Tm/SimpleGrid/E3/
    Ignored:    output/Calculate_Tm/SimpleGrid/E4/
    Ignored:    output/Calculate_Tm/SimpleGrid/E5/
    Ignored:    output/PGA_test/
    Ignored:    output/PRS/
    Ignored:    output/Run_GWAS/
    Ignored:    output/Simulate_Genotypes/.DS_Store
    Ignored:    output/Simulate_Genotypes/4PopSplit/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/A1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E2/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E3/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E4/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E5/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E6/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E7/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E8/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/S1/
    Ignored:    output/Simulate_Phenotypes/

Untracked files:
    Untracked:  _workflowr.yml~
    Untracked:  analysis/Calculate_TGWAS.Rmd
    Untracked:  analysis/Run_GWAS.Rmd
    Untracked:  code/Snakefiles/SimpleGrid/North-South-Meta/
    Untracked:  plink2.acount
    Untracked:  plink2.fst.summary
    Untracked:  plink2.log
    Untracked:  temp.R

Unstaged changes:
    Modified:   _workflowr.yml
    Modified:   analysis/Midparent.Rmd
    Modified:   analysis/Projection_plink2.Rmd
    Modified:   analysis/Simulate_Genotypes.Rmd
    Modified:   analysis/Simulate_Phenotypes.Rmd
    Modified:   analysis/Simulation_Methods_Reproducible.Rmd
    Modified:   analysis/index.Rmd
    Modified:   code/Calculate_Tm/4PopSplit_make_tvec.R
    Modified:   code/Calculate_Tm/calc_Tm.R
    Modified:   code/Calculate_Tm/format_ID_covars.R
    Modified:   code/PRS/clump.R
    Modified:   code/Simulate_Genotypes/generate_genotypes_4PopSplit.py
    Modified:   code/Simulate_Genotypes/split_gwas-test_4PopSplit.R
    Modified:   code/Simulate_Phenotypes/draw_effect_sizes_4PopSplit.R
    Modified:   code/Simulate_Phenotypes/simulate_phenotypes_4PopSplit.R
    Deleted:    code/Snakefiles/4PopSplit/Signal/snakefile
    Modified:   snakefile

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="run-gwas-and-ascertain-snps" class="section level2">
<h2>Run GWAS and Ascertain SNPs</h2>
<p>Here we will go through the steps to run the 3 different GWASs shown in the main text and outline our procedure to select SNPs to include in the polygenic score estimation done in the selection test.</p>
<div id="format-covariates" class="section level3">
<h3>Format Covariates</h3>
<p>We want to run 3 separate GWASs:</p>
<ol style="list-style-type: decimal">
<li>No correction<br />
</li>
<li>Including <span class="math inline">\(\vec{T}^{GWAS}\)</span><br />
</li>
<li>Including population ID</li>
</ol>
<p>This will allow us to compare our correction method, <span class="math inline">\(\vec{T}^{GWAS}\)</span>, to both no using any correction method and to the best case scenario of including the confounder (in this case population ID).</p>
<p>First we will format a covariate file that includes <span class="math inline">\(\vec{T}^{GWAS}\)</span> and population ID.</p>
<pre class="python"><code># Snakemake rule
rule format_covars:
    input:
      pops=&quot;output/Simulate_Genotypes/4PopSplit/{rep}/genos.pop&quot;,
      fam=&quot;output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-gwas_common.psam&quot;,
      Tm=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tm.txt&quot;
    output:
      &quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tm-ID_covars.txt&quot;,
    shell:
      &quot;Rscript code/Calculate_Tm/format_ID_covars.R {input.pops} {input.Tm} {input.fam} {output}&quot;</code></pre>
<pre class="r"><code># Read in GWAS panel information
pops &lt;- fread(&quot;../output/Simulate_Genotypes/4PopSplit/B1/genos.pop&quot;, header = T)
fam &lt;- fread(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.psam&quot;)

# Read in TGWAS
Tm &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/Tm.txt&quot;)

# Format covariate file
tmp &lt;- dplyr::inner_join(pops, fam, by = c(&quot;IID&quot;= &quot;IID&quot;)) %&gt;% select(&quot;IID&quot;, &quot;POP&quot;)
df &lt;- as.data.frame(cbind(tmp$IID, tmp$IID, Tm, tmp$POP))
colnames(df) &lt;- c(&quot;FID&quot;,&quot;IID&quot;, &quot;Tm&quot;, &quot;PopID&quot;)
pops &lt;- unique(df$PopID)
df &lt;- df %&gt;% mutate(PopID = case_when((PopID == pops[1]) ~ 1, (PopID == pops[2]) ~ 0))

kable(head(df))</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">FID</th>
<th align="left">IID</th>
<th align="right">Tm</th>
<th align="right">PopID</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">tsk_0</td>
<td align="left">tsk_0</td>
<td align="right">0.2734113</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">tsk_1</td>
<td align="left">tsk_1</td>
<td align="right">0.2863926</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">tsk_2</td>
<td align="left">tsk_2</td>
<td align="right">0.2909277</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">tsk_3</td>
<td align="left">tsk_3</td>
<td align="right">0.2087834</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">tsk_4</td>
<td align="left">tsk_4</td>
<td align="right">0.1100468</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">tsk_5</td>
<td align="left">tsk_5</td>
<td align="right">0.3162004</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
<div id="run-gwas" class="section level3">
<h3>Run GWAS</h3>
<p>We first want to run a GWAS with no correction using the model below where <span class="math inline">\(\vec{Y}\)</span> are phenotypes and <span class="math inline">\(\vec{g}_{\ell}\)</span> are the genotype counts of the alternate allele.<br />
<span class="math display">\[\vec{Y} \sim \vec{g}_{\ell}\]</span></p>
<p>Here we demonstrate how this GWAS can be done using R and show that it is equivalent to using <code>--glm</code> function in plink2. We do not that plink2 returns the effect size estimate of the minor allele rather than the alternate allele. We account for this fact by flipping the sign of the effect size when ascertaining SNPs in the next step.</p>
<pre class="r"><code># Read in GWAS Matrix 
pvar &lt;- NewPvar(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pvar&quot;)
d1 &lt;- NewPgen(&quot;../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pgen&quot;)
G &lt;- ReadList(d1,seq(1,10905), meanimpute=F)

# Read in phenotypes 
phenos &lt;- fread(&quot;../output/Simulate_Phenotypes/4PopSplit/B1/C1/h2-0.3/p-0.50/env_0.0/genos-gwas_common.phenos.txt&quot;)

# Run GWAS 
betas &lt;- rep(0, ncol(G))
for (i in 1:ncol(G)) {
  mod &lt;- lm(phenos$pheno_strat ~ G[,i])
  betas[i] &lt;- coef(mod)[2]
}

# Read in betas from plink 
plinkB &lt;- fread(&quot;../output/Run_GWAS/4PopSplit/B1/C1/h2-0.3/p-0.50/env_0.0/genos-gwas_common.pheno_strat.glm.linear&quot;)

# Copmare to estimated effect sizes from plink 
plot(abs(betas), abs(plinkB$BETA), xlab = &quot;Estimated Effect Sizes&quot;,  ylab = &quot;Plink estimated effect sizes&quot;)
abline(a = 0, b = 1, col = &quot;red&quot;)</code></pre>
<p><img src="figure/Run_GWAS.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We then repeat the GWAS using our covariate <span class="math inline">\(\vec{T}^{GWAS}\)</span>.<br />
<span class="math display">\[\vec{Y} \sim \vec{g}_{\ell} + \vec{T}^{GWAS}\]</span></p>
<pre class="r"><code># Read in covariate file 
covars &lt;- fread(&quot;../output/Calculate_Tm/4PopSplit/B1/C1/Tm-ID_covars.txt&quot;)

# Run GWAS 
betas &lt;- rep(0, ncol(G))
for (i in 1:ncol(G)) {
  mod &lt;- lm(phenos$pheno_strat ~ G[,i] + covars$Tm)
  betas[i] &lt;- coef(mod)[2]
}

# Read in betas from plink 
plinkB &lt;- fread(&quot;../output/Run_GWAS/4PopSplit/B1/C1/h2-0.3/p-0.50/env_0.0/genos-gwas_common-Tm.pheno_strat.glm.linear&quot;)

# Copmare to estimated effect sizes from plink 
plot(abs(betas), abs(plinkB$BETA), xlab = &quot;Estimated Effect Sizes&quot;,  ylab = &quot;Plink estimated effect sizes&quot;)
abline(a = 0, b = 1, col = &quot;red&quot;)</code></pre>
<p><img src="figure/Run_GWAS.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Finally, we repeat the GWAS a third time using population ID as a covariate.<br />
<span class="math display">\[\vec{Y} \sim \vec{g}_{\ell} + \vec{\text{ID}}\]</span></p>
<pre class="r"><code># Run GWAS 
betas &lt;- rep(0, ncol(G))
for (i in 1:ncol(G)) {
  mod &lt;- lm(phenos$pheno_strat ~ G[,i] + covars$PopID)
  betas[i] &lt;- coef(mod)[2]
}

# Read in betas from plink 
plinkB &lt;- fread(&quot;../output/Run_GWAS/4PopSplit/B1/C1/h2-0.3/p-0.50/env_0.0/genos-gwas_common-ID.pheno_strat.glm.linear&quot;)

# Copmare to estimated effect sizes from plink 
plot(abs(betas), abs(plinkB$BETA), xlab = &quot;Estimated Effect Sizes&quot;,  ylab = &quot;Plink estimated effect sizes&quot;)
abline(a = 0, b = 1, col = &quot;red&quot;)</code></pre>
<p><img src="figure/Run_GWAS.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Here are the snakemake rules to run these GWASs using plink2.</p>
<pre class="python"><code># Snakemake rule  
rule gwas_no_correction:
  input:
      genos=&quot;output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-gwas_common.psam&quot;,
      pheno=&quot;output/Simulate_Phenotypes/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.phenos.txt&quot;
  output:
      &quot;output/Run_GWAS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.pheno_strat.glm.linear&quot;
  shell:
      &quot;&quot;&quot;
      ~/plink2 \
      --pfile output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/{wildcards.config}/genos-gwas_common \
      --glm allow-no-covars\
      --pheno {input.pheno} \
      --pheno-name pheno_strat \
      --out output/Run_GWAS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common
      &quot;&quot;&quot;

rule gwas_Tm:
    input:
      genos=&quot;output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-gwas_common.psam&quot;,
      pheno=&quot;output/Simulate_Phenotypes/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.phenos.txt&quot;,
      Tm=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tm-ID_covars.txt&quot;
    output:
      &quot;output/Run_GWAS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-Tm.pheno_strat.glm.linear&quot;
    shell:
      &quot;&quot;&quot;
      plink2 \
      --pfile output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/{wildcards.config}/genos-gwas_common \
      --glm hide-covar \
      --covar {input.Tm} \
      --covar-col-nums 3 \
      --pheno {input.pheno} \
      --pheno-name pheno_strat \
      --out output/Run_GWAS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common-Tm
      &quot;&quot;&quot;

rule gwas_PopID:
    input:
      genos=&quot;output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-gwas_common.psam&quot;,
      pheno=&quot;output/Simulate_Phenotypes/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.phenos.txt&quot;,
      Tm=&quot;output/Calculate_Tm/4PopSplit/{rep}/{config}/Tm-ID_covars.txt&quot;
    output:
      &quot;output/Run_GWAS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-ID.pheno_strat.glm.linear&quot;
    shell:
      &quot;&quot;&quot;
      plink2 \
      --pfile output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/{wildcards.config}/genos-gwas_common \
      --glm hide-covar \
      --covar {input.Tm} \
      --covar-col-nums 4 \
      --pheno {input.pheno} \
      --pheno-name pheno_strat \
      --out output/Run_GWAS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common-ID
      &quot;&quot;&quot;</code></pre>
</div>
<div id="ascertain-snps" class="section level3">
<h3>Ascertain SNPs</h3>
<p>Next we want to ascertain SNPs to include in the polygenic scores we compute in the selection test. We ascertain 3 sets of SNPs that can be used to build the PGSs. Recall that we have simulated 200 separate chromosomes and will always choose one SNP per chromosome for a maxi um of 200 SNPs. In practice we set the p-value threshold to 1 so we are always ascertaining 200 variants to include.</p>
<ol style="list-style-type: decimal">
<li>Causal sites (note if the trait is not heritable these are randomly chosen sites)<br />
</li>
<li>Causal sites whose p-value is bellow a given threshold<br />
</li>
<li>Non-causal sites that are chosen as the minimum p-value (that passes a given threshold) per chromosome.</li>
</ol>
<p>First we select the causal variants</p>
<pre class="r"><code># Read in causal effects
causal &lt;- fread(&quot;../output/Simulate_Phenotypes/4PopSplit/B1/C1/h2-0.3/p-0.50/env_0.0/genos-gwas_common.effects.txt&quot;)
colnames(causal)=c(&quot;rsid&quot;,&quot;allele&quot;,&quot;esize&quot;)
causal=causal%&gt;%
  tidyr::separate(rsid,into=c(&quot;CHROM&quot;,&quot;POS&quot;,&quot;ref&quot;,&quot;alt&quot;),sep=&quot;_&quot;,remove=F)
causal$POS=as.numeric(causal$POS)
causal$CHROM=as.numeric(causal$CHROM)

# Read in the GWAS estimates 
gwas &lt;- fread(&quot;../output/Run_GWAS/4PopSplit/B1/C1/h2-0.3/p-0.50/env_0.0/genos-gwas_common.pheno_strat.glm.linear&quot;)
colnames(gwas)[1]=&quot;CHROM&quot;
gwas$Index &lt;- seq(1, nrow(gwas))

# Flip effects sizes to get effect size of the ALT allele 
flip_effect = function(gwas_df,beta_colname){
  gwas_df = gwas_df[A1==&quot;A&quot;, beta_colname := -BETA]
  gwas_df = gwas_df[A1==&quot;T&quot;, beta_colname := BETA]
  gwas_df$A1=&quot;T&quot;
  gwas_df = gwas_df[,.(CHROM,POS,ID,A1,beta_colname,P, Index)]
  colnames(gwas_df)[5] = beta_colname
  return(gwas_df)
}
gwas = flip_effect(gwas,beta_colname = &quot;BETA1&quot;)

# Select effect sizes for causal variants
gwas.causal= gwas[ID%in%causal$rsid]
gwas.causal=gwas.causal[,c(&quot;ID&quot;,&quot;A1&quot;,&quot;BETA1&quot;, &quot;P&quot;, &quot;Index&quot;)]
colnames(gwas.causal) &lt;- c(&quot;ID&quot;, &quot;A1&quot;, &quot;BETA_Strat&quot;, &quot;P&quot;, &quot;Index&quot;)

# Plot GWAS results, highlighting ascertained SNPS 
plot(gwas$Index, -log10(gwas$P), xlab = &quot;Index&quot;, ylab = &quot;-log10(P)&quot;)
points(gwas.causal$Index, -log10(gwas.causal$P), col = &quot;red&quot;, pch = 18)</code></pre>
<p><img src="figure/Run_GWAS.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We can also compare estimated effect sizes to causal effect sizes</p>
<pre class="r"><code>plot(causal$esize,gwas.causal$BETA_Strat, xlab = &quot;Causal Effect Size&quot;, ylab = &quot;Estimated Effect Size&quot;)
abline(0, 1, col = &quot;red&quot;)</code></pre>
<p><img src="figure/Run_GWAS.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Next, we can filter these causal sites to only ascertain those under a given p-value (ex. 1e-3).</p>
<pre class="r"><code># Function to select variants under a threshold
fcausal_p = function(df,pvalue=pval_threshold){
  df=df%&gt;%
    filter(P &lt; pvalue)
  return(df)
}
gwas.thresh = fcausal_p(gwas.causal,1e-3)
gwas.thresh=gwas.thresh[,c(&quot;ID&quot;,&quot;A1&quot;,&quot;BETA_Strat&quot;, &quot;P&quot;, &quot;Index&quot;)]

# Plot GWAS results, highlighting ascertained SNPS 
plot(gwas$Index, -log10(gwas$P), xlab = &quot;Index&quot;, ylab = &quot;-log10(P)&quot;)
points(gwas.thresh$Index, -log10(gwas.thresh$P), col = &quot;green&quot;, pch = 18)</code></pre>
<p><img src="figure/Run_GWAS.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Finally, we will ascertain a third set of SNPs where we select the SNP with the lowest p-value, below a threshold, per chromosome.</p>
<pre class="r"><code># Function to select lowest p-value per chromosome
fclump &lt;- function(df, pt, CHR) {

  # Select only SNPS under threshold
  df &lt;-  gwas1 %&gt;% filter(P &lt; pt) %&gt;% filter(CHROM == CHR)

  # Select lowest p-value
  min_p &lt;- df %&gt;% slice_min(P, with_ties = F)

  return(min_p)
}

# Extract chromosome info
tmp = gwas %&gt;% mutate(ID2 = ID) %&gt;% separate(ID2, c(&quot;chr&quot;, &quot;pos&quot;, &quot;a1&quot;, &quot;a2&quot;), &quot;_&quot;)
gwas1= tmp %&gt;% mutate(CHROM = chr) %&gt;% select(&quot;CHROM&quot;, &quot;POS&quot;, &quot;ID&quot;, &quot;A1&quot;, &quot;BETA1&quot;, &quot;P&quot;, &quot;Index&quot;)

# Set pvalue threshold
pval_threshold = 1

# Pick the minimum p-value SNP on chrom 1 
nchrms = unique(gwas1$CHROM)
gwas.red = fclump(gwas1, pval_threshold, 1)

# Repeat selection for each chromosome
for (i in 2:length(nchrms)) {
  new = fclump(gwas1, pval_threshold, nchrms[i])
  new = new[order(new), ]
  gwas.red = rbind(gwas.red, new)
}

# Plot GWAS results, highlighting ascertained SNPS 
plot(gwas$Index, -log10(gwas$P), xlab = &quot;Index&quot;, ylab = &quot;-log10(P)&quot;)
points(gwas.red$Index, -log10(gwas.red$P), col = &quot;purple&quot;, pch = 18)</code></pre>
<p><img src="figure/Run_GWAS.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Here are the snakemake rules to run all three ascertainment steps.</p>
<pre class="python"><code>rule pick_SNPS:
    input:
      causal_effect=&quot;output/Simulate_Phenotypes/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.effects.txt&quot;,
      gwas_strat=&quot;output/Run_GWAS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.pheno_strat.glm.linear&quot;
    output:
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.c.betas&quot;,
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.c.p.betas&quot;,
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.nc.betas&quot;
    params:
      pt = PVALUE_THRESHOLD
    shell:
      &quot;&quot;&quot;
      Rscript code/PRS/clump.R {input.causal_effect} output/Run_GWAS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common {params.pt} output/PRS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common
      &quot;&quot;&quot;

rule pick_SNPS_Tm:
    input:
      causal_effect=&quot;output/Simulate_Phenotypes/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.effects.txt&quot;,
      gwas_strat=&quot;output/Run_GWAS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-Tm.pheno_strat.glm.linear&quot;
    output:
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-Tm.c.betas&quot;,
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-Tm.c.p.betas&quot;,
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-Tm.nc.betas&quot;
    params:
      pt = PVALUE_THRESHOLD
    shell:
      &quot;&quot;&quot;
      Rscript code/PRS/clump.R {input.causal_effect} output/Run_GWAS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common-Tm {params.pt} output/PRS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common-Tm
      &quot;&quot;&quot;

rule pick_SNPS_ID:
    input:
      causal_effect=&quot;output/Simulate_Phenotypes/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common.effects.txt&quot;,
      gwas_strat=&quot;output/Run_GWAS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-ID.pheno_strat.glm.linear&quot;
    output:
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-ID.c.betas&quot;,
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-ID.c.p.betas&quot;,
      &quot;output/PRS/4PopSplit/{rep}/{config}/{h2}/{ts}/{env}/genos-gwas_common-ID.nc.betas&quot;
    params:
      pt = PVALUE_THRESHOLD
    shell:
      &quot;&quot;&quot;
      Rscript code/PRS/clump.R {input.causal_effect} output/Run_GWAS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common-ID {params.pt} output/PRS/4PopSplit/{wildcards.rep}/{wildcards.config}/{wildcards.h2}/{wildcards.ts}/{wildcards.env}/genos-gwas_common-ID
      &quot;&quot;&quot;</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] tidyr_1.1.3       pgenlibr_0.3.1    data.table_1.14.2 dplyr_1.0.7      
[5] knitr_1.37       

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       highr_0.9        pillar_1.6.4     compiler_3.6.2  
 [5] bslib_0.3.0      later_1.3.0      jquerylib_0.1.4  git2r_0.28.0    
 [9] workflowr_1.6.2  tools_3.6.2      digest_0.6.29    lattice_0.20-44 
[13] jsonlite_1.7.2   evaluate_0.14    lifecycle_1.0.1  tibble_3.1.6    
[17] png_0.1-7        pkgconfig_2.0.3  rlang_0.4.12     Matrix_1.3-4    
[21] DBI_1.1.1        yaml_2.2.1       xfun_0.29        fastmap_1.1.0   
[25] stringr_1.4.0    generics_0.1.0   fs_1.5.2         vctrs_0.3.8     
[29] sass_0.4.0       grid_3.6.2       rprojroot_2.0.2  tidyselect_1.1.1
[33] reticulate_1.20  glue_1.6.0       R6_2.5.1         fansi_0.5.0     
[37] rmarkdown_2.10   purrr_0.3.4      magrittr_2.0.1   promises_1.2.0.1
[41] ellipsis_0.3.2   htmltools_0.5.2  assertthat_0.2.1 httpuv_1.6.2    
[45] utf8_1.2.2       stringi_1.7.6    crayon_1.4.2    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
