<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jennifer Blanc" />


<title>Simulate Genotypes</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Simulate Genotypes</h1>
<h4 class="author">Jennifer Blanc</h4>
<h4 class="date">9/28/2021</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-10-05
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201015code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201015)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201015code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201015)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtreecba0a794d92676b31dc4ac30147503c02840d20dtargetblankcba0a79a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/cba0a794d92676b31dc4ac30147503c02840d20d" target="_blank">cba0a79</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtreecba0a794d92676b31dc4ac30147503c02840d20dtargetblankcba0a79a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/cba0a794d92676b31dc4ac30147503c02840d20d" target="_blank">cba0a79</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/
    Ignored:    code/.DS_Store
    Ignored:    code/Calculate_Tm/.DS_Store
    Ignored:    code/Debug/.DS_Store
    Ignored:    code/PGA_test/.DS_Store
    Ignored:    code/Simulate_Phenotypes/.DS_Store
    Ignored:    code/Snakefiles/.DS_Store
    Ignored:    code/Snakefiles/SimpleGrid/.DS_Store
    Ignored:    data/.DS_Store
    Ignored:    data/projection_example/
    Ignored:    output/Calculate_Tm/4PopSplit/E2/
    Ignored:    output/Calculate_Tm/4PopSplit/E20/
    Ignored:    output/Calculate_Tm/4PopSplit/E3/
    Ignored:    output/Calculate_Tm/4PopSplit/E4/
    Ignored:    output/Calculate_Tm/4PopSplit/E5/
    Ignored:    output/Calculate_Tm/4PopSplit/E6/
    Ignored:    output/Calculate_Tm/4PopSplit/E7/
    Ignored:    output/Calculate_Tm/4PopSplit/E8/
    Ignored:    output/Calculate_Tm/4PopSplit/E9/
    Ignored:    output/Calculate_Tm/SimpleGrid/E1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E2/
    Ignored:    output/Calculate_Tm/SimpleGrid/E3/
    Ignored:    output/Calculate_Tm/SimpleGrid/E4/
    Ignored:    output/Calculate_Tm/SimpleGrid/E5/
    Ignored:    output/PGA_test/4PopSplit/E2/
    Ignored:    output/PGA_test/4PopSplit/E20/
    Ignored:    output/PGA_test/4PopSplit/E3/
    Ignored:    output/PGA_test/4PopSplit/E4/
    Ignored:    output/PGA_test/4PopSplit/E5/
    Ignored:    output/PGA_test/4PopSplit/E6/
    Ignored:    output/PGA_test/4PopSplit/E7/
    Ignored:    output/PGA_test/4PopSplit/E8/
    Ignored:    output/PGA_test/4PopSplit/E9/
    Ignored:    output/PGA_test/SimpleGrid/
    Ignored:    output/PRS/4PopSplit/E2/
    Ignored:    output/PRS/4PopSplit/E20/
    Ignored:    output/PRS/4PopSplit/E3/
    Ignored:    output/PRS/4PopSplit/E4/
    Ignored:    output/PRS/4PopSplit/E5/
    Ignored:    output/PRS/4PopSplit/E6/
    Ignored:    output/PRS/4PopSplit/E7/
    Ignored:    output/PRS/4PopSplit/E8/
    Ignored:    output/PRS/4PopSplit/E9/
    Ignored:    output/PRS/SimpleGrid/
    Ignored:    output/Run_GWAS/
    Ignored:    output/Simulate_Genotypes/4PopSplit/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E2/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E3/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E4/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E5/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E6/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E7/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E8/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E2/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E20/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E3/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E4/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E5/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E6/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E7/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E8/
    Ignored:    output/Simulate_Phenotypes/4PopSplit/E9/
    Ignored:    output/Simulate_Phenotypes/SimpleGrid/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Simulate_Genotypes.Rmd</code>) and HTML (<code>docs/Simulate_Genotypes.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/4a478051738fc7e6fb2d027da74d1a0b02c9722c/analysis/Simulate_Genotypes.Rmd" target="_blank">4a47805</a>
</td>
<td>
jgblanc
</td>
<td>
2021-09-29
</td>
<td>
organizing website
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/4a478051738fc7e6fb2d027da74d1a0b02c9722c/docs/Simulate_Genotypes.html" target="_blank">4a47805</a>
</td>
<td>
jgblanc
</td>
<td>
2021-09-29
</td>
<td>
organizing website
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="simulate-genotypes" class="section level2">
<h2>Simulate Genotypes</h2>
<p>Here we outline the steps of the snakemake pipeline that generate the genotypes used for stratification simulations (Fig 1 and 2).</p>
<div id="msprime-simulations" class="section level3">
<h3>Msprime simulations</h3>
<p>The example below is for the 4 population split toy model but the snakefiles for the stepping-stone grid model use the same code but swap <code>code/Simulate_Genotypes/generate_genotypes_4PopSplit.py</code> for <code>code/Simulate_Genotypes/generate_genotypes_SimpleGrid.py</code>.</p>
<p>Here we are calling a python script that runs a separate msprime demographic simulation for <code>chr_num</code> chromosomes. <code>-NA</code>, <code>NB</code>, <code>NC</code>, and <code>ND</code> are the population sizes for each of the 4 populations while <code>-a</code>, <code>-b</code>, <code>-c</code>, and <code>d</code> are the sample sizes. <code>-Nac</code> is the ancestral popultaion size while <code>-L</code> is the length of the chromosom. The rest of the parameters are set to their default values as defined in the script itself. The output of this rule is <code>chr_num</code> separate vcf files. We also output a text file that indicates which population (or deme) each individuals belongs to.</p>
<pre class="python"><code>rule simulate_genotypes_4popsplit:
  output:
    expand(&quot;output/Simulate_Genotypes/4PopSplit/{{rep}}/genos_{chr}.vcf&quot;, chr=CHR),
&quot;output/Simulate_Genotypes/4PopSplit/{rep}/genos.pop&quot;
  params:
    chr_num = len(CHR)
  shell:
    &quot;python code/Simulate_Genotypes/generate_genotypes_4PopSplit.py \
    --outpre output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/genos \
    --chr {params.chr_num} \
    --Nanc 10000 \
    --NA 10000 \
    --NB 10000 \
    --NC 10000 \
    --ND 10000 \
    -a 2000 \   
    -b 2000 \
    -c 2000 \
    -d 2000 \
    -s1 22000 \
    -s2 11000 \
    -L 100000&quot;</code></pre>
</div>
<div id="formating-vcfs" class="section level3">
<h3>Formating vcfs</h3>
<p>The fist rule below takes each individual .vcf file and reformats it so the SNP ID is “chr_genomic-location_A_T” and the reference allele is always A and the alternate allele is T. The second rule uses bcftools to first convert all the CHR in the .vcfs to say “1”. We are simulating 200 separate chromosome but plink does not allow for that many chromosomes so we relabel all of them “1” and retain the true chromosome info in the SNP ID. We then concatenate all chromosome .vcf files into a single .vcf.</p>
<pre class="python"><code>rule format_VCF:
  input:
    &quot;output/Simulate_Genotypes/{model}/{rep}/genos_{chr}.vcf&quot;
  output:
    gz=&quot;output/Simulate_Genotypes/{model}/{rep}/genos_{chr}.ids.vcf.gz&quot;
    #csi=&quot;output/Simulate_Genotypes/{model}/{rep}/genos_{chr}.ids.vcf.gz.csi&quot;
  shell:
    &quot;&quot;&quot;
    head -n6 {input} &gt; output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/header_{wildcards.chr}.txt
    cat output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/header_{wildcards.chr}.txt &lt;(cat output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/genos_{wildcards.chr}.vcf | awk -v OFS=&quot;\t&quot; &#39;NR&gt;6 {{$3=$1&quot;_&quot;$2&quot;_A_T&quot;;$4=&quot;A&quot;; $5=&quot;T&quot;; print ;}}&#39;) | bgzip &gt; {output.gz}
    rm output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/header_{wildcards.chr}.txt
    &quot;&quot;&quot;

rule concat_vcfs:
  input:
    expand(&quot;output/Simulate_Genotypes/{{model}}/{{rep}}/genos_{chr}.ids.vcf.gz&quot;, chr=CHR)
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/genos.ids.vcf.gz&quot;
  shell:
    &quot;&quot;&quot;
    bcftools concat {input} -o output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/temp.vcf.gz -O z
    bcftools annotate --rename-chrs code/Simulate_Genotypes/convert_chr.txt
    output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/temp.vcf.gz -o {output} -O z
    rm output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/temp.vcf.gz
    &quot;&quot;&quot;</code></pre>
</div>
<div id="convert-to-plink2-format" class="section level3">
<h3>Convert to plink2 format</h3>
<p>The code below converts the .vcf file into the plink2 file format (.psam, .pgen, .pgen).</p>
<pre class="python"><code>rule convert_vcf_to_plink:
  input:
    &quot;output/Simulate_Genotypes/{model}/{rep}/genos.ids.vcf.gz&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/genos.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/genos.pgen&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/genos.pvar&quot;
  shell:
    &quot;plink2 \
    --double-id \
    --make-pgen \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/genos \
    --vcf {input}&quot;</code></pre>
</div>
<div id="create-testgwas-panels" class="section level3">
<h3>Create test/GWAS panels</h3>
<p>We first use a custom R script to split to individuals into GWAS and test panels. In one configuration “C1” inviduals from A and C are the GWAS panel and B and D are the test. In “C2” A and B are the GWAS panel and C and D are the test. <code>config</code> becomes a wildcard and all subsequent code is run on both configurations. In the next we rule we use plink to create two sets of plink files for each panel. We then have an optional rule that downsamples the number of individuals in the test panel - we did not use this rule in our analysis.</p>
<pre class="python"><code>rule create_panels_4PopSplit:
  input:
    &quot;output/Simulate_Genotypes/{model}/{rep}/genos.pop&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/C1/ids.gwas&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/C1/ids.test&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/C2/ids.gwas&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/C2/ids.test&quot;
  script:
    &quot;code/Simulate_Genotypes/split_gwas-test_4PopSplit.R&quot;

rule split_into_test_gwas:
  input:
    gwas=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/ids.gwas&quot;,
    test=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/ids.test&quot;,
    psam=&quot;output/Simulate_Genotypes/{model}/{rep}/genos.psam&quot;,
    pvar=&quot;output/Simulate_Genotypes/{model}/{rep}/genos.pvar&quot;,
    pgen=&quot;output/Simulate_Genotypes/{model}/{rep}/genos.pgen&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test-big.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test-big.pgen&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test-big.pvar&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.pgen&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.pvar&quot;
  shell:
    &quot;&quot;&quot;
    plink2 \
    --keep {input.gwas} \
    --make-pgen \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-gwas \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/genos \
    --rm-dup exclude-all

    plink2 \
    --keep {input.test} \
    --make-pgen \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-test-big \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/genos \
    --rm-dup exclude-all
    &quot;&quot;&quot;

rule downsample_test:
  input:
    psam=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test-big.psam&quot;,
    pgen=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test-big.pgen&quot;,
    pvar=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test-big.pvar&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.pvar&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.pgen&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/downsample.id&quot;
  params:
    size = SIZE
  shell:
    &quot;&quot;&quot;
    set +o pipefail;
    awk &#39;NR &gt; 1&#39; {input.psam} | cut -f 1,2 | sort -R | head -n {params.size} &gt; output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/downsample.id

    plink2 \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-test-big \
    --keep output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/downsample.id \
    --make-pgen \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-test
    &quot;&quot;&quot;</code></pre>
</div>
<div id="filter-on-variant-frequency" class="section level3">
<h3>Filter on variant frequency</h3>
<p>Finally we need to filter the variants included in our downstream analysis. First we use plink to get the frequency of every SNP in both panels. In the second rule, we use a custom R script to get a list of SNPs that are at least 5% frequency in both panels. Then we use this list to remake the plink files for each panel only including these SNPs. Finally, we use plink to get the frequency of these SNPs in each panel which we will use in subsequent steps.</p>
<pre class="python"><code>rule get_variant_freq:
  input:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.pvar&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.pgen&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.afreq&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.afreq&quot;
  shell:
    &quot;&quot;&quot;
    plink2 \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-test \
    --freq \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-test

    plink2 \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-gwas \
    --freq \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-gwas
    &quot;&quot;&quot;

rule get_common_snp_list:
  input:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.afreq&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.afreq&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/common_snp_ids.txt&quot;
  script:
    &quot;code/Simulate_Genotypes/get_common_snp_list.R&quot;

rule remake_panels_with_common_snps:
  input:
    common_id=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/common_snp_ids.txt&quot;,
    test_psam=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.psam&quot;,
    test_pvar=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.pvar&quot;,
    test_pgen=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test.pgen&quot;,
    gwas_psam=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.psam&quot;,
    gwas_pvar=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.pvar&quot;,
    gwas_pgen=&quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas.pgen&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test_common.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test_common.pvar&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-test_common.pgen&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas_common.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas_common.pvar&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas_common.pgen&quot;
  shell:
    &quot;&quot;&quot;
    plink2 \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-test \
    --extract {input.common_id} \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-test_common \
    --make-pgen

    plink2 \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-gwas \
    --extract {input.common_id} \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-gwas_common \
    --make-pgen
    &quot;&quot;&quot;

rule common_snp_freq:
  input:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas_common.psam&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas_common.pvar&quot;,
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas_common.pgen&quot;
  output:
    &quot;output/Simulate_Genotypes/{model}/{rep}/{config}/genos-gwas_common.afreq&quot;
  shell:
    &quot;plink2 \
    --pfile output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-gwas_common \
    --out output/Simulate_Genotypes/{wildcards.model}/{wildcards.rep}/{wildcards.config}/genos-gwas_common \
    --freq&quot;</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
