<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jennifer Blanc" />


<title>Simulation_Methods_Reproducible</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Simulation_Methods_Reproducible</h1>
<h4 class="author">Jennifer Blanc</h4>
<h4 class="date">1/3/2022</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-05-02
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of the R Markdown file created these results, you’ll want to first commit it to the Git repo. If you’re still working on the analysis, you can ignore this warning. When you’re finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201015code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201015)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201015code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201015)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree155709963e77c0feb9471d1f9512c9ab2809e5cctargetblank1557099a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/155709963e77c0feb9471d1f9512c9ab2809e5cc" target="_blank">1557099</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree155709963e77c0feb9471d1f9512c9ab2809e5cctargetblank1557099a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/155709963e77c0feb9471d1f9512c9ab2809e5cc" target="_blank">1557099</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/
    Ignored:    analysis/Calculate_TGWAS_cache/
    Ignored:    analysis/Polygenic_Selection_Test_cache/
    Ignored:    code/.DS_Store
    Ignored:    code/1kg/.DS_Store
    Ignored:    code/Calculate_Tm/.DS_Store
    Ignored:    code/Debug/.DS_Store
    Ignored:    code/PGA_test/.DS_Store
    Ignored:    code/PRS/.DS_Store
    Ignored:    code/Simulate_Genotypes/.DS_Store
    Ignored:    code/Simulate_Phenotypes/.DS_Store
    Ignored:    code/Snakefiles/.DS_Store
    Ignored:    code/Snakefiles/4PopSplit/.DS_Store
    Ignored:    code/Snakefiles/4PopSplit/Signal/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/conda-archive/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/conda/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/locks/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/metadata/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/shadow/
    Ignored:    code/Snakefiles/SimpleGrid/.snakemake/singularity/
    Ignored:    data/.DS_Store
    Ignored:    data/1kg/
    Ignored:    data/cgd_example/
    Ignored:    data/projection_example/
    Ignored:    data/proof/
    Ignored:    output/.DS_Store
    Ignored:    output/Calculate_TGWAS/
    Ignored:    output/Calculate_Tm/4PopSplit/
    Ignored:    output/Calculate_Tm/SimpleGrid/A1/
    Ignored:    output/Calculate_Tm/SimpleGrid/B1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E1/
    Ignored:    output/Calculate_Tm/SimpleGrid/E2/
    Ignored:    output/Calculate_Tm/SimpleGrid/E3/
    Ignored:    output/Calculate_Tm/SimpleGrid/E4/
    Ignored:    output/Calculate_Tm/SimpleGrid/E5/
    Ignored:    output/PGA_test/
    Ignored:    output/PRS/
    Ignored:    output/Run_GWAS/
    Ignored:    output/Simulate_Genotypes/.DS_Store
    Ignored:    output/Simulate_Genotypes/4PopSplit/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/A1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E1/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E2/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E3/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E4/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E5/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E6/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E7/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/E8/
    Ignored:    output/Simulate_Genotypes/SimpleGrid/S1/
    Ignored:    output/Simulate_Phenotypes/

Unstaged changes:
    Modified:   analysis/Simulation_Methods_Reproducible.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Simulation_Methods_Reproducible.Rmd</code>) and HTML (<code>docs/Simulation_Methods_Reproducible.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/51065b39b2903aa1b80e64ca9c56a19bb0356ea5/analysis/Simulation_Methods_Reproducible.Rmd" target="_blank">51065b3</a>
</td>
<td>
jgblanc
</td>
<td>
2022-04-29
</td>
<td>
ready for final 4 pop run
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/931f81811a70a5965859780201a24c6e17e2fab8/analysis/Simulation_Methods_Reproducible.Rmd" target="_blank">931f818</a>
</td>
<td>
jgblanc
</td>
<td>
2022-04-10
</td>
<td>
ready to run meta
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/b9219567664670675ba0fdadc737553b46a013dc/analysis/Simulation_Methods_Reproducible.Rmd" target="_blank">b921956</a>
</td>
<td>
jgblanc
</td>
<td>
2022-03-31
</td>
<td>
ready for full run
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/587c88ec6456a223d81d1736be154d64444d254c/analysis/Simulation_Methods_Reproducible.Rmd" target="_blank">587c88e</a>
</td>
<td>
jgblanc
</td>
<td>
2022-03-29
</td>
<td>
set up simple run
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/73fd8f8aad934f339fe4093d54fa6489071444e8/analysis/Simulation_Methods_Reproducible.Rmd" target="_blank">73fd8f8</a>
</td>
<td>
jgblanc
</td>
<td>
2022-03-24
</td>
<td>
ready to run 2 pops
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/1cb285705e3f702ae4657cf42de80418bf96a72f/analysis/Simulation_Methods_Reproducible.Rmd" target="_blank">1cb2857</a>
</td>
<td>
jgblanc
</td>
<td>
2022-03-03
</td>
<td>
added conjugate gradient descent - to test tomorrow
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/db79e2ed4f64a80e540c758f89ecac59b34e9fd0/docs/Simulation_Methods_Reproducible.html" target="_blank">db79e2e</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/1e10dec627b5ea23e6a139befadaf18de7b4dfbc/analysis/Simulation_Methods_Reproducible.Rmd" target="_blank">1e10dec</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
<td>
wflow_publish(c(“analysis/Simulation_Methods_Reproducible.Rmd”,
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Here I will reccord the details for the simulations presented in the main text and the supplement. I will also include the code to make the the figures.</p>
</div>
<div id="popsplit" class="section level2">
<h2>4PopSplit</h2>
<div id="msprime-simulation" class="section level3">
<h3>Msprime simulation</h3>
<p>Generate diplois individuals according to the demographic model below</p>
<ul>
<li>T2 = 2200<br />
</li>
<li>T1 (deeper split) = 4400</li>
<li>Ne (ancestral = A = B = C = D) = 10000</li>
<li>sample size = 10000 diploids per population<br />
</li>
<li>chromosome length = 100000<br />
</li>
<li>200 chromosomes<br />
</li>
<li><span class="math inline">\(\mu\)</span> = 1e-08<br />
</li>
<li><span class="math inline">\(\rho\)</span> = 1e-08</li>
</ul>
</div>
<div id="create-genotypes-panels-and-filter-snps" class="section level3">
<h3>Create genotypes panels and filter SNPs</h3>
<ol style="list-style-type: decimal">
<li>Split individuals into 2 configurations<br />
</li>
</ol>
<ol style="list-style-type: lower-alpha">
<li>C1: GWAS Panel = A and C, Test Panel = B and D (overlapping structure)<br />
</li>
<li>C2: GWAS Panel = A and B, Test Panel = C and D (no overlapping structure)<br />
</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>Randomly downsample 1,000 individuals from each test panel population. Finale sample sizes: GWAS panel n=20,000 and Test panel n=2,000.<br />
</li>
<li>Restrict snps to sites segregating at greater than 5% frequency within each panel.</li>
</ol>
<p>See <a href="Simulate_Genotypes.html">link</a> for more simluations details.</p>
</div>
<div id="draw-phenotypes" class="section level3">
<h3>Draw phenotypes</h3>
<p>We simulated both heritable <span class="math inline">\(h^2 = 0.3\)</span> and non-heritable <span class="math inline">\(h^2=0\)</span> phenotypes with a ranger of environmental confounders <span class="math inline">\(\Delta_{AB}\)</span>. Additionally, for the heritable simulations, we also conducted simulations where there was a true signal of polygenic selection by flipping the signs of causal effect sizes depending on allele frequency difference in the test panel.</p>
<p>See <a href="Simulate_Phenotypes.html">link</a> for more simluations details.</p>
</div>
<div id="project-test-vector" class="section level3">
<h3>Project Test Vector</h3>
<ol style="list-style-type: decimal">
<li>Set test vector as a mean centered <span class="math inline">\(n \times 1\)</span> vector with each entry as an indicator of the test panel individual’s popultion ID.<br />
</li>
<li>Project test vector in GWAS panel, resulting in an <span class="math inline">\(m \times 1\)</span> (see <a href="Calculate_TGWAS.html">link</a> for more detail)</li>
</ol>
</div>
<div id="run-gwass" class="section level3">
<h3>Run GWASs</h3>
<ol style="list-style-type: decimal">
<li>Use plink to run GWAS with and without including projected test vector (see <a href="Run_GWAS.html">link</a> for more detail)</li>
</ol>
</div>
<div id="calculate-pgs-and-qx" class="section level3">
<h3>Calculate PGS and Qx</h3>
<p><strong>Ascertained Sites</strong></p>
<ol style="list-style-type: decimal">
<li>Pick the SNP with lowest p-value per chromosome to include in PGS (200 SNPs total)<br />
</li>
<li>Compute PGS in test panel individuals</li>
<li>Calculate <span class="math inline">\(V_a\)</span></li>
<li>Calculate the amount of variance explained by the test vector (<span class="math inline">\(\lambda_T\)</span>)<br />
</li>
<li>Calculate chi-squared p-vale</li>
<li>Calculate empirical null p-value by permuting the signs of the SNP effect sizes 1,000 times and computing the proportion of permutations greater than the true <span class="math inline">\(Q_x\)</span> value.</li>
</ol>
<p><strong>Causal sites</strong></p>
<p>For heritable traits we can also repeat the above steps with the acual causal sites as opposed to choosing the site with the lowest p-value.</p>
<p>See <a href="Polygenic_Selection_Test.html">link</a> for more detail.</p>
</div>
<div id="figures" class="section level3">
<h3>Figures</h3>
<div id="figure-1" class="section level4">
<h4>Figure 1</h4>
<ul>
<li><span class="math inline">\(h^2 = 0\)</span><br />
</li>
<li>Ascertained SNPs (minimum p-value per chromosome)</li>
</ul>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-2-1">
Past versions of unnamed-chunk-2-1.png
</button>
</p>
<div id="fig-unnamed-chunk-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/db79e2ed4f64a80e540c758f89ecac59b34e9fd0/docs/figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-2-1.png" target="_blank">db79e2e</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can also compare the above results using <span class="math inline">\(T^{GWAS}\)</span> to a GWAS done using population ID as covariate.</p>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-3-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/db79e2ed4f64a80e540c758f89ecac59b34e9fd0/docs/figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-3-1.png" target="_blank">db79e2e</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="heritable-simulations" class="section level4">
<h4>Heritable Simulations</h4>
<ul>
<li><span class="math inline">\(h^2 = 0.3\)</span></li>
<li>Asceratined SNPs</li>
</ul>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/db79e2ed4f64a80e540c758f89ecac59b34e9fd0/docs/figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-4-1.png" target="_blank">db79e2e</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/db79e2ed4f64a80e540c758f89ecac59b34e9fd0/docs/figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-5-1.png" target="_blank">db79e2e</a>
</td>
<td>
jgblanc
</td>
<td>
2022-01-03
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
</div>
<div id="simple-grid" class="section level2">
<h2>Simple Grid</h2>
<div id="msprime-simulation-1" class="section level3">
<h3>Msprime simulation</h3>
<p>Generate diploid individuals using a stepping stone model on a 6x6 grid with the parameters below:</p>
<p>Parameters:</p>
<ul>
<li>200 chromosomes<br />
</li>
<li>sample size = 80 diploids<br />
</li>
<li>Ne = 1000</li>
<li><span class="math inline">\(\mu\)</span> = 1e-07<br />
</li>
<li><span class="math inline">\(\rho\)</span> = 1e-07</li>
<li>m = 0.01<br />
</li>
<li>chromosome length = 10000</li>
<li>tmove = -9 (perpetual structure)</li>
</ul>
<p>The simulation was run 100 times total.</p>
</div>
<div id="create-genotypes-panels-and-filter-snps-1" class="section level3">
<h3>Create genotypes panels and filter SNPs</h3>
<ol style="list-style-type: decimal">
<li>Sample 20 individuals per deme for test panels (36 * 20 individuals total)</li>
<li>Remaining individuals (60 per deme) become the GWAS panel<br />
</li>
<li>Restrict snps to sites segregating at greater than 5% frequency within each panel</li>
</ol>
</div>
<div id="draw-phenotype" class="section level3">
<h3>Draw phenotype</h3>
<p>Notes:</p>
<ul>
<li>All simulations have <span class="math inline">\(h^2 = 0\)</span><br />
</li>
<li>ENV is the magnitude of the phenotype shift</li>
</ul>
<ol style="list-style-type: decimal">
<li>Sample random phenotype <span class="math inline">\(N(0,1)\)</span> for all GWAS indviduals</li>
<li>Scale variance within deme to match full size GWAS variance (see __ for more details)</li>
<li>Simulate stratification:<br />
</li>
<li>Lat/Lat: Shift average phenotype per Latitude (if env is the total shift, shift 1 lat by env/5, 2 lat by 2*(env/5), etc.) and use Latitude as the test vector</li>
<li>stratD/Lat: Shift the average phenotype on the diagonals of the grid going lower left to upper right and use latitude as the test vector</li>
</ol>
</div>
<div id="project-test-vector-1" class="section level3">
<h3>Project Test Vector</h3>
<ol style="list-style-type: decimal">
<li>Set test vector as a mean centered <span class="math inline">\(n \times 1\)</span> vector with each entry as the test panel individuals latitude/longitude/popID<br />
</li>
<li>Project test vector in GWAS panel, resulting in an <span class="math inline">\(m \times 1\)</span> (see ___ for more detail)</li>
</ol>
</div>
<div id="run-gwass-1" class="section level3">
<h3>Run GWASs</h3>
<ol style="list-style-type: decimal">
<li>Use plink to run GWAS with and without including projected test vector</li>
</ol>
</div>
<div id="calculate-pgs-and-qx-1" class="section level3">
<h3>Calculate PGS and Qx</h3>
<ol style="list-style-type: decimal">
<li>Pick the SNP with lowest p-value per chromosome to include in PGS (200 SNPs total)<br />
</li>
<li>Compute PGS in test panel individuals</li>
<li>Calculate <span class="math inline">\(V_a\)</span></li>
<li>Calculate the amount of variance explained by the test vector (<span class="math inline">\(\lambda_T\)</span>; note that we used the mean-centered test vector to compute this)<br />
</li>
<li>Calculate empirical null p-value by permuting the signs of the SNP effect sizes 1,000 times and computing the proportion of permutations greater than the true <span class="math inline">\(Q_x\)</span> value.</li>
</ol>
</div>
<div id="figures-1" class="section level3">
<h3>Figures</h3>
<div id="latlat" class="section level4">
<h4>Lat/Lat</h4>
<ul>
<li>Simulated phenotype according to latitudinal gradient</li>
<li>Test vector is latitude</li>
</ul>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-7-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="stratdlat" class="section level4">
<h4>stratD/Lat</h4>
<ul>
<li>Simulated phenotype according to gradient along the diagonal going from lower left to upper right<br />
</li>
<li>Project latitude as the test vector</li>
<li>For each simulation do 2 polygenic adaptation tests one with latitude as the test vector and one with longitude</li>
</ul>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-8-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="singledemelat" class="section level4">
<h4>singleDeme/Lat</h4>
<ul>
<li>Simulated phenotype difference in a single deme</li>
<li>Project latitude as the test vector</li>
<li>For each simulation do 1 polygenic adaptation test with latitude as the test vector</li>
</ul>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-9-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="singledemesingledeme" class="section level4">
<h4>singleDeme/singleDeme</h4>
<ul>
<li>Simulated phenotype difference in a single deme</li>
<li>Project single deme vs everyone else as test vector</li>
<li>For each simulation do 1 polygenic adaptation test with single deme vs everybody as the test vector</li>
</ul>
<p><img src="figure/Simulation_Methods_Reproducible.Rmd/unnamed-chunk-10-1.png" width="960" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] ggpubr_0.4.0      latex2exp_0.5.0   forcats_0.5.1     stringr_1.4.0    
 [5] dplyr_1.0.7       purrr_0.3.4       readr_2.1.1       tidyr_1.1.3      
 [9] tibble_3.1.6      ggplot2_3.3.5     tidyverse_1.3.1   data.table_1.14.2

loaded via a namespace (and not attached):
 [1] fs_1.5.2          lubridate_1.7.10  httr_1.4.2        rprojroot_2.0.2  
 [5] tools_3.6.2       backports_1.4.1   bslib_0.3.0       utf8_1.2.2       
 [9] R6_2.5.1          DBI_1.1.1         colorspace_2.0-2  withr_2.4.3      
[13] gridExtra_2.3     tidyselect_1.1.1  curl_4.3.2        compiler_3.6.2   
[17] git2r_0.28.0      cli_3.1.0         rvest_1.0.1       xml2_1.3.3       
[21] labeling_0.4.2    sass_0.4.0        scales_1.1.1      digest_0.6.29    
[25] foreign_0.8-72    rmarkdown_2.10    R.utils_2.10.1    rio_0.5.27       
[29] pkgconfig_2.0.3   htmltools_0.5.2   highr_0.9         dbplyr_2.1.1     
[33] fastmap_1.1.0     rlang_0.4.12      readxl_1.3.1      rstudioapi_0.13  
[37] jquerylib_0.1.4   generics_0.1.0    farver_2.1.0      jsonlite_1.7.2   
[41] zip_2.2.0         car_3.0-11        R.oo_1.24.0       magrittr_2.0.1   
[45] Rcpp_1.0.7        munsell_0.5.0     fansi_0.5.0       abind_1.4-5      
[49] lifecycle_1.0.1   R.methodsS3_1.8.1 stringi_1.7.6     whisker_0.4      
[53] yaml_2.2.1        carData_3.0-4     grid_3.6.2        promises_1.2.0.1 
[57] crayon_1.4.2      haven_2.4.3       cowplot_1.1.1     hms_1.1.1        
[61] knitr_1.37        pillar_1.6.4      ggsignif_0.6.2    reprex_2.0.1     
[65] glue_1.6.0        evaluate_0.14     modelr_0.1.8      vctrs_0.3.8      
[69] tzdb_0.2.0        httpuv_1.6.2      cellranger_1.1.0  gtable_0.3.0     
[73] assertthat_0.2.1  xfun_0.29         openxlsx_4.2.4    broom_0.7.9      
[77] rstatix_0.7.0     later_1.3.0       workflowr_1.6.2   ellipsis_0.3.2   </code></pre>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
