<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jennifer Blanc" />

<meta name="date" content="2020-10-15" />

<title>Theory</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">polygenic_adaptation_stratification</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jgblanc/polygenic_adaptation_stratification">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Theory</h1>
<h4 class="author">Jennifer Blanc</h4>
<h4 class="date">10/15/2020</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-01-13
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>polygenic_adaptation_stratification/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20201015code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20201015)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20201015code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20201015)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree2e49bfde494f3ea5988d664ac7d715a278f62162targetblank2e49bfda"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/2e49bfde494f3ea5988d664ac7d715a278f62162" target="_blank">2e49bfd</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjgblancpolygenicadaptationstratificationtree2e49bfde494f3ea5988d664ac7d715a278f62162targetblank2e49bfda" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/jgblanc/polygenic_adaptation_stratification/tree/2e49bfde494f3ea5988d664ac7d715a278f62162" target="_blank">2e49bfd</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .snakemake/conda-archive/
    Ignored:    .snakemake/conda/
    Ignored:    .snakemake/locks/
    Ignored:    .snakemake/scripts/
    Ignored:    .snakemake/shadow/
    Ignored:    .snakemake/singularity/
    Ignored:    code/.DS_Store
    Ignored:    code/.ipynb_checkpoints/
    Ignored:    code/Simulate_Genotypes/.ipynb_checkpoints/
    Ignored:    data/.DS_Store
    Ignored:    data/4PopSplit/
    Ignored:    output/Simulate_Genotypes/
    Ignored:    output/Simulate_Phenotypes/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Theory.Rmd</code>) and HTML (<code>docs/Theory.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/b01edbd051801fc1603fe413cda1a4f831ad68c5/docs/Theory.html" target="_blank">b01edbd</a>
</td>
<td>
jgblanc
</td>
<td>
2021-01-07
</td>
<td>
started pipeline
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/83fe4e9f43b93bf928f007a2bf0fe3a8a5af709f/analysis/Theory.Rmd" target="_blank">83fe4e9</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-26
</td>
<td>
added jupyter notebook
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/83fe4e9f43b93bf928f007a2bf0fe3a8a5af709f/docs/Theory.html" target="_blank">83fe4e9</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-26
</td>
<td>
added jupyter notebook
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/401d8dd52aefb13d571e5fc0109732544b24bf6b/docs/Theory.html" target="_blank">401d8dd</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jgblanc/polygenic_adaptation_stratification/854d4f7080c4700bbd1f0e700d7372a0c10e8628/docs/Theory.html" target="_blank">854d4f7</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jgblanc/polygenic_adaptation_stratification/blob/f10649af9ea0c80e583d8c9f590ac45a0863ce45/analysis/Theory.Rmd" target="_blank">f10649a</a>
</td>
<td>
jgblanc
</td>
<td>
2020-10-19
</td>
<td>
Theory
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="problem" class="section level2">
<h2>Problem</h2>
<p>Consider a true model where an indvidual’s phenotype is composed of their genetic component, an environmental variable, and random uncorrelated noise, such that the vector of phenotypes can be written as,</p>
<p><span class="math display">\[\mathrm{\vec{Y} = \vec{\beta}}\mathbf{M} \mathrm{+ \vec{E}+\vec{\epsilon}}.\]</span></p>
<p>Due to computational constraints and the fact that we typically have more SNPs than individuals <span class="math inline">\((\mathrm{\ell &gt; m})\)</span> in our sample, we do a series a univariate regressions, fitting the model below,</p>
<p><span class="math display">\[\mathrm{\vec{Y} = \beta_{\ell}}\mathbf{M_{\cdot \ell}} \mathrm{+ \vec{C}+\vec{\varepsilon}}, \]</span></p>
<p>where <span class="math inline">\(\mathbf{M_{\cdot \ell}}\)</span> is the <span class="math inline">\(\ell^{\text{th}}\)</span> column of the genotype matrix <span class="math inline">\(\mathbf{M}\)</span> and the coufounder <span class="math inline">\(\mathrm{\vec{C}}\)</span> is a vector of length <span class="math inline">\(m\)</span> that includes each individuals environmental variable <span class="math inline">\(\mathrm{E_m}\)</span> plus their genetic value at all other loci <span class="math inline">\(\mathrm{\sum\limits_{i\neq\ell}\beta_i}\mathbf{M_{\cdot i}}\)</span>.</p>
<p>For each site, the resulting regression coeffcient is an estimtate of the true effect size, <span class="math inline">\(\mathrm{\beta_{\ell}}\)</span> and can be written as,</p>
<p><span class="math display">\[\mathrm{\hat{\beta_{\ell}} = \mathrm{\frac{cov(\mathbf{M_{\cdot \ell}}, \vec{Y})}{var(\mathbf{M_{\cdot \ell}})}}}.\]</span></p>
<p>Replacing <span class="math inline">\(\mathrm{\vec{Y}}\)</span> with <span class="math inline">\(\mathrm{\beta_{\ell}}\mathbf{M_{\cdot \ell}} \mathrm{+ \vec{C}+\vec{\varepsilon}}\)</span> we can re-write the estimated effect size as,</p>
<p><span class="math display">\[\mathrm{\hat{\beta_{\ell}} = \mathrm{\frac{cov(\mathbf{M_{\cdot \ell}}, \beta_{\ell} \mathbf{M_{\cdot \ell}} \mathrm{+ \vec{C}+\vec{\varepsilon}})}{var(\mathbf{M_{\cdot \ell}})}}}\]</span></p>
<p><span class="math display">\[\mathrm{= \frac{cov(\mathbf{M_{\cdot \ell}},\beta_{\ell} \mathbf{M_{\cdot \ell}}) + cov(\mathbf{M_{\cdot \ell}},\vec{C}) + cov(\mathbf{M_{\cdot \ell}},\vec{\varepsilon})}{var(\mathbf{M_{\cdot \ell}})}}\]</span></p>
<p><span class="math display">\[\mathrm{= \frac{\beta_{\ell}var(\mathbf{M_{\cdot \ell}}) + cov(\mathbf{M_{\cdot \ell}},\vec{C}) + cov(\mathbf{M_{\cdot \ell}},\vec{\varepsilon})}{var(\mathbf{M_{\cdot \ell}})}}\]</span></p>
<p><span class="math display">\[\mathrm{= \beta_{\ell} + \frac{cov(\mathbf{M_{\cdot \ell}},\vec{C})}{var(\mathbf{M_{\cdot \ell}})} + \frac{cov(\mathbf{M_{\cdot \ell}},\vec{\varepsilon})}{var(\mathbf{M_{\cdot \ell}})}}.\]</span></p>
<p>Looking at the final line above, we can see that the estimated effect size from GWAS contains three components: the true effect size, the stratification effect, which comes from the covariance between genotype at SNP <span class="math inline">\(\mathrm{\ell}\)</span> and the counfounder <span class="math inline">\(\mathrm{\vec{C}}\)</span>, and the covariance between the genotype and the noise, which has expectation zero. Effect sizes will be biased if <span class="math inline">\(\mathbb{E}[\mathrm{\frac{cov(\mathbf{M_{\cdot \ell}},\vec{C})}{var(\mathbf{M_{\cdot \ell}})}] \neq 0}\)</span>.</p>
</div>
<div id="set-up" class="section level2">
<h2>Set-up</h2>
<p>We have two panels of genotyped individuals. First is the panel of individuals which the original GWAS was done in. The genotype matrix is denoted <span class="math inline">\(\underset{m\times L}{\mathbf{M}}\)</span> which consists of <span class="math inline">\(m\)</span> individuals genotyped at <span class="math inline">\(L\)</span> loci. Phenotypes for these <span class="math inline">\(m\)</span> individuals are stored in the vector <span class="math inline">\(\underset{1\times m}{\mathrm{\vec{Y}}}\)</span>.</p>
<p>Now we have a second set of non-overlapping (?) individuals that we want to use to test for evidence of polygenic adaptation along some axis. The genotype matrix for these <span class="math inline">\(n\)</span> individuals genotypes at the same <span class="math inline">\(L\)</span> loci is <span class="math inline">\(\underset{n\times L}{\mathbf{X}}\)</span>. Additionally, we have our test vector <span class="math inline">\(\underset{1\times n}{\mathrm{\vec{T}}}\)</span>. This vector represents any axis along which we want to test for polygenic adapation. For example, if we wanted to test if members of a specific population show evidence of polygenic adaptation, this vector would be a series of 1’s and 0’s indicating whether each individual belonged to that population or not. In another scenario we could use a continous variable (ex. an inidvidual’s latitude or position along a principle componet) as our test vector.</p>
<p>In order to test for polygenic adaptation, we ask if there is a significant covariance between an indvidials polygenic score (built from GWAS summary statistics) and the test vector in our test panel. In other words, is <span class="math inline">\(\text{cov}(\mathrm{\vec{\hat{\beta}}}\mathbf{X},\mathrm{\vec{T}})\)</span> significantly different from zero. Here the vector of effect sizes <span class="math inline">\(\underset{1\times L}{\mathrm{\vec{\hat{\beta}}}}\)</span>, are estimated from doing a GWAS in the original GWAS panel. <strong>conclusion sentence</strong></p>
</div>
<div id="theory" class="section level2">
<h2>Theory</h2>
<p>Consider our test statistic, the covariance bewteen polygenic scores calculated in the test panel using effect sizes estimated in the GWAS panel, <span class="math display">\[\text{cov}(\mathrm{\vec{\hat{\beta}}}\mathbf{X},\mathrm{\vec{T}}).\]</span> Noting that <span class="math inline">\(\text{cov}(\mathrm{\vec{\hat{\beta}}}\mathbf{X},\mathrm{\vec{T}}) =\mathrm{\hat{\beta}(T\mathbf{X})^T}\)</span> we can see that our test statistic amounts to multiplying the loading of genotypes in the test panel on the test vector by estimated effect size and suming across sites. In a situtation of tue polygenic adaptation along the test vector, we expect there to be a non-zero covariance between SNPs that effect the trait and the test vector at sites that have a true causal effect, creating a on-zero test statistic after summing accross sites. Here we ask if under of model no polygenic adaptation, if biased effect sizes can mimic the signal of polygenic adaptation.</p>
<p>In order to examine the effect of population stratification on our test statistics, let’s decompose the estimated effect, <span class="math display">\[\mathrm{(\vec{\beta} + \vec{\sigma_s} + \vec{\sigma_e})(\vec{T}\mathbf{X})^T}.\]</span></p>
<p>Here <span class="math inline">\(\mathrm{\vec{\beta}}\)</span> is a vector of true effect sizes while <span class="math inline">\(\mathrm{\vec{\sigma_S}}\)</span> is the effect size of stratification for each SNP. In other words, each entry of <span class="math inline">\(\mathrm{\vec{\sigma}}\)</span> is equal to, <span class="math display">\[\mathrm{\vec{\sigma_{s, \ell}} = \frac{cov(\mathbf{M_{\cdot \ell}},\vec{C})}{var(\mathbf{M_{\cdot \ell}})}}.\]</span></p>
<p>Similarly, <span class="math inline">\(\mathrm{\vec{\sigma_e}}\)</span>, is the effect size of random noise for each SNP where each entry is equal to,<br />
<span class="math display">\[\mathrm{\vec{\sigma_{e,\ell}} = \frac{cov(\mathbf{M_{\cdot \ell}},\vec{\varepsilon})}{var(\mathbf{M_{\cdot \ell}})}}.\]</span></p>
<p>Distributing <span class="math inline">\(\mathrm{(\vec{T}\mathbf{X})^T}\)</span> and recognizing that <span class="math inline">\(\mathrm{\vec{\sigma_s} = \vec{E}\mathbf{M}}\)</span> we can rewrite the test statistic as, <span class="math display">\[\mathrm{(\vec{\beta}\mathbf{X}^T\vec{T}^T + \vec{E} \mathbf{M}\mathbf{X}^T\vec{T}^T + \vec{\sigma_e}\mathbf{X}^T\vec{T}^T)}.\]</span></p>
<p>To examine the effect of stratification, let’s focus on the second term that includes the environmental confounder. This environmental variable causes bias in the GWAS effect size estimates when it covaries with genotypes in the GWAS panel. We want to know if this same variable also covaries with genotypes in the test panel. To do this we will decompose the environmental variable in to two parts, one part that is correlated with the loading of test panel genotypes on the test vector and another part that is uncorrelated,</p>
<p><span class="math display">\[\mathrm{\vec{E} = \gamma\vec{T}\mathbf{XM}^T + \vec{\mathcal{E}}}.\]</span></p>
<p>Plugging this expression for <span class="math inline">\(\mathrm{\vec{E}}\)</span> into the second term in our test statistic we get, <span class="math display">\[\mathrm{(\gamma\vec{T}\mathbf{XM}^T + \vec{\varepsilon})\mathbf{M}\mathbf{X}^T\vec{T}^T}\]</span> <span class="math display">\[\mathrm{\gamma\vec{T}\mathbf{XM}^T\mathbf{M}\mathbf{X}^T\vec{T}^T + \vec{\mathcal{E}}\mathbf{M}\mathbf{X}^T\vec{T}^T}.\]</span></p>
<p>We can now write our final test statistic as, <span class="math display">\[\mathrm{\vec{\beta}\mathbf{X}^T\vec{T}^T + \gamma\vec{T}\mathbf{XM}^T\mathbf{M}\mathbf{X}^T\vec{T}^T + \epsilon_f}\]</span></p>
<p>Looking at the equation above, we expect popultaion stratification in the GWAS panel to bias our test for polygenic adaptation if <span class="math inline">\(\mathrm{\gamma \neq 0}\)</span> and <span class="math inline">\(\vec{T}\mathbf{XM}^T\mathbf{M}\mathbf{X}^T\vec{T}^T \neq 0\)</span>.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] workflowr_1.6.2 Rcpp_1.0.4.6    rprojroot_1.3-2 digest_0.6.25  
 [5] later_1.1.0.1   R6_2.4.1        backports_1.1.8 git2r_0.27.1   
 [9] magrittr_1.5    evaluate_0.14   stringi_1.4.6   rlang_0.4.6    
[13] fs_1.4.1        promises_1.1.1  whisker_0.4     rmarkdown_2.3  
[17] tools_3.6.2     stringr_1.4.0   glue_1.4.1      httpuv_1.5.4   
[21] xfun_0.15       yaml_2.2.1      compiler_3.6.2  htmltools_0.5.0
[25] knitr_1.29     </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
