---
title: "Calculate TGWAS"
author: "Jennifer Blanc"
date: "4/14/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(autodep = TRUE)
library(pgenlibr)
library(data.table)
library(dplyr)
library(pracma)
```

## Calculate $T^{GWAS}$

Here we will walk through our procedure for estimating $T^{GWAS}$.

### Make Test Vector 

First we need to make the test vector along which will we test for selection. In our 4 population toy model this is simply population ID so we create a $n \times1$ vector, where $n$ is the number of individuals in the test panel, where individuals in the first test population are coded as $1$ and individuals in the second test population are $0$. We then mean center the test vector. 

```{python, eval = FALSE}
# Snakemake rule 
rule make_test_vector:
    input:
        pops="output/Simulate_Genotypes/4PopSplit/{rep}/genos.pop",
        fam="output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-test_common.psam"
    output:
        "output/Calculate_Tm/4PopSplit/{rep}/{config}/Tvec.txt"
    shell:
        "Rscript code/Calculate_Tm/4PopSplit_make_tvec.R {input.pops} {input.fam} {output}"
```

Example:  
```{r, warning=FALSE, message=FALSE}
# Read in Fam file
fam <- fread("../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.psam")
colnames(fam) <- c("FID", "IID", "SEX")

# Read in population information 
pops <- fread("../output/Simulate_Genotypes/4PopSplit/B1/genos.pop", header = F)
colnames(pops)  <- c("FID", "IID", "POP")

# Join tables
pop <- dplyr::inner_join(pops, fam) %>% dplyr::select(c("FID", "IID", "POP"))
test_pops <- unique(pop$POP)

# Make test vector as a new column
pop <- pop %>% mutate(Tvec = case_when(POP == test_pops[1] ~ 1, POP == test_pops[2] ~ 0)) %>% mutate(Tvec = Tvec - mean(Tvec))

# Dispaly test vector 
table(pop$POP, pop$Tvec)
```


### Compute covariate 

In the main text we derive a single covariate, $T^{GWAS}$, to control for stratification specific to the test of interest. 

$$\vec{T}^{GWAS} = GX^T\vec{T}$$

Here $G$ is the GWAS genotype matrix while $X$ is the test genotype matrix. Due to the size of the genotype matrices we cannot simply multiply the matrices in R. Instead we use plink2 to do the matrix multiplication. 

Specfically, we can first use `--glm` to get the product of $X^{T}\vec{T}$. This amounts to running a general linear model, regressing the test vector onto test panel genotypes, to get the covariance between the test vector and each SNP divided by the variance of the genotype, $\frac{Cov(g_{\ell}, \vec{T})}{Var(g_{\ell})}$. Because we only want the covariance, we use `--geno-counts` to get the proportion of each genotype in the test panel. This allows us to compute the variance at each site and adjust the effect sizes estimated by plink2. We can then use `--score` and these adjusted effect sizes to compute $GX^{T}\vec{T}$. 


```{python, eval = FALSE}
# Snakemake rules
rule proj_T:
    input:
        test="output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-test_common.psam",
        gwas="output/Simulate_Genotypes/4PopSplit/{rep}/{config}/genos-gwas_common.psam",
        tvec="output/Calculate_Tm/4PopSplit/{rep}/{config}/Tvec.txt"
    output:
        "output/Calculate_Tm/4PopSplit/{rep}/{config}/Tm.txt"
    shell:
        """
        Rscript code/Calculate_Tm/calc_GXT.R  output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/{wildcards.config}/genos-test_common output/Simulate_Genotypes/4PopSplit/{wildcards.rep}/{wildcards.config}/genos-gwas_common {input.tvec} output/Calculate_Tm/4PopSplit/{wildcards.rep}/{wildcards.config}/
        """
```


Next we walk through our 4 population example. In this example we are working in the configuration where there is shared structure between the GWAS and test panels (i.e Figure 1A). Here our test vector is separating populations on opposite sides of the deepest split. Because there is overlapping structure and shared drift between panels, we expect that $\vec{T}^{GWAS}$ will also separate the two population that comprise the GWAS panel.   

Let's read in both the test and GWAS genotype matrices. 
```{r}
# Read GWAS Matrix 
pvar <- NewPvar("../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pvar")
d1 <- NewPgen("../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common.pgen")
G <- ReadList(d1,seq(1,10905), meanimpute=F)


# Read in Test Matrix 
pvar <- NewPvar("../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.pvar")
d1 <- NewPgen("../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common.pgen")
X <- ReadList(d1,seq(1, 10905), meanimpute=F)
```

First, let's calculate $\vec{T}^{GWAS}$ just by doing matrix multiplication. 
```{r, cache=TRUE}
Tvec <- pop$Tvec
TGWAS_mat <-  G %*% t(X) %*% Tvec
TGWAS_mat <- scale(TGWAS_mat)
plot(TGWAS_mat)
```
We can see that $\vec{T}^{GWAS}$ is recapitulating the population structure we expect. 

Now let's check that the our plink procedure get's us the same answer.  
```{r, results=FALSE}
# Compute t(X)T
cmd_XT <- paste0("~/plink2 --glm omit-ref allow-no-covars --pfile ../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-test_common --pheno ../output/Calculate_Tm/4PopSplit/B1/C1/Tvec.txt --pheno-name Tvec --geno-counts --out ../output/Calculate_Tm/4PopSplit/B1/C1/xt_temp")
system(cmd_XT)

# Adjust Betas to account for variance in x

beta_plink <- fread(paste0("../output/Calculate_Tm/4PopSplit/B1/C1/", "xt_temp.Tvec.glm.linear"))
count_plink <- fread(paste0("../output/Calculate_Tm/4PopSplit/B1/C1/", "xt_temp.gcount"))

# Calculate variance of each genotype 
nOBS <- (count_plink$HOM_REF_CT + count_plink$HET_REF_ALT_CTS + count_plink$TWO_ALT_GENO_CTS)
counts <- (count_plink$HOM_REF_CT * 0) + (count_plink$HET_REF_ALT_CTS * 1) + (count_plink$TWO_ALT_GENO_CTS * 2)
mean_gc <- counts / nOBS
length_mc_genos <- (count_plink$HOM_REF_CT * (mean_gc)^2) + (count_plink$HET_REF_ALT_CTS * (1 - mean_gc)^2) +  (count_plink$TWO_ALT_GENO_CTS * (2 - mean_gc)^2)

# Fix betas
betas_plink_norm <- beta_plink$BETA * length_mc_genos

#  Re-write .linear file with correct betas
beta_plink$BETA <- betas_plink_norm
beta_reformat <- beta_plink %>% dplyr::select(ID, A1, BETA)
fwrite(beta_reformat, paste0("../output/Calculate_Tm/4PopSplit/B1/C1/", "xt_temp.Tvec.glm.linear"), sep = "\t")

# Compute Gt(X)T
cmd_b <- paste("~/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/B1/C1/genos-gwas_common --score ../output/Calculate_Tm/4PopSplit/B1/C1/xt_temp.Tvec.glm.linear header-read center cols=dosagesum,scoresums --out ../output/Calculate_Tm/4PopSplit/B1/C1/b")
system(cmd_b)

# Read in and return TGWAS
TGWAS_plink = fread(paste0("../output/Calculate_Tm/4PopSplit/B1/C1/", "b.sscore"))
TGWAS_plink = as.matrix(TGWAS_plink$BETA_SUM)
TGWAS_plink = scale(TGWAS_plink)
```

```{r}
plot(TGWAS_mat, TGWAS_plink)
abline(0,1, col = "red")
```


