---
title: "Home"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
editor_options:
  chunk_output_type: console
---

# Guaranteeing unbiasedness in selection tests based on polygenic scores

Link to manuscript: https://www.overleaf.com/project/5ff4eec3e5325e0db841f6eb 

## Introduction 

This website contains all the code used for the analysis.  

## Simulation Pipeline 

This pipline is written in snakemake. To run the entire simulation pipeline you simply need to swap in the correct snakefile.   

* `code/Snakefiles/4PopSplit/snakefile` is the correct file to run the 4 population toy model from Figure 1.  
* `code/Snakefiles/SimpleGrid/snakefile` runs the stepping stone grid simulations and has options to simulate the confounders and do the polygenic adaptation tests from Figure 2.    

The detailed methods and results for each simulation can be found [here](Simulation_Methods_Reproducible.html).  

The pages below explain the 5 major steps of the pipeline:  

1. [Simulate Genotypes](Simulate_Genotypes.html) 
2. [Simulate Phenotypes](Simulate_Phenotypes.html) 
3. [Project Test Vector](Calculate_TGWAS.html) 
4. [Run GWAS and Ascertain SNPs](Run_GWAS.html)
5. [Polygenic Adaptation Test](Polygenic_Selection_Test.html) 


<br />
<br />
<br />
<br />
<br />
<br />
<br />
<br />

  
WIP links  
Reproducible figures [link](Simulation_Methods_Reproducible.html)
Theory [link](Theory.html)  
4popslit [link](4Popsplit.html)  
scratch [link](scratch.html)  
test_vec [link](Test_vec.html)  
Plot_4PopSplit [link](Plot_4PopSplit.html)   
projection_example [link](projection_example.html)  
plink2 projection [link](Projection_plink2.html)  
calc Qx [link](Calc_Qx.html)   
Plot_SimpleGrid [link](Plot_SimpleGrid.html)   
methods [link](Simulations_methods.html)  
Meeting Notes [link](Meetings_Notes.html)  
Proof [link](Proof.html)

