---
title: "Plot_SimpleGrid"
author: "Jennifer Blanc"
date: "6/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list =ls())
library(snpStats)
library(data.table)
library(ggplot2)
library(dplyr)
library(pgenlibr)
library(tidyverse)
library(knitr)
library(latex2exp)
```


# Grid Genotype Simulaltion 

* Fst = 0.015  
* 36 demes on a 6 x 6 grid  
* Number of SNPs = 184,228  
* Used a pruned set of 2402 SNPS to do PCA  
* 432 GWAS individuals 


```{r}
Mod <- read.plink("../output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-gwas_common.cmpruned.b")
M <- as.matrix(Mod$genotypes) 
cov_mat <- xxt(M)
myE <- eigen(cov_mat)

pops <- fread("~/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E2/genos.pop")
colnames(pops) <- c("IID", "#FID", "Pop", "Lat", "Long")

df <- as.data.frame(cbind(myE$vectors[,1], myE$vectors[,2], Mod$fam[,1]))
colnames(df) <- c("PC1", "PC2", "IID")
df <- inner_join(df, pops)
df$PC1 <- as.numeric(as.character(df$PC1))
df$PC2 <- as.numeric(as.character(df$PC2))
ggplot(df, aes(x=PC1, y=PC2, color = Lat)) + geom_point() + theme_classic()
```

# Phenotype simulalation 

No phenotypic difference across Latitude 
```{r}
phenos <- fread("../output/Simulate_Phenotypes/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.phenos.txt")
df <- suppressMessages(inner_join(phenos, pops))
ggplot(data = df, aes(Lat, pheno_strat)) + geom_point() + geom_smooth() + theme_bw() + ylab("Phenotype")
```

```{r}
ggplot(df, aes(x = Lat, y = Long, fill = pheno_strat)) + geom_tile() + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```

2 unit phenotypic difference across Latitude
```{r}
phenos <- fread("../output/Simulate_Phenotypes/SimpleGrid/E1/C1/h2-0/env-2/genos-gwas_common.phenos.txt")
df <- suppressMessages(inner_join(phenos, pops))
ggplot(data = df, aes(Lat, pheno_strat)) + geom_point() + geom_smooth() + theme_bw() + ylab("Phenotype")
```

```{r}
ggplot(df, aes(x = Lat, y = Long, fill = pheno_strat)) + geom_tile() + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```


# PGS in test panel 

No stratification
```{r}
pgs <- fread("../output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/PGS.txt")
df <- inner_join(pgs, pops) %>% gather(key = "Type", value = "PRS" , c, c.p, nc, c_Tm, c.p_Tm, nc_Tm)
df <- arrange(transform(df,Type=factor(Type,levels=c("c", "c.p", "nc", "c_Tm", "c.p_Tm", "nc_Tm"))),Type)
ggplot(data = df, aes(Lat, PRS)) + geom_point() + geom_smooth() + theme_bw()  + facet_wrap(~Type)
```

```{r}
df_pop <- df %>% group_by(Type, Pop) %>% mutate(avg_PRS = mean(PRS)) %>% ungroup()
ggplot(df_pop, aes(x = Lat, y = Long, fill = avg_PRS)) + geom_tile() + facet_wrap(~Type) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```

```{r}
Qx <- fread("../output/PGA_test/SimpleGrid/E1/C1/h2-0/env-0.0/Qx.txt")
Qx
```


Stratification
```{r}
pgs <- fread("../output/PRS/SimpleGrid/E1/C1/h2-0/env-2/PGS.txt")
df <- inner_join(pgs, pops) %>% gather(key = "Type", value = "PRS" , c, c.p, nc, c_Tm, c.p_Tm, nc_Tm)
df <- arrange(transform(df,Type=factor(Type,levels=c("c", "c.p", "nc", "c_Tm", "c.p_Tm", "nc_Tm"))),Type)
ggplot(data = df, aes(Lat, PRS)) + geom_point() + geom_smooth() + theme_bw()  + facet_wrap(~Type)
```

```{r}
df_pop <- df %>% group_by(Type, Pop) %>% mutate(avg_PRS = mean(PRS)) %>% ungroup()
df_pop <- arrange(transform(df_pop,Type=factor(Type,levels=c("c", "c.p", "nc", "c_Tm", "c.p_Tm", "nc_Tm"))),Type)
ggplot(df_pop, aes(x = Lat, y = Long, fill = avg_PRS)) + geom_tile() + facet_wrap(~Type) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```

```{r}
Qx <- fread("../output/PGA_test/SimpleGrid/E1/C1/h2-0/env-2/Qx.txt")
kable(Qx)
```

# 100 Simulations 

Look at number of false positives for $Q_x$ test using a p-value calculated from an empirical null (randomly flipping effect size signs).  

Clumped sites
```{r}
All <- fread("../code/plot/F100_SimpleGrid.txt")

nc <- All %>% filter(type == "nc"| type == "nc-Tm") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-") %>% filter(case == "C1")
nc$diff <- as.numeric(nc$diff)

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3) + xlab("Env Shift") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0) + geom_hline(yintercept = 0.05, color = "red")
```

Random Sites
```{r}
All <- fread("../code/plot/F100_SimpleGrid.txt")

c <- All %>% filter(type == "c"| type == "c-Tm") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-") %>% filter(case == "C1")
c$diff <- as.numeric(c$diff)

ggplot(data = c, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3) + xlab("Env Shift") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0) + geom_hline(yintercept = 0.05, color = "red")
```

# Look at code used to calculate emprical null for a single example

Calculate Qx
```{r}
num = 1000
read_genos <- function(geno_prefix, betas) {

  pvar <- pgenlibr::NewPvar(paste0(geno_prefix, ".pvar"))
  d1 <- pgenlibr::NewPgen(paste0(geno_prefix, ".pgen"))
  var.ids <- betas$ID
  var.indx <- rep(0, length(var.ids))
  for (i in 1:length(var.indx)) {
    var.indx[i] <- pgenlibr::GetVariantsById(pvar,var.ids[i])
  }
  X <- ReadList(d1,var.indx, meanimpute=F)

  return(X)
}

pgs <- function(X, betas) {

  Bhat_strat <- betas$BETA_Strat
  Z_strat <- X %*% Bhat_strat

  out <- cbind(Z_strat)
  colnames(out) <- c("STRAT")

  return(out)
}

# Read in PGS w/ clumped sites
betas <- fread("../output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.nc.betas")
X <- read_genos("../output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-test_common", betas)
sscore <- pgs(X, betas)

# Calculate Qx
calc_Qx <- function(mprs, tvec, Va, lambda_T) {

  # Compute Qx Strat
  Ztest <- t(tvec) %*% mprs$strat.adjusted
  Qx_strat <- (t(Ztest) %*% Ztest) / (Va*lambda_T)

  return(Qx_strat)
}


tvec <- fread("../output/Calculate_Tm/SimpleGrid/E1/C1/Tvec.txt")
tvec <- tvec$V1
lambda_T <- fread("../output/Calculate_Tm/SimpleGrid/E1/C1/Lambda_T.txt")
lambda_T <- as.numeric(lambda_T[1,1])
mprs <- as.data.frame(sscore)
colnames(mprs) <- "strat.adjusted"
Va <- fread("../output/PGA_test/SimpleGrid/E1/C1/h2-0/env-0.0/Va.txt")

Qx = calc_Qx(mprs, tvec,Va[3,2] , lambda_T)
Qx
```

Generate Empirical Null 
```{r}
# Function to flip effect sizes
flip <- function(betas) {
  new_betas <- sample(c(-1,1), length(betas),  replace = T) * betas
  return(new_betas)
}

# Function to flip effect sizes and recompute Qx
en <- function(betas, tvec, Va, X, lambda_T) {

  # Flip effect sizes
  betas$BETA_Strat <- flip(betas$BETA_Strat)

  # Calculate PGS
  prs <- pgs(X, betas)
  prs <- as.data.frame(prs)
  colnames(prs) <- "strat.adjusted"

  # Calculate Qx
  Qx <- t(calc_Qx(prs, tvec, Va, lambda_T))

  return(Qx)
}

# Generate Empirical null
redraws <- matrix(0, ncol = 1, nrow = num)
for (i in 1:num){
    redraws[i,] <- en(betas, tvec, Va[3,2], X, lambda_T)
}

hist(redraws)
abline(v = Qx, col = "red")

all_strat <- redraws[,1]
p_strat_en <- length(all_strat[all_strat > as.numeric(Qx)])/length(all_strat)
p_strat_en
```

# Looking at allele frequencies 

```{r}
betas <- fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.nc.betas")
Mnc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-gwas_common", betas)

hist(colMeans(Mnc)/2, breaks = seq(from=0, to=1, by=.05))
```

```{r}
betas <- fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.c.betas")
Mc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-gwas_common", betas)

hist(colMeans(Mc)/2, breaks = seq(from=0, to=1, by=.05))
```

```{r}
frq_c <- sort(colMeans(Mc)/2)
frq_nc <- sort(colMeans(Mnc)/2)

plot(sort(log10(frq_c)))
points(sort(log10(frq_nc)), col = "red")
```

```{r}
betas_c <- fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.c.betas")
betas_nc <- fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.nc.betas")


betas_c <- betas_c %>% mutate(SNP_A = ID) %>% separate(ID, c("CHR", "POS", "A", "B"), "_") %>% filter(CHR == 1) 
betas_c$POS <- as.numeric(as.character(betas_c$POS))
betas_c <- betas_c[order(POS),]
betas_nc <- betas_nc %>% mutate(SNP_A = ID) %>% separate(ID, c("CHR", "POS", "A", "B"), "_") %>% filter(CHR == 1)
betas_nc$POS <- as.numeric(as.character(betas_nc$POS))
betas_nc <- betas_nc[order(POS),]

plot(betas_c$POS, betas_c$BETA_Strat)
points(betas_nc$POS, betas_nc$BETA_Strat, col = "red")
```


```{r}
test_c <- betas_c%>% 
  mutate(pairwise = POS - lag(POS), COV = BETA_Strat * lag(BETA_Strat), SNP_B = lag(SNP_A) )

test_nc <- betas_nc%>% 
  mutate(pairwise = POS - lag(POS), COV = BETA_Strat * lag(BETA_Strat), SNP_B = lag(SNP_A)  )

plot(test_nc$pairwise, test_nc$COV)
points(test_c$pairwise, test_c$COV, col = "red")
```

```{r}
X <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-test_common", fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.nc.betas"))

frq <- colMeans(X) /2
LD <- rep(NA, 100)
for (i in 1:99) {
  LD[i+1] <- cor(X[,i], X[,i+1]) 
}

test_nc$LD <- LD
test_nc$prod <- test_nc$COV * test_nc$LD

plot(test_nc$pairwise, test_nc$prod)
```

```{r}
Xc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-test_common", fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.c.betas"))

LD <- rep(NA, 100)
for (i in 1:99) {
  LD[i+1] <- cor(Xc[,i], Xc[,i+1]) 
}

test_c$LD <- LD
test_c$prod <- test_c$COV * test_c$LD

plot(test_c$pairwise, test_c$prod)
```


```{r}
plot(test_nc$pairwise, test_nc$prod)
points(test_c$pairwise, test_c$prod, col = "red")
```

