---
title: "Plot_4PopSplit"
author: "Jennifer Blanc"
date: "1/21/2021"
output: html_document
---

```{r setup, include=FALSE}
rm(list=ls())
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(tidyverse)
library(knitr)
library(latex2exp)
```

## Introduction


# Configuration 1

A/C are the GWAS panel  
B/D are the test panel  

Stratification in the GWAS pannel should effect PGS in the test population. Since there was an environmental effedt in A Population B should have greater PGS than D in the test panel. 

```{r}
# Read in all Reps and Pick important columns 
agg_data <- function(rep, file_path, file_name) {
  gvalue <- fread(paste0(file_path, rep,file_name ))
  gwas.pop <- fread(paste0("../output/Simulate_Genotypes/4PopSplit/", rep, "/C1/ids.gwas"), header = F)
  colnames(gwas.pop) <- c("#IID", "FID", "POP")
  gvalue <- suppressMessages(full_join(gvalue,gwas.pop))
  gvalue$REP <- rep
  return(gvalue)
}

reps <- list("E1", "E2")
cols_gwas <- c("red2", "mediumblue")

df_pheno_env0 <- plyr::ldply(reps, agg_data, file_path='../output/Simulate_Phenotypes/4PopSplit/', file_name='/C1/h2-0/env-0.0/genos-gwas_common.phenos.txt')
df_pheno_env0$env <- 0
df_pheno_env0.1 <- plyr::ldply(reps, agg_data, file_path='../output/Simulate_Phenotypes/4PopSplit/', file_name='/C1/h2-0/env-0.10/genos-gwas_common.phenos.txt')
df_pheno_env0.1$env <- 0.1
df_pheno_env0.5 <- plyr::ldply(reps, agg_data, file_path='../output/Simulate_Phenotypes/4PopSplit/', file_name='/C1/h2-0/env-0.50/genos-gwas_common.phenos.txt')
df_pheno_env0.5$env <- 0.5
df_pheno_env1 <- plyr::ldply(reps, agg_data, file_path='../output/Simulate_Phenotypes/4PopSplit/', file_name='/C1/h2-0/env-1.00/genos-gwas_common.phenos.txt')
df_pheno_env1$env <- 1

df_pheno <- rbind(df_pheno_env0, df_pheno_env0.1, df_pheno_env0.5, df_pheno_env1)

ggplot(data = df_pheno, aes(x=pheno_strat, fill = POP)) + geom_histogram(data = subset(df_pheno, POP == "A"), alpha = 0.75, bins = 30) + geom_histogram(data = subset(df_pheno, POP == "C"), alpha = 0.75, bins = 30) + facet_grid(REP ~ env) + theme_bw() + xlab("Stratified Phenotype") + scale_fill_manual(values = cols_gwas)
```

Polygenic Score in Test Population (clumping) 
```{r}
# Read in all Reps and Pick important columns 
agg_data <- function(rep, file_path, file_name) {
  gvalue <- fread(paste0(file_path, rep,file_name ))
  gwas.pop <- fread(paste0("../output/Simulate_Genotypes/4PopSplit/", rep, "/C1/ids.test"), header = F)
  colnames(gwas.pop) <- c("IID", "FID", "POP")
  gvalue <- suppressMessages(full_join(gvalue,gwas.pop))
  gvalue$REP <- rep
  return(gvalue)
}

df_test_env0 <- plyr::ldply(reps, agg_data, file_path='../output/PGA_test/4PopSplit/', file_name='/C1/h2-0/env-0.0/PGS.txt')
df_test_env0$env <- 0
df_test_env0.1 <- plyr::ldply(reps, agg_data, file_path='../output/PGA_test/4PopSplit/', file_name='/C1/h2-0/env-0.10/PGS.txt')
df_test_env0.1$env <- 0.1
df_test_env0.5 <- plyr::ldply(reps, agg_data, file_path='../output/PGA_test/4PopSplit/', file_name='/C1/h2-0/env-0.50/PGS.txt')
df_test_env0.5$env <- 0.5
df_test_env1 <- plyr::ldply(reps, agg_data, file_path='../output/PGA_test/4PopSplit/', file_name='/C1/h2-0/env-1.00/PGS.txt')
df_test_env1$env <- 1
cols_test <- c("darkgoldenrod1", "darkorchid1")

df_test <- rbind(df_test_env0, df_test_env0.1, df_test_env0.5, df_test_env1)

# Histogram 
ggplot(data = df_test, aes(x=nc, fill = POP)) + geom_histogram(data = subset(df_test, POP == "B"), alpha = 0.75, bins = 30) + geom_histogram(data = subset(df_test, POP == "D"), alpha = 0.75, bins = 30) + facet_grid(REP ~ env) + theme_bw() + xlab("Stratified Phenotype") + scale_fill_manual(values = cols_test)
```

Polygenic Score in Test Population, including Tm (clumping) 
```{r}
ggplot(data = df_test, aes(x=nc_Tm, fill = POP)) + geom_histogram(data = subset(df_test, POP == "B"), alpha = 0.75, bins = 30) + geom_histogram(data = subset(df_test, POP == "D"), alpha = 0.75, bins = 30) + facet_grid(REP ~ env) + theme_bw() + xlab("Stratified Phenotype") + scale_fill_manual(values = cols_test)
```

Qx stats
```{r}
reps <- rep(NA, 2)
for (i in 1:2){reps[i] <- paste0("E", i)}
h2 <- "h2-0"
envs <- c("env-0.0", "env-0.10", "env-0.50", "env-1.00")
cases <- c("C1")
dat <- expand.grid(reps, cases, h2, envs)
colnames(dat) <- c("rep", "case", "h2", "env")

agg_all_data <- function(rep, dir_path, case, type, h2, env) {
  
  Qx <- fread(paste0(dir_path, rep,"/", case, "/", h2, "/", env, "/", "Qx.txt"))
  Qx$type <- c("c", "c.p", "nc", "c-Tm", "c.p-Tm", "nc-Tm")
  
  return(Qx)
}
df <- plyr::mdply(dat, agg_all_data, dir_path = '../output/PGA_test/4PopSplit/' )
df_null <- df %>% filter(type == "nc" & env == "env-0.0")
kable(df_null)


df_Qx <- df %>% group_by(env, type) %>% summarise(avg_Ax = mean(Ax))
ggplot(data = df_Qx, aes(x = type, y = env, fill = avg_Ax)) +geom_tile() + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient(low = "lightgoldenrod1", high = "blue")
```

# Look over 100 simulations 

Original Simulations  

- 100 Inidividuals per population (200 per panel)  
- N = 10000 (ancestral and individual pops)  
- L = 10000000 (1 MB) - 2 chromosomes  
- mu = 1e-8  
- rho = 1e-8  

Clumping  

- Pick lowested p-value in predefined 100Kb windows  
- Always results in 200 sites included in PRS   

```{r}
df <- fread("../code/plot/K100_chrs.txt")
nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
kable(nc[1:2,1:5])
nc$diff <- as.numeric(nc$diff)


ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+ xlab("Mean(A) - Mean(C)") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") 
#+ geom_smooth(method = "loess", se = FALSE) 
```

New Clumping Procedure  

- Iteratively pick lowest possible p-value and eliminates SNPs within 100KB on either side  
- Keep going until all SNPs have been eliminated  
- Results in ~150 SNPs included  

```{r}
df <- fread("../code/plot/F100_NewClump.txt")
nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
kable(nc[1:2,1:5])
nc$diff <- as.numeric(nc$diff)

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+ xlab("Mean(A) - Mean(C)") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red")
```

New Genotype Simulations  

- 100 Inidividuals per population (100 per panel)  
- N = 10000 (ancestral and individual pops)  
- L = 10000000 (1 MB) - 2 chromosomes  
- mu = 1e-8  
- rho = 1e-7   

Same Clumping  

- ~150 SNPs included in PRS  

```{r}
df <- fread("../code/plot/J100.txt")
nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
kable(nc[1:2,1:5])
nc$diff <- as.numeric(nc$diff)

test <- df %>% filter(type == "nc") %>% filter(case == "C1" & env == "env-0.0")

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+ xlab("Mean(A) - Mean(C)") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red")
```

New Genotype Simulations  

- 200 Inidividuals per population (400 per panel)  
- N = 20000 (ancestral and individual pops)  
- L = 10000000 (1 MB) - 2 chromosomes  
- mu = 1e-8  
- rho = 1e-7  

Same Clumping  

- ~150 SNPs included in PRS  

```{r}
df <- fread("../code/plot/K100.txt")
nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
kable(nc[1:2,1:5])
nc$diff <- as.numeric(nc$diff)

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+ xlab("Mean(A) - Mean(C)") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red")
```


New Clumping  

- Eliminate 200 KB on either side of included SNP  

```{r}
df <- fread("../code/plot/K100_widewindow.txt")
nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
kable(nc[1:2,1:5])
nc$diff <- as.numeric(nc$diff)

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+ xlab("Mean(A) - Mean(C)") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red")
```

New Clumping  

- Eliminate 300 KB on either side of included SNP  

```{r}
df <- fread("../code/plot/K100_widewindow3.txt")
nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
kable(nc[1:2,1:5])
nc$diff <- as.numeric(nc$diff)

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+ xlab("Mean(A) - Mean(C)") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red")
```

New Clumping  

- Eliminate 500 KB on either side of included SNP  
- ~ 30 SNPs included in PRS  

```{r}
df <- fread("../code/plot/K100_widewindow5.txt")
nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
kable(nc[1:2,1:5])
nc$diff <- as.numeric(nc$diff)

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+ xlab("Mean(A) - Mean(C)") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red")
```

Yay? Seems like a huge window   
Implement LD based clumping instead of physical distance?  
Increase sample size?  




```{r}
mat <- rep(NA, 5000)
for (i in 1:5000) {
  mat[i] <- paste0(i, "\t", 1)
}
```


```{r}
mat <- as.data.frame(mat)
fwrite(mat, file = "~/Desktop/convert_chr.txt", row.names = F, col.names = F, quote = F)
```


# Looking at LD 

```{r}
read_genos <- function(geno_prefix, betas) {

  pvar <- NewPvar(paste0(geno_prefix, ".pvar"))
  d1 <- NewPgen(paste0(geno_prefix, ".pgen"))
  var.ids <- betas$ID
  var.indx <- rep(0, length(var.ids))
  for (i in 1:length(var.indx)) {
    var.indx[i] <- pgenlibr::GetVariantsById(pvar,var.ids[i])
  }
  X <- ReadList(d1,var.indx, meanimpute=F)

  return(X)
}
```

```{r}
betas_nc <- fread("~/polygenic_adaptation_stratification/output/PRS/4PopSplit/E1/C1/h2-0/env-0.0/genos-gwas_common.nc.betas")
Mnc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/E1/C1/genos-gwas_common", betas_nc)

betas_c <- fread("~/polygenic_adaptation_stratification/output/PRS/4PopSplit/E1/C1/h2-0/env-0.0/genos-gwas_common.c.betas")
Mc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/E1/C1/genos-gwas_common", betas_c)
```



```{r}
X_nc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/E4/C1/genos-test_common",  betas = fread("~/polygenic_adaptation_stratification/output/PRS/4PopSplit/E4/C1/h2-0/env-0.0/genos-gwas_common.nc.betas"))

frq_nc <- colMeans(X_nc) /2
LD_nc <- rep(NA, length(frq_nc))
for (i in 1:(length(frq_nc) -1)) {
  LD_nc[i+1] <- cor(X_nc[,i], X_nc[,i+1]) 
}

X_c <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/4PopSplit/E4/C1/genos-test_common",  betas = fread("~/polygenic_adaptation_stratification/output/PRS/4PopSplit/E4/C1/h2-0/env-0.0/genos-gwas_common.c.betas"))

frq_c <- colMeans(X_c) /2
LD_c <- rep(NA, length(frq_c))
for (i in 1:(length(frq_c) -1)) {
  LD_c[i+1] <- cor(X_c[,i], X_c[,i+1]) 
}

hist(LD_c)
hist(LD_nc)

plot(sort(LD_c), sort(LD_nc))
abline(a=0, b=1)

hist(frq_c)
hist(frq_nc)

plot(sort(frq_c), sort(frq_nc))
abline(a=0, b=1)
```




