---
title: "Plot_SimpleGrid"
author: "Jennifer Blanc"
date: "6/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list =ls())
library(snpStats)
library(data.table)
library(ggplot2)
library(dplyr)
library(pgenlibr)
library(tidyverse)
library(knitr)
library(latex2exp)
```


# Grid Genotype Simulaltion 

* Fst = 0.015  
* 36 demes on a 6 x 6 grid  
* Number of SNPs = 184,228  
* Used a pruned set of 2402 SNPS to do PCA  
* 432 GWAS individuals 


```{r}
Mod <- read.plink("../output/Simulate_Genotypes/SimpleGrid/E6/C1/genos-gwas_common.cmpruned.b")
M <- as.matrix(Mod$genotypes) 
cov_mat <- xxt(M)
myE <- eigen(cov_mat)

pops <- fread("~/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E6/genos.pop")
colnames(pops) <- c("IID", "#FID", "Pop", "Lat", "Long")

df <- as.data.frame(cbind(myE$vectors[,1], myE$vectors[,2], Mod$fam[,1]))
colnames(df) <- c("PC1", "PC2", "IID")
df <- inner_join(df, pops)
df$PC1 <- as.numeric(as.character(df$PC1))
df$PC2 <- as.numeric(as.character(df$PC2))
ggplot(df, aes(x=PC1, y=PC2, color = Lat)) + geom_point() + theme_classic()
```

# Phenotype simulalation 

No phenotypic difference across Latitude 
```{r}
phenos <- fread("../output/Simulate_Phenotypes/SimpleGrid/E5/C1/h2-0/env-0.0/genos-gwas_common.phenos.txt")
df <- suppressMessages(inner_join(phenos, pops))
ggplot(data = df, aes(Lat, pheno_strat)) + geom_point() + stat_summary(fun.y=mean, colour="red", geom="line", size = 1)+ theme_bw() + ylab("Phenotype")
```

```{r}
ggplot(df, aes(x = Lat, y = Long, fill = pheno_strat)) + geom_tile() + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```

2 unit phenotypic difference across Latitude
```{r}
phenos <- fread("../output/Simulate_Phenotypes/SimpleGrid/E5/C1/h2-0/env-0.1/genos-gwas_common.phenos.txt")
df <- suppressMessages(inner_join(phenos, pops))
ggplot(data = df, aes(Lat, pheno_strat)) + geom_point() +stat_summary(fun.y=mean, colour="red", geom="line", size = 3) + theme_bw() + ylab("Phenotype")
```

```{r}
ggplot(df, aes(x = Lat, y = Long, fill = pheno_strat)) + geom_tile() + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```


# PGS in test panel 

No stratification
```{r}
pgs <- fread("../output/PRS/SimpleGrid/E5/C1/h2-0/env-0.0/PGS.txt")
df <- inner_join(pgs, pops) %>% gather(key = "Type", value = "PRS" , c, c.p, nc, c_Tm, c.p_Tm, nc_Tm)
df <- arrange(transform(df,Type=factor(Type,levels=c("c", "c.p", "nc", "c_Tm", "c.p_Tm", "nc_Tm"))),Type)
ggplot(data = df, aes(Lat, PRS)) + geom_point() + geom_smooth() + theme_bw()  + facet_wrap(~Type)
```

```{r}
df_pop <- df %>% group_by(Type, Pop) %>% mutate(avg_PRS = mean(PRS)) %>% ungroup()
ggplot(df_pop, aes(x = Lat, y = Long, fill = avg_PRS)) + geom_tile() + facet_wrap(~Type) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```

```{r}
Qx <- fread("../output/PGA_test/SimpleGrid/E5/C1/h2-0/env-0.0/Qx.txt")
Qx
```


Stratification
```{r}
pgs <- fread("../output/PRS/SimpleGrid/E5/C1/h2-0/env-0.1/PGS.txt")
df <- inner_join(pgs, pops) %>% gather(key = "Type", value = "PRS" , c, c.p, nc, c_Tm, c.p_Tm, nc_Tm)
df <- arrange(transform(df,Type=factor(Type,levels=c("c", "c.p", "nc", "c_Tm", "c.p_Tm", "nc_Tm"))),Type)
ggplot(data = df, aes(Lat, PRS)) + geom_point() + geom_smooth() + theme_bw()  + facet_wrap(~Type)
```

```{r}
df_pop <- df %>% group_by(Type, Pop) %>% mutate(avg_PRS = mean(PRS)) %>% ungroup()
df_pop <- arrange(transform(df_pop,Type=factor(Type,levels=c("c", "c.p", "nc", "c_Tm", "c.p_Tm", "nc_Tm"))),Type)
ggplot(df_pop, aes(x = Lat, y = Long, fill = avg_PRS)) + geom_tile() + facet_wrap(~Type) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")
```

```{r}
Qx <- fread("../output/PGA_test/SimpleGrid/E1/C1/h2-0/env-0.5/Qx.txt")
kable(Qx)
```

# 100 Simulations 

Look at number of false positives for $Q_x$ test using a p-value calculated from an empirical null (randomly flipping effect size signs).  

Clumped sites
```{r}
All <- fread("../code/plot/F100_SimpleGrid.txt")

nc <- All %>% filter(type == "nc"| type == "nc-Tm") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-") %>% filter(case == "C1")
nc$diff <- as.numeric(nc$diff)

ggplot(data = nc, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3) + xlab("Env Shift") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0) + geom_hline(yintercept = 0.05, color = "red")
```

Random Sites
```{r}
All <- fread("../code/plot/F100_SimpleGrid.txt")

c <- All %>% filter(type == "c"| type == "c-Tm") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-") %>% filter(case == "C1")
c$diff <- as.numeric(c$diff)

ggplot(data = c, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3) + xlab("Env Shift") + ylab("False Postive Rate") + theme_classic(base_size = 14) + scale_color_manual(values = c("goldenrod2", "navy"), labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(0,1) + labs(colour="GWAS") + theme(legend.text.align = 0) + geom_hline(yintercept = 0.05, color = "red")
```


# Looking at LD 

```{r}
read_genos <- function(geno_prefix, betas) {

  pvar <- NewPvar(paste0(geno_prefix, ".pvar"))
  d1 <- NewPgen(paste0(geno_prefix, ".pgen"))
  var.ids <- betas$ID
  var.indx <- rep(0, length(var.ids))
  for (i in 1:length(var.indx)) {
    var.indx[i] <- pgenlibr::GetVariantsById(pvar,var.ids[i])
  }
  X <- ReadList(d1,var.indx, meanimpute=F)

  return(X)
}
```

```{r}
betas_nc <- fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.nc.betas")
Mnc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-gwas_common", betas_nc)

betas_c <- fread("~/polygenic_adaptation_stratification/output/PRS/SimpleGrid/E1/C1/h2-0/env-0.0/genos-gwas_common.c.betas")
Mc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-gwas_common", betas_c)
```



```{r}
betas_c <- betas_c %>% mutate(ID2 = ID) %>% separate(ID, c("CHR", "POS", "A", "B"), "_") %>% filter(CHR == 1) %>% mutate(ID = ID2)
betas_c$POS <- as.numeric(as.character(betas_c$POS))
betas_c <- betas_c[order(POS),]
betas_nc <- betas_nc %>% mutate(ID2 = ID) %>% separate(ID, c("CHR", "POS", "A", "B"), "_") %>% filter(CHR == 1) %>% mutate(ID = ID2)
betas_nc$POS <- as.numeric(as.character(betas_nc$POS))
betas_nc <- betas_nc[order(POS),]

plot(betas_c$POS, betas_c$BETA_Strat)
points(betas_nc$POS, betas_nc$BETA_Strat, col = "red")
```


```{r}
test_c <- betas_c%>% 
  mutate(pairwise = POS - lag(POS), es_prod = BETA_Strat * lag(BETA_Strat))

test_nc <- betas_nc%>% 
  mutate(pairwise = POS - lag(POS), es_prod = BETA_Strat * lag(BETA_Strat))

plot(test_nc$pairwise, test_nc$es_prod)
points(test_c$pairwise, test_c$es_prod, col = "red")
```

```{r}
X_nc <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-test_common",  betas = betas_nc)

frq_nc <- colMeans(X_nc) /2
LD_nc <- rep(NA, length(frq_nc))
for (i in 1:(length(frq_nc) -1)) {
  LD_nc[i+1] <- cor(X_nc[,i], X_nc[,i+1]) 
}

X_c <- read_genos("/Users/jenniferblanc/polygenic_adaptation_stratification/output/Simulate_Genotypes/SimpleGrid/E1/C1/genos-test_common",  betas = betas_c)

frq_c <- colMeans(X_c) /2
LD_c <- rep(NA, length(frq_c))
for (i in 1:(length(frq_c) -1)) {
  LD_c[i+1] <- cor(X_c[,i], X_c[,i+1]) 
}

hist(LD_c)
hist(LD_nc)
```

```{r}
test_c$LD <- LD_c
test_nc$LD <- LD_nc

plot(test_c$LD, test_c$es_prod, xlim = c(-0.7,0.7), ylim = c(-5,5))
points(test_nc$LD, test_nc$es_prod, col = "red")
```


```{r}
qx_k <- function(k, frq, X, betas) {
  cov <- rep(NA, length(frq)-k)
  prod <- rep(NA, length(frq)-k)
  pos <- rep(NA, length(frq)-k)
  for (i in 1:(length(frq) -k)) {
    cov[i] <- cov(X[,i], X[,i+k]) 
    prod[i] <- betas$BETA_Strat[i] * betas$BETA_Strat[i+k]
    pos[i] <- betas$POS[i+k] - betas$POS[i]
  }
  return(c(sum(cov* prod), mean(pos)))
}

mat_nc <- matrix(NA, nrow = 10, ncol = 2)
for (i in 1:10) {
  mat_nc[i,] <- qx_k(i, frq_nc, X_nc, betas_nc)
}

mat_c <- matrix(NA, nrow = 10, ncol = 2)
for (i in 1:10) {
  mat_c[i,] <- qx_k(i, frq_c, X_c, betas_c)
}
```

```{r}
plot(mat_nc[,2], mat_nc[,1], col = "red",xlab = "pQx (numerator)", ylab = "avg. physical distance" )
points(mat_c[,2], mat_c[,1])
```

```{r}
cov_c <- rep(NA, length(frq_c))
for (i in 1:(length(frq_c) -1)) {
  cov_c[i+1] <- cov(X_c[,i], X_c[,i+1]) 
}

qx2 <- sum(cov_c[2:length(cov_c)] * test_c$es_prod[2:length(cov_c)])

cov_c <- rep(NA, length(frq_c)-2)
prod <- rep(NA, length(frq_c)-2)
for (i in 1:(length(frq_c) -2)) {
  cov_c[i] <- cov(X_c[,i], X_c[,i+2]) 
  prod[i] <- betas_c$BETA_Strat[i] * betas_c$BETA_Strat[i+2]
}
qx3 <- sum(cov_c * prod)

cov_c <- rep(NA, length(frq_c)-3)
prod <- rep(NA, length(frq_c)-3)
for (i in 1:(length(frq_c) -3)) {
  cov_c[i] <- cov(X_c[,i], X_c[,i+3]) 
  prod[i] <- betas_c$BETA_Strat[i] * betas_c$BETA_Strat[i+3]
}
qx4 <- sum(cov_c * prod)


cov_c <- rep(NA, length(frq_c)-4)
prod <- rep(NA, length(frq_c)-4)
for (i in 1:(length(frq_c) -4)) {
  cov_c[i] <- cov(X_c[,i], X_c[,i+4]) 
  prod[i] <- betas_c$BETA_Strat[i] * betas_c$BETA_Strat[i+4]
}
qx5 <- sum(cov_c * prod)
```

```{r}
plot(c(qx2, qx3, qx4, qx5))
```



## Concatenate cluster results 

```{r}
n <- 100
reps <- rep(NA, n)
for (i in 1:n){reps[i] <- paste0("T", i)}
print(reps)
h2 <- "h2-0"
envs <- c("env-0.0", "env-0.01", "env-0.02", "env-0.03", "env-0.04", "env-0.05", "env-0.06", "env-0.07", "env-0.08", "env-0.09", "env-0.1")
cases <- c("C1")
dat <- expand.grid(reps, cases, h2, envs)
colnames(dat) <- c("rep", "case", "h2", "env")

agg_all_data <- function(rep, dir_path, case, type, h2, env) {
  
  PGS <- fread(paste0(dir_path, rep,"/", case, "/", h2, "/", env, "/", "PGS.txt"))
  pop <- fread(paste0('../../output/Simulate_Genotypes/SimpleGrid/', rep,"/", "genos.pop"))
  colnames(pop) <- c("IID", "#FID", "POP", "LAT", "LONG" )
  
  df <- inner_join(PGS, pop)%>% select("nc", "nc_Tm", "POP", "LAT", "LONG")
  return(df)
}
df <- plyr::mdply(dat, agg_all_data, dir_path = '../../output/PRS/SimpleGrid/' )

```


```{r}
n <- 100
reps <- rep(NA, n)
for (i in 1:n){reps[i] <- paste0("T", i)}
print(reps)
h2 <- "h2-0"
envs <- c("env-0.0", "env-0.01", "env-0.02", "env-0.03", "env-0.04", "env-0.05", "env-0.06", "env-0.07", "env-0.08", "env-0.09", "env-0.1")
cases <- c("C1")
dat <- expand.grid(reps, cases, h2, envs)
colnames(dat) <- c("rep", "case", "h2", "env")

agg_all_data <- function(rep, dir_path, case, type, h2, env) {
  
  pheno <- fread(paste0(dir_path, rep,"/", case, "/", h2, "/", env, "/", "genos-gwas_common.phenos.txt"))
  pop <- fread(paste0('../../output/Simulate_Genotypes/SimpleGrid/', rep,"/", "genos.pop"))
  colnames(pop) <- c("IID", "FID", "POP", "LAT", "LONG" )
  
  df <- inner_join(pheno, pop)%>% select("IID", "pheno_strat", "POP", "LAT", "LONG")
  return(df)
}
df <- plyr::mdply(dat, agg_all_data, dir_path = '../../output/Simulate_Phenotypes/SimpleGrid/' )
```










```{r}
df <- fread("~/polygenic_adaptation_stratification/code/plot/Simple_Grid_T100_pgs.txt")
df2 <- df %>% group_by(rep, case, env) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP, env) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = LAT, LONG = LONG)

ggplot(df2, aes(x = LONG, y = LAT, fill = avg_PRS)) + geom_tile() + facet_wrap(~env) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white", limits = c(-1,1))

ggplot(df2, aes(x = LONG, y = LAT, fill = avg_PRS_Tm)) + geom_tile() + facet_wrap(~env) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white", limits = c(-1,1)) 
```


```{r}
df0 <- subset(df, df$env == "env-0.0" & df$rep == "T6") 
df0 <- df0%>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% group_by(LAT, LONG) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan), c = n())

ggplot(df0, aes(x = LONG, y = LAT, fill = avg_PRS)) + geom_tile()  + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")

```


```{r}
df_pheno <- fread("~/polygenic_adaptation_stratification/code/plot/Simple_Grid_T100_phenos.txt")
df_pheno <- subset(df_pheno, df_pheno$rep == "T6")
df_pheno <- df_pheno %>% group_by(POP) %>% summarise(avg_pheno = mean(pheno_strat))


ggplot(df_pheno, aes(x = LONG, y = LAT, fill = avg_pheno)) + geom_tile()  + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")

ggplot(df_pheno, aes(x = LAT, y = avg_pheno)) +geom_point()
```



```{r}
df <- fread("~/polygenic_adaptation_stratification/code/plot/SimpleGrid_cell_T100_pgs.txt")
df2 <- df %>% group_by(rep, case, env) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP, env) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = LAT, LONG = LONG)

#ind <- df %>% filter(rep == "T47")
#ggplot(ind, aes(x = LONG, y = LAT, fill = nc)) + geom_tile() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")

ggplot(df2, aes(x = LONG, y = LAT, fill = avg_PRS)) + geom_tile() + facet_wrap(~env) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white", limits = c(-1,1))

ggplot(df2, aes(x = LONG, y = LAT, fill = avg_PRS_Tm)) + geom_tile() + facet_wrap(~env) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white", limits = c(-1,1)) 
```

```{r}
df <- fread("~/polygenic_adaptation_stratification/code/plot/SimpleGrid_cell_T100.txt") 

nc <- df %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), avg_Ax = mean(Ax)) %>% separate(env, c("env", "diff"), "-")
nc$diff <- as.numeric(nc$diff)

nc
```


```{r}
phenos <- fread("~/polygenic_adaptation_stratification/code/plot/SimpleGrid_cell_T100_phenos.txt")
phenos <- phenos %>% group_by(POP) %>% summarise(avg_pheno = mean(pheno_strat), LAT = LAT, LONG = LONG)

ggplot(phenos, aes(x = LONG, y = LAT, fill = avg_pheno)) + geom_tile()  + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white")

```

