---
title: "Projection_plink2"
author: "Jennifer Blanc"
date: "3/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list=ls())
library(pgenlibr)
library(data.table)
```

## Projecting the Test vector (original) 

$$T_m = G X^T (XX^T)^{-1}T$$

```{r}
# Read GWAS Matrix 
pvar <- NewPvar("../output/Simulate_Genotypes/4PopSplit/E2/C1/genos-gwas_common.pvar")
d1 <- NewPgen("../output/Simulate_Genotypes/4PopSplit/E2/C1/genos-gwas_common.pgen")
G <- ReadList(d1,seq(1,20848), meanimpute=F)
#G <- scale(G)

# Read in Test Matrix 
pvar <- NewPvar("../output/Simulate_Genotypes/4PopSplit/E2/C1/genos-test_common.pvar")
d1 <- NewPgen("../output/Simulate_Genotypes/4PopSplit/E2/C1/genos-test_common.pgen")
X <- ReadList(d1,seq(1, 20848), meanimpute=F)
X <- scale(X)

# Read in Test Vector 
Tvec <- fread("../output/Calculate_Tm/4PopSplit/A2/C1/Tvec.txt")$V1
L <- ncol(X)
```


```{r}
mat_Tm <- matrix(NA, ncol = L, nrow = nrow(G))
for (i in 1:L) {
        mat_Tm[,i] <- G[,i] %*% t(X[,i]) %*% Tvec        
}

plot(Tm, mat_Tm[,100])
plot(Tm, rowSums(mat_Tm))
```

Direct multiplication 

$$T_m = G X^T (XX^T)^{-1}T$$

```{r}
Tm <- (G %*% t(X) %*% solve(X %*% t(X)) %*% Tvec) 
plot(Tm)
```


Use SVD decompostion of $X$ in R
```{r}
# Do SVD
s <- svd(X)
u <- s$u
d <- s$d
v <- s$v

# Square singular values and divide by L
svd_vals <- (d^2 / L)
n <- nrow(X)

# Calculate Tm
K = (M %*% t(X)) / (L)
Tm_svd = K %*% u[,1:(n-1)] %*% diag(1/svd_vals[1:(n-1)]) %*% t(u[,1:(n-1)]) %*% Tvec

plot(Tm_svd, Tm)
```

Check loadings
```{r}
t <- t(u) %*% u[,1]

plot(t(X) %*% u[,1] %*% (1/d[1]), v[,1])

plot(t(X) %*% u %*% diag(1/d) %*% t(u) %*% u[,1], v[,1])

plot(G %*% v[,1:199] %*% diag(1/d[1:199]) %*% t(u[,1:199]) %*% Tvec, Tm)
abline(a = 0, b= 1)
```


Calculate Tm via projecting each eigenvector individually and weighting by the Betas
```{r}
betas <- t(u) %*% Tvec

tmp <- M %*% t(X)%*% u[,1:(n-1)] %*% diag(1/svd_vals[1:(n-1)]) %*% t(u[,1:(n-1)]) %*% u[,1:(n-1)]
Tm_weighted <- tmp %*% betas[1:(n-1)] / L

plot(Tm, Tm_weighted)
abline(a = 0, b =1 , col = "red")
```

Calculate Tm via projecting each eigenvector individually using right singular values (loadings) and weighting by the Betas  
```{r}
tmp <- G %*% v[,1:(n-1)] %*% diag(1/d[1:(n-1)])
Tm_weighted_v <- tmp %*% betas[1:(n-1)] 

plot(Tm, Tm_weighted_v)
abline(a = 0, b =1 , col = "red")
```

Check that the $Tm$ from plink 2 matches 
```{r}
Tm_plink <- fread("../output/Calculate_Tm/4PopSplit/A1/C1/Tm.txt")
plot(Tm_plink$Tm, Tm)
abline(0,1, col = "red")
```


```{r}
TGX <- (1/(L-1)) * G %*% t(X) %*% Tvec
TGX <- scale(TGX, scale = F)
plot(TGX, Tm)

betas <- t(X) %*% Tvec
out <- G %*% betas 
plot(scale(out), TGX)
```

```{r}
TGGX <- G %*% pinv(t(G) %*% G) %*% t(X) %*% Tvec


plot(TGX)
```





```{r}
pvar <- NewPvar("../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.pvar")
d1 <- NewPgen("../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.pgen")
X <- ReadList(d1,seq(1,96278), meanimpute=F)
XA <- M[1:100,]
XB <- M[101:200,]
freq_diff <- (colSums(XA)/(2*100)) - (colSums(XB)/(2*100))

s <- svd(scale(X, scale = F))
u <- s$u
d <- s$d
v <- s$v
plot(t(X) %*% u[,1] %*% (1/d[1]), v[,1])

scale_loadings <- v[,1] * d[1]
plot(v[,1], freq_diff)

plot(colSums(XA)/(2*100), colSums(XB)/(2*100) )
```

```{r}
pop1 <- c(rep(2, 500), rep(1, 500))
pop1 <- matrix(rep(pop1,each=200),nrow=200)

pop2 <- c(rep(0, 500), rep(1, 500))
pop2 <- matrix(rep(pop2,each=200),nrow=200)

X <- rbind(pop1, pop2)

s <- svd(scale(X, scale = F))
u <- s$u
d <- s$d
v <- s$v
plot(t(X) %*% u[,1] %*% (1/d[1]), v[,1])
abline(a=0, b=1)

freq_diff <- (colSums(pop2)/(2*200)) - (colSums(pop1)/(2*200))
plot(freq_diff * 0.5, v[,1]^2)
abline(a=0, b=1)

# FST 
((((colSums(pop2)/(2*200)) - 0.5)^2 / 0.25) + (((colSums(pop1)/(2*200)) - 0.5)^2 / 0.25)) / 2
# Freq diff
freq_diff <- (colSums(pop2)/(2*200)) - (colSums(pop1)/(2*200))

```



Do PCA in plink (including loadings)
```{bash}
~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common \
       --pca biallelic-var-wts 199 \
       --out ../data/projection_example/S2_PC
```

Calculating Tm, creating pseudoinverse with plink vecs 
```{r}
plink_vecs <-fread("../data/projection_example/S2_PC.eigenvec")[,3:201]
plink_vecs <- apply(plink_vecs, 2, as.numeric)
plink_vals <- fread("../data/projection_example/S2_PC.eigenval")$V1

betas_plink <- t(plink_vecs) %*% Tvec
tmp <- M %*% t(X)%*% plink_vecs[,1:(n-1)] %*% diag(1/plink_vals[1:(n-1)]) %*% t(plink_vecs[,1:(n-1)]) %*% plink_vecs[,1:(n-1)]
Tm_plink <- tmp %*% betas_plink[1:(n-1)] / L

plot(Tm, Tm_plink)
abline(a = 0, b =1 , col = "red")
```

Calculate Tm, using loadings from plink
```{r}
plink_loadings <- fread("../data/projection_example/S2_PC.eigenvec.var")[,5:203]
plink_loadings <- apply(plink_loadings, 2, as.numeric)

tmp <- M %*% plink_loadings %*% diag(1/sqrt(plink_vals[1:(n-1)])) 
Tm_plink <- tmp %*% diag(1/sqrt(plink_vals)) %*% betas_plink[1:(n-1)] 

plot(Tm, Tm_plink)
abline(a = 0, b =1 , col = "red")
```

Calculate using projection step from plink2 
```{bash}
~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-test_common \
       --pca allele-wts 199 \
       --out ../data/projection_example/S2_PC_proj

~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S2/C1/genos-gwas_common \
       --score ../data/projection_example/S2_PC_proj.eigenvec.allele 2 5 header-read no-mean-imputation \
               variance-standardize \
       --score-col-nums 6-204 \
       --out ../data/projection_example/S2_projection
```

```{r}
betas_plink <- t(plink_vecs) %*% Tvec
proj_vecs_plink <- fread("../data/projection_example/S2_projection.sscore")[,5:203]
proj_vecs_plink <- apply(proj_vecs_plink, 2, as.numeric)

Tm_plink <- proj_vecs_plink %*% diag(1/sqrt(plink_vals)) %*% betas_plink[1:199] 

plot(Tm, -2 * Tm_plink)
abline(a = 0, b =1 , col = "red")
```

```{r}
allele <- fread("../data/projection_example/S2_PC_proj.eigenvec.allele")
allele_random <- allele %>% group_by(ID) %>% slice(1) %>%ungroup
allele_random <- allele_random[,6:204]
allele_random <- apply(allele_random, 2, as.numeric)
 
tmp <- M %*% allele_random %*% diag(1/sqrt(plink_vals[1:199]))
Tm_plink <- tmp %*% betas_plink[1:199] 
 
plot(Tvec, Tm_plink)
abline(a = 0, b =1 , col = "red")
```














Calculate Tm via plink2
```{bash}
#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
#       --pca allele-wts 199 \
#       --out ../data/projection_example/S1_PC
```

```{bash}
#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
#       --score ../data/projection_example/S1_PC.eigenvec.allele 2 5 header-read no-mean-imputation \
#               variance-standardize \
#       --score-col-nums 6-204 \
#       --out ../data/projection_example/S1_projection
```

```{bash}
#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common \
#       --score ../data/projection_example/S1_PC.eigenvec.allele 2 5 header-read no-mean-imputation \
#               variance-standardize \
#       --score-col-nums 6-204 \
#       --out ../data/projection_example/S1_projection
```

```{r}
#plink_vecs <- fread("../data/projection_example/S1_PC.eigenvec")[,3:201]
#plink_vecs <- apply(plink_vecs, 2, as.numeric)

#betas <- t(plink_vecs) %*% Tvec

#proj_vecs_plink <- fread("../data/projection_example/S1_projection.sscore")[,5:203]
#proj_vecs_plink <- apply(proj_vecs_plink, 2, as.numeric)
#plink_vals <- fread("../data/projection_example/S1_PC.eigenval")
#tst <- proj_vecs_plink %*%  diag(1/sqrt(plink_vals$V1))

#Tm_plink <- tst %*% betas[1:199] 

#plot(Tm, -1.8 * Tm_plink)
#abline(a = 0, b =1 , col = "red")
```

```{r}
# plink_loadings <- fread("../data/projection_example/S1_PC.eigenvec.var")[,5:203]
# plink_loadings <- apply(plink_loadings, 2, as.numeric)
# W <- scale(M) %*% plink_loadings 
# betas <- t(plink_vecs) %*% Tvec
# Tm_man <- W %*% betas[1:199] / (L)
# 
# plot(Tm_plink, Tm_man)
# abline(a = 0, b =1 , col = "red")
```














```{bash}
#~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
#       --score ../data/projection_example/S1_PC.eigenvec.allele 2 5 header-read no-mean-imputation \
#               variance-standardize \
#       --score-col-nums 6-204 \
#       --out ../data/projection_example/S1_projection_test
```

```{r}
# plink_vecs <- fread("../data/projection_example/S1_PC.eigenvec")[,3:201]
# plink_vecs <- apply(plink_vecs, 2, as.numeric)
# 
# betas <- t(plink_vecs) %*% Tvec
# 
# proj_vecs_plink <- fread("../data/projection_example/S1_projection_test.sscore")[,5:203]
# proj_vecs_plink <- apply(proj_vecs_plink, 2, as.numeric)
# plink_vals <- fread("../data/projection_example/S1_PC.eigenval")
# tst <- proj_vecs_plink %*%  diag(1/sqrt(plink_vals$V1))
# 
# Tm_plink_t <- tst %*% betas[1:199] 
# 
# plot(-2 *Tm_plink_t, Tvec)
# abline(a = 0, b =1 , col = "red")
```






























```{bash}
# Make GRM
#~/Desktop/gcta64 --bfile ../data/projection_example/genos-test_common --make-grm --out ../data/projection_example/GRM_S1

# Do PCA
#~/Desktop/gcta64 --grm ../data/projection_example/GRM_S1 --pca 199 --out ../data/projection_example/GRM_S1

# Get Loadings for n-1 PCs in test panel  
#~/Desktop/gcta64 --bfile ../data/projection_example/genos-test_common --pc-loading ../data/projection_example/GRM_S1 --out ../data/projection_example/GCTA_loadings
```

```{bash}
# Project n-1 PCs into GWAS panel
#~/Desktop/gcta64 --bfile ../data/projection_example/genos-test_common --project-loading  #../data/projection_example/GCTA_loadings 199 --out ../data/projection_example/projection_gcta
```

```{r}
# gcta_vecs <- fread("../data/projection_example/GRM_S1.eigenvec")[,3:201]
# gcta_vecs <- apply(gcta_vecs, 2, as.numeric)
# 
# betas <- t(gcta_vecs) %*% Tvec
# 
# proj_vecs <- fread("../data/projection_example/projection_gcta.proj.eigenvec")[,3:201]
# proj_vecs <- apply(proj_vecs, 2, as.numeric)
# 
# Tm_gcta <- proj_vecs %*% betas 
# 
# plot(Tvec, Tm_gcta)
# abline(a = 0, b =1 , col = "red")
```

```{r}
# plot(Tm, -2 * Tm_plink)
# abline(a = 0, b =1 , col = "red")
```

```{r}
# allele <- fread("../data/projection_example/S2_PC.eigenvec.allele")
# allele_random <- allele %>% group_by(ID) %>% slice(1) %>%ungroup
# allele_random <- allele_random[,6:204]
# allele_random <- apply(allele_random, 2, as.numeric)
# 
# tmp <- X %*% allele_random %*% diag(1/sqrt(plink_vals[1:199]))
# Tm_plink <- tmp %*% betas_plink[1:199] 
# 
# plot(Tvec, Tm_plink)
# abline(a = 0, b =1 , col = "red")
```


```{r}
deltaAB <- 1

E <- c(rep((-1 *deltaAB), 50), rep((1 *deltaAB), 50))
g <- c(rbinom(50, 2, 0.2), rbinom(50, 2, 0.8))
gs <- scale(g)
pA <- mean(g[1:50])/2
pB <- mean(g[51:100])/2
p <- mean(g)/2

cov(E, g)
pB - pA 

fB <- pB  / sd(g)
fA <- pA  / sd(g)

cov(E, gs)
fB - fA

(pA - pB) / var(g)


```


```{r}
p1 <- 0.5
p2 <- 0.5
pbar <- mean(c(p1, p2))

Fst1 <- ((pbar * (1-pbar)) - (0.5 * p1 * (1-p1)) - (0.5 * p2 * (1 -p2))) / (pbar * (1-pbar))
Fst1

Fst <- (0.5 * (p1 + p2)^2) / (pbar * (1 - pbar))
Fst
```


### Test projection

```{r}
# Read in GWAS matrix
pvar <- NewPvar("../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common.pvar")
d1 <- NewPgen("../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common.pgen")
G <- ReadList(d1,seq(1,932), meanimpute=F)
G <- scale(G, scale = F)

# Read in Test Matrix 
pvar <- NewPvar("../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.pvar")
d1 <- NewPgen("../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common.pgen")
X <- ReadList(d1,seq(1, 932), meanimpute=F)
X <- scale(X, scale = F)
Xs <- scale(X)

# Read in Test Vector 
Tvec <- fread("../output/Calculate_Tm/4PopSplit/S1/C1/Tvec.txt")$Tvec
L <- ncol(X)
```

Compute $(XX^\top)^{-1}T$ using solve
```{r}
## Unscaled 
FXX <-(1 / (L)) * X %*% t(X)
FXXinv <- solve(FXX)

## Scaled
FXXs <- (1 / (L)) * Xs %*% t(Xs)
FXXinvs <- solve(FXXs)

Ts <- FXXinv %*% Tvec
Tss <- FXXinvs %*% Tvec

plot(Ts, Tss, xlab="Raw",ylab= "Scaled" )
abline(0,1, col = "red")
```
Compute $(XX^\top)^{-1}T$ using eigen
```{r}
myE <- eigen(FXX)
myEs <- eigen(FXXs)
n <- nrow(X)
n1 <- n-1
FXXinv_pca <- myE$vectors[,1:n1] %*% diag(1/myE$values[1:n1]) %*% t(myE$vectors[,1:n1] )
FXXinv_pcas <- myEs$vectors[,1:n1] %*% diag(1/myEs$values[1:n1]) %*% t(myEs$vectors[,1:n1] )

Ts_pca <- FXXinv_pca %*% Tvec
Ts_pcas <- FXXinv_pcas %*% Tvec

plot(Ts_pca, Ts_pcas, xlab="Raw",ylab= "Scaled" )
abline(0,1, col = "red")
```
Compare eigen to solve 
```{r}
plot(Ts, Ts_pca, xlab="Solve",ylab= "PCA" , main = "Raw")
abline(0,1, col = "red")
plot(Tss, Ts_pcas, xlab="Solve",ylab= "PCA" , main = "Scaled")
abline(0,1, col = "red")
```
Compare to plink PCA  
```{bash}
~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
       --pca 199 \
       --out ~/Desktop/test_pca
```

```{r}
plink_vecs <- fread("~/Desktop/test_pca.eigenvec")
plink_vecs <- apply(plink_vecs[,3:ncol(plink_vecs)], 2, as.numeric)
plink_vals <- fread("~/Desktop/test_pca.eigenval")

ts_plink <- plink_vecs %*% diag(1/plink_vals$V1) %*% t(plink_vecs) %*% Tvec

plot(Ts_pca,ts_plink, main = "Raw",xlab = "Eigen", ylab= "Plink")
abline(0,1, col = "red")
plot(Ts_pcas,ts_plink, main = "Scaled", xlab = "Eigen",ylab = "Plink")
abline(0,1, col = "red")
```

Compute $(XX^T)$ in plink 
```{bash}
~/Desktop/plink2 --pfile ../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common \
       --make-rel cov square \
       --out ~/Desktop/test_pca
```

```{r}
cov_mat <- fread("~/Desktop/test_pca.rel")
cov_mat <- apply(cov_mat, 2, as.numeric)
myE_cov <-eigen(cov_mat)


ts_cov_pca <- myE_cov$vectors[,1:n1] %*% diag(1/myE_cov$values[1:n1]) %*% t(myE_cov$vectors[,1:n1]) %*% Tvec
#ts_cov_solve <- solve(cov_mat) %*% Tvec


#plot(Ts, ts_cov_solve, xlab = "Solve R Cov", ylab = "Solve Plink Cov")
#abline(0,1, col = "red")
plot(Ts_pca,ts_cov_pca, xlab = "Eigen R Cov", ylab = "Eigen Plink Cov")
abline(0,1, col = "red")
```

Compute $r = \tilde{X}^\top(XX^\top)^{-1}T$ using eigen
```{r}
r <- t(X) %*% ts_cov_pca
```

Compute r using plink 
```{r}
ids <- fread("~/Desktop/test_pca.rel.id")
ids$Tvec <- Tvec
ids$InTvec <- ts_cov_pca
fwrite(ids, "~/Desktop/InvTvec.txt", quote = F, col.names = T, row.names = F, sep ="\t")

outfile_XT <- paste0("~/Desktop/", "xt_temp")
cmd_XT <- paste("sh ../code/Calculate_Tm/compute_XTinv.sh", "../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-test_common", "~/Desktop/InvTvec.txt", outfile_XT, sep = " ")
system(cmd_XT)
```

```{r}
outpath <- "~/Desktop/"
# Read in betas and genotype counts
beta_plink <- fread(paste0(outpath, "xt_temp.InTvec.glm.linear"))
count_plink <- fread(paste0(outpath, "xt_temp.gcount"))

# Calculate length of mean centered genotypes from counts
nOBS <- (count_plink$HOM_REF_CT + count_plink$HET_REF_ALT_CTS + count_plink$TWO_ALT_GENO_CTS)
counts <- (count_plink$HOM_REF_CT * 0) + (count_plink$HET_REF_ALT_CTS * 1) + (count_plink$TWO_ALT_GENO_CTS * 2)
mean_gc <- counts / nOBS
length_mc_genos <- (count_plink$HOM_REF_CT * (-1 * mean_gc)^2) + (count_plink$HET_REF_ALT_CTS * (1 - mean_gc)^2) +  (count_plink$TWO_ALT_GENO_CTS * (2 - mean_gc)^2)

# Fix betas
betas_plink_norm <- beta_plink$BETA * length_mc_genos

# Compute r using plink 
r_plink <- betas_plink_norm

plot(r,r_plink)
abline(0,1, col ="red")
```

Compute $D^{-1}r$
```{r}
colVars <- apply(G, 2, var)
D <- 1/sqrt(colVars)
Dr <- D * r
```

Compute $D^{-1}r$ using plink
```{r}
outfile_count <- paste0("~/Desktop/", "G_count")
cmd_count <- paste("sh ../code/Calculate_Tm/compute_GWAScount.sh", "../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common", outfile_count, sep = " ")
system(cmd_count)

count_plink <- fread(paste0(outpath, "G_count.gcount"))

# Calculate length of mean centered genotypes from counts
nOBS <- (count_plink$HOM_REF_CT + count_plink$HET_REF_ALT_CTS + count_plink$TWO_ALT_GENO_CTS)
counts <- (count_plink$HOM_REF_CT * 0) + (count_plink$HET_REF_ALT_CTS * 1) + (count_plink$TWO_ALT_GENO_CTS * 2)
mean_gc <- counts / nOBS
length_mc_genos <- (count_plink$HOM_REF_CT * (-1 * mean_gc)^2) + (count_plink$HET_REF_ALT_CTS * (1 - mean_gc)^2) +  (count_plink$TWO_ALT_GENO_CTS * (2 - mean_gc)^2)
length_mc_genos <- length_mc_genos * (1/(m-1))

Dr_plink <- (1/sqrt(length_mc_genos)) * r_plink
plot(Dr_plink, Dr)
abline(0,1, col = "red")
```
Compute TGWAS $GD^{-1}r$
```{r}
TGWAS <- scale(G) %*% Dr
```

Compute TGWAS in plink 
```{r}
#  Re-write .linear file with correct betas
beta_plink$BETA <- (1/sqrt(length_mc_genos)) * Dr_plink
#beta_plink$BETA <- Dr_plink
#beta_plink$BETA <- rep(1,nrow(beta_plink))
beta_reformat <- beta_plink %>% dplyr::select(ID, A1, BETA)
fwrite(beta_reformat, paste0(outpath, "xt_scaled.InTvec.glm.linear"), sep = "\t")

# Compute b
outfile_b <- paste0(outpath, "b")
cmd_b <- paste("sh ../code/Calculate_Tm/GWAS_score.sh", "../output/Simulate_Genotypes/4PopSplit/S1/C1/genos-gwas_common", paste0(outpath, "xt_scaled.InTvec.glm.linear"), outfile_b, sep = " ")
system(cmd_b)

# Read in and return b
b = fread(paste0(outpath, "b.sscore"))
TGWAS_plink = as.matrix(b$BETA_SUM)


plot(TGWAS,TGWAS_plink)
abline(0,1, col="red")

plot(scale(TGWAS),scale(TGWAS_plink))

plot(b, TGWAS)
abline(0,1, col="red")
```


```{r}
colVars <- apply(G, 2, var)

sqrt(colVars[1])

var(G[,1]/sqrt(colVars[1]))

t<- G[,1]/sqrt(colVars[1])
t2 <- scale(G[,1])
t3 <- G[,1]/sqrt(length_mc_genos[1]) 
```


