---
title: "Simulations_Methods"
author: "Jennifer Blanc"
date: "7/23/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
library(data.table)
library(tidyverse)
library(latex2exp)
library(ggpubr)
```

## Introduction 

Here I will reccord the details for the simulations presented in the main text and the supplement. I will also include the code to make the the figures. The illustrative examples for the 3 simulation types can be found **here** (link later).  

## 4PopSplit 

### Msprime simulation 

Generate diplois individuals according to the demographic model below 

* T2 = 11000  
* T1 (deeper split) = 22000 
* Ne (ancestral = A = B = C = D)  = 10000
* sample size = 1000 diploids per population  
* chromosome length = 100000  
* 200 chromosomes  
* $\mu$ = 1e-08  
* $\rho$ = 1e-08  

The simulation was run 100 times total. 

### Create genotypes panels and filter SNPs

1. Set up 2 parallel sets of 2  
  1. C1: GWAS panel has the 1000 individuals from populations A and C (2000 total), test panel has the 1000 inidividuals from B and D (2000 total)
  2. C2: GWAS panel has the 1000 individuals from populations A and B (2000 total), test panel has the 1000 inidividuals from C and D (2000 total)
2. Filter snps to sites segregating at greater than 5% frequency within each panel  

### Draw phenotype 

Notes:  

* All simulations have $h^2 = 0$  

1. Sample random phenotype $N(0,1)$ for all GWAS indviduals 
2. Scale variance to match realistic GWAS (see __ for more details)  
3. For all individuals in population A shift the mean phenotype by a variable amount 

### Project Test Vector  

1. Set test vector as a standardized, mean-centered $n \times 1$ vector with each entry as the test panel individuals population ID (1 vs 0) encoding   
2. Project test vector in GWAS panel, resulting in an $m \times 1$ (see ___ for more detail)

### Run GWASs  

1. Use plink to run GWAS with and without including projected test vector 

### Calculate PGS and Qx 

1. Pick the SNP with lowest p-value per chromosome to include in PGS (200 SNPs total)  
2. Compute PGS in test panel individuals 
3. Calculate $V_a$ 
4. Calculate the amount of variance explained by the test vector ($\lambda_T$; note that we used a test vector with $1/2*n_{pop}$,$ -1/2 *n_{pop}$ encoding)  
5. Calculate empirical null p-value by permuting the signs of the SNP effect sizes 1,000 times and computing the proportion of permutations greater than the true $Q_x$ value.  

### Figures  

```{r,eval = TRUE, echo=FALSE, warning=FALSE}
plot_data <- function(data, color_name) {
  
  pl <- ggplot(data = data, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+   xlab(TeX("$\\Delta_{AB}$")) + ylab("False Postive Rate") + theme_classic(base_size = 12) + scale_color_manual(values = c(color_name)) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(-0.1,1) + labs(colour="") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme( axis.title = element_text(size = 12),legend.position = "none") + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width = 0.005)
  
  return(pl)
}
```

```{r,eval = TRUE, echo=FALSE, warning=FALSE}
df <- fread("../code/plot/4PopSplit_F100_fullgrid.txt")

# C1 - no correction
nc <- df %>% filter(type == "nc") %>% filter(case == "C1") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-")  
nc$diff <- as.numeric(nc$diff)
pl1 <- plot_data(nc, "goldenrod2")

# C1 - correction 
nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C1") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-")
nc$diff <- as.numeric(nc$diff)
pl2 <- plot_data(nc, "navy")

# C2 - No correction 
nc <- df %>% filter(type == "nc") %>% filter(case == "C2") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-")
nc$diff <- as.numeric(nc$diff)
pl3 <- plot_data(nc, "goldenrod2")

# C2 - correction 
nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C2") %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-")
nc$diff <- as.numeric(nc$diff)
pl4 <- plot_data(nc, "navy")

ggarrange(pl1, pl2, pl3, pl4)
```


## Simple Grid 

### Msprime simulation 

Generate diploid individuals using a stepping stone model on a 6x6 grid with the parameters below: 

Parameters: 

* 200 chromosomes  
* sample size = 80 diploids  
* Ne = 1000 
* $\mu$ = 1e-07  
* $\rho$ = 1e-07 
* m = 0.01  
* chromosome length = 100000 
* tmove = -9 (perpetual structure)  

The simulation was run 100 times total.  

### Create genotypes panels and filter SNPs

1. Sample 20 individuals per deme for test panels (36 * 20 individuals total) 
2. Remaining individuals (60 per deme) become the GWAS panel  
3. Restrict snps to sites segregating at greater than 5% frequency within each panel  

### Draw phenotype 

Notes:  

* All simulations have $h^2 = 0$  
* ENV is the magnitude of the phenotype shift 

1. Sample random phenotype $N(0,1)$ for all GWAS indviduals 
2. Scale variance within deme to match full size GWAS variance (see __ for more details)
3. Simulate stratification:  
  1. Lat/Lat: Shift average phenotype per Latitude (if env is the total shift, shift 1 lat by env/5, 2 lat by 2*(env/5), etc.) and use Latitude as the test vector 
  2. stratD/Lat: Shift the average phenotype on the diagonals of the grid going lower left to upper right and use latitude as the test vector 
  
### Project Test Vector  

1. Set test vector as a mean centered $n \times 1$ vector with each entry as the test panel individuals latitude/longitude/popID  
2. Project test vector in GWAS panel, resulting in an $m \times 1$ (see ___ for more detail)

### Run GWASs  

1. Use plink to run GWAS with and without including projected test vector 

### Calculate PGS and Qx 

1. Pick the SNP with lowest p-value per chromosome to include in PGS (200 SNPs total)  
2. Compute PGS in test panel individuals 
3. Calculate $V_a$ 
4. Calculate the amount of variance explained by the test vector ($\lambda_T$; note that we used the mean-centered test vector to compute this)  
5. Calculate empirical null p-value by permuting the signs of the SNP effect sizes 1,000 times and computing the proportion of permutations greater than the true $Q_x$ value.  


### Figures

```{r, eval = TRUE, echo=FALSE, warning=FALSE}
make_phenotype_plot <- function(df) {
  pl <- ggplot(df, aes(x = LONG, y = LAT, fill = value)) + geom_tile() + theme_classic(base_size = 12) +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1), legend.position = "none") + scale_fill_gradient2(low = "palegoldenrod", high = "purple4") + xlab("Longitude") + ylab("Latitude") + scale_y_continuous(breaks=seq(0,6,1)) + scale_x_continuous(breaks=seq(0,6,1)) 
  return(pl)
}

make_pgs_plot <- function(df, ll, up) {
  
  plu <- df %>% filter(variable == "avg_PRS") %>% ggplot(aes(x = LONG, y = LAT, fill = value)) + geom_tile() +  theme_classic(base_size = 12) +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "brown4", high = "royalblue4", limits = c(ll, up), name = "PGS") + xlab("Longitude") + ylab("Latitude") + scale_y_continuous(breaks=seq(0,6,1)) + scale_x_continuous(breaks=seq(0,6,1)) 
  
  plc <- df %>% filter(variable == "avg_PRS_Tm") %>% ggplot(aes(x = LONG, y = LAT, fill = value)) + geom_tile() +  theme_classic(base_size = 12) +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "brown4", high = "royalblue4", limits = c(ll, up), name = "PGS") + xlab("Longitude") + ylab("Latitude") + scale_y_continuous(breaks=seq(0,6,1)) + scale_x_continuous(breaks=seq(0,6,1)) 
  
  pl <- ggarrange(plu, plc, common.legend = T, legend="right")
  
  return(pl)
}

make_fp_plot <- function(df) {
  
  pl <- ggplot(data = df, aes(x = type, y = fp_strat, fill = type)) + geom_col()+ geom_point(fill = "black", size = 3)+ xlab("") + ylab("False Postive Rate") + theme_classic(base_size = 12) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.text.x = element_text(size = 12, color = "black",  angle = 35, hjust=1), legend.position = "none") + ylim(-0.01,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red")+ geom_errorbar(aes(ymin = lowerci, ymax = upperci), width =0.25) + scale_x_discrete(breaks=c("nc","nc-Tm"),labels = unname(TeX(c("No covariate", " Including $T^{GWAS}")))) + scale_fill_manual(values = c("goldenrod2", "navy")) + facet_wrap(~test)
  
  return(pl)
}
```


#### Lat/Lat 

* Simulated phenotype according to latitudinal gradient 
* Test vector is latitude 

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
# Phenotype plot
df <- fread("../code/plot/SimpleGrid_scale_phenos.txt") %>% filter(env == "env-0.05") %>% group_by(rep, case, env) %>% mutate(pheno_strat = (pheno_strat - mean(pheno_strat)) / sd(pheno_strat)) %>% ungroup() %>%  group_by(POP, env) %>% summarise(avg_pheno = mean(pheno_strat),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,3:5] %>%  gather(variable, value, -c(LAT, LONG))
pl_phenotype <- make_phenotype_plot(df)

# PGS plot
df <- fread("../code/plot/SimpleGrid_C1_scale_T100_pgs.txt") %>% filter(env == "env-0.05") %>% group_by(rep, case, env) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP, env) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,2:6] %>%  gather(variable, value, -c(LAT, LONG,env))
pl_pgs <- make_pgs_plot(df, -0.5, 0.5)

# False positive rate plot 
nc <- fread("../code/plot/SimpleGrid_T100_rescale.txt") %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1" & env == "env-0.05")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-") %>% mutate(test = "Latitude")
pl_fp <- make_fp_plot(nc)

ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.25, 0.5, 0.25))
```


#### stratD/Lat 

* Simulated phenotype according to gradient along the diagonal going from lower left to upper right  
* Project latitude as the test vector 
* For each simulation do 2 polygenic adaptation tests one with latitude as the test vector and one with longitude  

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
# Phenotype plot
df <- fread("../code/plot/SimpleGrid_stratD_phenos.txt") %>% filter(env == "env-0.3") %>% group_by(rep, case, env) %>% mutate(pheno_strat = (pheno_strat - mean(pheno_strat)) / sd(pheno_strat)) %>% ungroup() %>%  group_by(POP, env) %>% summarise(avg_pheno = mean(pheno_strat),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,3:5] %>%  gather(variable, value, -c(LAT, LONG))
pl_phenotype <- make_phenotype_plot(df)

# PGS plot
df <- fread("../code/plot/SimpleGrid_C1_sratD_T100_pgs.txt") %>% filter(env == "env-0.3") %>% group_by(rep, case, env) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP, env) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,2:6] %>%  gather(variable, value, -c(LAT, LONG,env))
pl_pgs <- make_pgs_plot(df, -0.5, 1)

# False positive rate plot 
nc_lat <- fread("../code/plot/SimpleGrid_T100_stratD_Lat.txt") %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1" & env == "env-0.3")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()))  %>% separate(env, c("env", "diff"), "-") %>% mutate(test = "Latitude")
nc_lat$diff <- as.numeric(nc_lat$diff)
nc_long <- fread("../code/plot/SimpleGrid_T100_stratD_Long.txt") %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1" & env == "env-0.3")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-") %>% mutate(test = "Longitude")
nc_long$diff <- as.numeric(nc_long$diff)
df <- rbind(nc_lat, nc_long)
pl_fp <- make_fp_plot(df)

ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.25, 0.5, 0.25))
```

#### singleDeme/Lat 

* Simulated phenotype difference in a single deme
* Project latitude as the test vector 
* For each simulation do 1 polygenic adaptation test with latitude as the test vector 

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
# Phenotype plot
df <- fread("../code/plot/SimpleGrid_ps-1.0_phenos.txt") %>% filter(env == "ps-1.0") %>% group_by(rep, case, env) %>% mutate(pheno_strat = (pheno_strat - mean(pheno_strat)) / sd(pheno_strat)) %>% ungroup() %>%  group_by(POP, env) %>% summarise(avg_pheno = mean(pheno_strat),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,3:5] %>%  gather(variable, value, -c(LAT, LONG))
pl_phenotype <- make_phenotype_plot(df)

# PGS plot
df <- fread("../code/plot/SimpleGrid_ps-1.0_pgs.txt") %>% filter(env == "ps-1.0") %>% group_by(rep, case, env) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP, env) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,2:6] %>%  gather(variable, value, -c(LAT, LONG,env))
pl_pgs <- make_pgs_plot(df, -0.7, 1.5)

# False positive rate plot 
nc <- fread("../code/plot/SimpleGrid_T100_ps.txt") %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C1" & env == "ps-1.0")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-")%>% mutate(test = "Latitude")
nc$diff <- as.numeric(nc$diff) 
pl_fp <- make_fp_plot(nc)

ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.25, 0.5, 0.25))
```

#### singleDeme/singleDeme 

* Simulated phenotype difference in a single deme
* Project single deme vs everyone else as test vector 
* For each simulation do 1 polygenic adaptation test with single deme vs everybody as the test vector 

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
# Phenotype plot
df <- fread("../code/plot/SimpleGrid_1Pop_phenos.txt") %>% filter(env == "env-0.5") %>% group_by(rep, case, env) %>% mutate(pheno_strat = (pheno_strat - mean(pheno_strat)) / sd(pheno_strat)) %>% ungroup() %>%  group_by(POP, env) %>% summarise(avg_pheno = mean(pheno_strat),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,3:5] %>%  gather(variable, value, -c(LAT, LONG))
pl_phenotype <- make_phenotype_plot(df)

# PGS plot
df <- fread("../code/plot/SimpleGrid_1Pop_pgs.txt") %>% filter(env == "env-0.5") %>% group_by(rep, case, env) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP, env) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = mean(LAT), LONG = mean(LONG))
df <- df[,2:6] %>%  gather(variable, value, -c(LAT, LONG,env))
pl_pgs <- make_pgs_plot(df, -0.7, 1.5)

# False positive rate plot 
nc <- fread("../code/plot/SimpleGrid_1Pop_ps.txt") %>% filter(type == "nc"| type == "nc-Tm") %>% filter(case == "C3" & env == "env-0.5")  %>% group_by(env, type, case) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(env, c("env", "diff"), "-")%>% mutate(test = "Single Deme")
nc$diff <- as.numeric(nc$diff) 
pl_fp <- make_fp_plot(nc)

ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.25, 0.5, 0.25))
```

## True Signal 

### Msprime simulation 

Generate diplois individuals according to the demographic model below 

* T2 = 2200  
* T1 (deeper split) = 4400 
* Ne (ancestral = A = B = C = D)  = 10000
* sample size A, C = 10000 diploids per population  
* sample size B, D = 1000 diploids per population  
* chromosome length = 100000  
* 200 chromosomes  
* $\mu$ = 1e-08  
* $\rho$ = 1e-08  

The simulation was run 100 times total. 

### Create genotypes panels and filter SNPs

GWAS panel has the 1000 individuals from populations A and C (20000 total), test panel has the 1000 inidividuals from B and D (2000 total)


### Draw phenotype 

Notes:  

* All simulations have $h^2 = 0.3$  

** TBD ** 

### Project Test Vector  

1. Set test vector as a standardized, mean-centered $n \times 1$ vector with each entry as the test panel individuals population ID (1 vs 0) encoding   
2. Project test vector in GWAS panel, resulting in an $m \times 1$ (see ___ for more detail)

### Run GWASs  

1. Use plink to run GWAS with and without including projected test vector 
2. Re-run the GWAS using population ID (A vs C) as a covariate, as all the confounding is between populations this is how we would control for stratification given we know the distribution of the confounder.   

### Calculate PGS and Qx 

For each replicate calculate 7 total $Q_X$ statistics 

1. True $Q_X$ computed using the true causal effects.  
2. $Q_X$ computed using the estimated effect sizes at causal sites.  
  a. No covariate included  
  b. Include $T^{GWAS}$  
  c. Include population ID  
3. $Q_X$ computed using the estimated effect sizes at ascertained sites (picking lowest p-value per chromosome)   
  a. No covariate included  
  b. Include $T^{GWAS}$  
  c. Include population ID  


1. Pick the SNP with lowest p-value per chromosome to include in PGS (200 SNPs total)  
2. Compute PGS in test panel individuals 
3. Calculate $V_a$ 
4. Calculate the amount of variance explained by the test vector ($\lambda_T$; note that we used a test vector with $1/2*n_{pop}$,$ -1/2 *n_{pop}$ encoding)  
5. Calculate empirical null p-value by permuting the signs of the SNP effect sizes 1,000 times and computing the proportion of permutations greater than the true $Q_x$ value.  

### Figures  

```{r, eval = TRUE, echo=FALSE, warning=FALSE}
df <- fread("../code/plot/4PopSplit_L100_true_same.txt")
nc <- df %>%filter(env == "env-0.0" | env == "env-1.0") %>% filter(type == "nc"| type == "nc-Tm" | type == "c" | type == "c-Tm" | type == "c-ID" | type == "nc-ID") %>% group_by(env, type, case, ts) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(ts, c("ts", "prob"), "-")
nc$prob <- as.numeric(nc$prob)
nc <- nc %>% mutate(causal = case_when((type == "c" | type == "c-Tm" | type == "c-ID") ~ "causal", (type == "nc" | type == "nc-Tm" | type == "nc-ID") ~ "ascertained")) %>% mutate(covariate_included = case_when((type == "c" | type == "nc") ~ "No", (type == "c-Tm" | type == "nc-Tm") ~ "Tm", (type == "c-ID" | type == "nc-ID") ~ "ID"))
nc_same <- nc %>% filter(prob < 1)
nc_same <- nc_same %>% mutate(direction = case_when(env == "env-0.0" ~ "no environmental stratification", env == "env-1.0" ~ "same direction"))

df <- fread("../code/plot/4PopSplit_L100_true_diff.txt")
nc <- df %>%filter(env == "env-0.0" | env == "env-1.0") %>% filter(type == "nc"| type == "nc-Tm" | type == "c" | type == "c-Tm"| type == "c-ID" | type == "nc-ID") %>% group_by(env, type, case, ts) %>% summarise(fp_strat = sum(`P-EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(ts, c("ts", "prob"), "-")
nc$prob <- as.numeric(nc$prob)
nc <- nc %>% mutate(causal = case_when((type == "c" | type == "c-Tm" | type == "c-ID") ~ "causal", (type == "nc" | type == "nc-Tm" | type == "nc-ID") ~ "ascertained")) %>% mutate(covariate_included = case_when((type == "c" | type == "nc") ~ "No", (type == "c-Tm" | type == "nc-Tm") ~ "Tm", (type == "c-ID" | type == "nc-ID") ~ "ID"))
nc_diff <- nc %>% filter(prob < 1)
nc_diff$direction <- "opposite direction"

df <- fread("../code/plot/4PopSplit_L100_tes.txt")
nc <- df %>%filter(env == "env-0.0" | env == "env-1.0")%>% group_by(env, h2, case, ts) %>% summarise(tp = sum(`P-EN` < 0.05)/ n()) %>% separate(ts, c("ts", "prob"), "-")  %>% filter(tp > 0) %>% mutate(direction = case_when(env == "env-0.0" ~ "no environmental stratification", (h2 == "same-0.3" & env == "env-1.0") ~ "same direction", (h2 == "diff-0.3" & env == "env-1.0") ~ "opposite direction")) 
nc <- nc[,2:ncol(nc)]
nc$prob <- as.numeric(nc$prob)

data <- rbind(nc_same, nc_diff)
data <- full_join(data, nc)


ggplot(data, aes(x = prob, y = fp_strat, color = covariate_included)) + geom_point() + facet_grid(causal ~ direction) + geom_point(aes(y = tp), color = "red", shape = 8) + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width = 0.02) + theme_classic(base_size = 12)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 12),legend.position = "bottom") + ylim(0,1) + geom_hline(yintercept = 0.05, color = "red") + scale_color_manual(values = c("darkgreen", "goldenrod", "navy"), label = c("Population ID", "Uncorrected", "T GWAS"), name = "Covariate Included") + xlab("True Signal") + ylab("Positive Rate") 


```




```{r,eval = FALSE, echo=FALSE}
A <- rnorm(10000, 0, 1)
C <- rnorm(10000, 3, 1)

df <- cbind(A,C)
df <- melt(df)
ggplot(df, aes(x = value, fill = Var2)) + geom_density(alpha=0.5) + theme_classic(base_size = 14) + xlab("Phenotype")+ ylab("") + scale_fill_manual(values = c("darkred", "goldenrod")) + theme(axis.ticks = element_blank(), axis.text = element_blank(), axis.line.y=element_blank(), , axis.title.x = element_text(size = 25)) + geom_vline(xintercept = 0, color = "darkred", size = 1.5)+ geom_vline(xintercept = 3, color = "goldenrod", size = 1.5)+ scale_y_continuous(expand = c(0, 0)) + theme(legend.position = "none")+ scale_y_continuous(expand = c(0, 0))
ggsave("~/Desktop/gwas.png", height = 5, width = 7)
```

```{r,eval = FALSE, echo=FALSE}
D <- rnorm(10000, 3, 1)
C <- rnorm(10000, 0, 1)

df <- cbind(C,D)
df <- melt(df)
ggplot(df, aes(x = value, fill = Var2)) + geom_density(alpha=0.5) + theme_classic(base_size = 14) + xlab("Polygenic Score")+ ylab("") + scale_fill_manual(values = c("darkblue","forestgreen")) + theme(axis.ticks = element_blank(), axis.text = element_blank(), axis.line.y=element_blank(), axis.title.x = element_text(size = 25)) + geom_vline(xintercept = 0, color = "darkblue", size = 1.5)+ geom_vline(xintercept = 3, color = "forestgreen", size = 1.5)  + scale_y_continuous(expand = c(0, 0)) + theme(legend.position = "none")
ggsave("~/Desktop/pgs_sep.png", height = 5, width = 7)

#ggplot(df, aes(x =Var2,y=value, color = Var2)) + geom_jitter() + theme_classic(base_size = 16) + scale_color_manual(values = c("darkblue","forestgreen")) + xlab("Population") + ylab("Polygenic Score") + stat_summary(fun.y=mean, colour="black", geom="line", aes(group = 1), size = 3) + stat_summary(fun.y=mean, colour="black", geom="point", aes(group = 1), size = 3)
```

```{r,eval = FALSE, echo=FALSE}
D <- rnorm(10000, 3, 1)
C <- rnorm(10000, 0, 1)

df <- cbind(C,D)
df <- melt(df)
ggplot(df, aes(x = value, fill = Var2)) + geom_density(alpha=0.5) + theme_classic(base_size = 14) + xlab("Polygenic Score")+ ylab("") + scale_fill_manual(values = c("darkblue","forestgreen")) + theme(axis.ticks = element_blank(), axis.text = element_blank(), axis.line.y=element_blank(), axis.title.x = element_text(size = 25)) + geom_vline(xintercept = 0, color = "darkblue", size = 1.5)+ geom_vline(xintercept = 3, color = "forestgreen", size = 1.5)  + scale_y_continuous(expand = c(0, 0)) + theme(legend.position = "none")
#ggsave("~/Desktop/pgs_sep.png", height = 5, width = 7)
```


```{r,eval = FALSE, echo=FALSE}
D <- rnorm(10000, 3, 1)
C <- rnorm(10000, 0, 1)
E <- rnorm(10000, 7, 1)

df <- cbind(C,D,E)
df <- melt(df)
ggplot(df, aes(x = value, fill = Var2)) + geom_density(alpha=0.5) + theme_classic(base_size = 14) + xlab("Polygenic Score")+ ylab("") + scale_fill_manual(values = c("darkblue","forestgreen", "darkseagreen1")) + theme(axis.ticks = element_blank(), axis.text = element_blank(), axis.line.y=element_blank(), axis.title.x = element_text(size = 25)) + geom_vline(xintercept = 0, color = "darkblue", size = 1.5)+ geom_vline(xintercept = 3, color = "forestgreen", size = 1.5) + geom_vline(xintercept = 7, color = "darkseagreen1", size = 1.5) + scale_y_continuous(expand = c(0, 0)) + theme(legend.position = "none")
#ggsave("~/Desktop/pgs_sep.png", height = 5, width = 7)
```




```{r,eval = FALSE, echo=FALSE}
Latitude <- seq(0, 100, 1)
PGS <- Latitude + rnorm(length(Latitude), 0, 10)

df <- as.data.frame(cbind(PGS, Latitude))
ggplot(df, aes(x = Latitude, y = PGS)) + geom_point(size = 2, color = "darkmagenta") + theme_classic(base_size = 16) +geom_smooth(method='lm', se = FALSE, color = "black", linetype = "dashed") + ylab("Polygenic Score") + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.x = element_text(size = 22))
```

```{r,eval = FALSE, echo=FALSE}
Temperature <- rep(0, 100)
PGS <- Temperature + rnorm(length(Temperature), 0, 5)

df <- as.data.frame(cbind(PGS, seq(37, 46.9, 0.1)))
ggplot(df, aes(x = V2, y = PGS)) + geom_point(size = 2, color = "deeppink3") + theme_classic(base_size = 16) +geom_smooth(method='lm', se = FALSE, color = "black", linetype = "dashed") + ylab("Polygenic Score") + xlab("Temperature") +theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.x = element_text(size = 22))
```

```{r,eval = FALSE, echo=FALSE}
Time <-  seq(0, 10000, 100)
PGS <- Time + rnorm(length(Time), 0, 1000)

df <- as.data.frame(cbind(PGS, Time))
ggplot(df, aes(x = -1*Time, y = PGS)) + geom_point(size = 2, color = "slateblue4") + theme_classic(base_size = 16) +geom_smooth(method='lm', se = FALSE, color = "black", linetype = "dashed") + ylab("Polygenic Score") + xlab("Time") +theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.x = element_text(size = 22))
```

```{r,eval = FALSE, echo=FALSE}
pop <-  rep(seq(1,8, 1),10)
PGS <- pop + rnorm(length(pop), 0, 1)

df <- as.data.frame(cbind(PGS, pop))
ggplot(df, aes(x = pop, y = PGS, color = pop)) + geom_point(size = 2) + theme_classic(base_size = 16) +geom_smooth(method='lm', se = FALSE, color = "black", linetype = "dashed") + ylab("Polygenic Score") + xlab("Population ID") +theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.x = element_text(size = 22))
```


```{r,eval = FALSE, echo=FALSE}
df <- fread("~/polygenic_adaptation_stratification/code/plot/SimpleGrid_C1_scale_T100_pgs.txt")

df2 <- df %>% filter(env == "env-0.05") %>% group_by(rep, case, env) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP, env) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = mean(LAT), LONG = mean(LONG))
df3 <- df2[,3:6] %>%  gather(variable, value, -c(LAT, LONG))

df3$variable <- factor(df3$variable ,labels=c('A'=parse(text=TeX('No covariate')),
                                               'B'=parse(text=TeX('Including $T^{GWAS}$'))))

ggplot(df3, aes(x = LONG, y = LAT, fill = value)) + geom_tile() + facet_wrap(~variable, labeller=label_parsed) + theme_classic() +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 20), axis.text = element_text(size = 12), legend.title = element_text(size = 16), strip.text.x = element_text(size = 16)) + scale_fill_gradient2(low = "red", high = "blue", mid = "white", limits = c(-0.5, 0.5), name = "PGS") + xlab("Longitude") + ylab("Latitude") + scale_y_continuous(breaks=seq(0,6,1)) + scale_x_continuous(breaks=seq(0,6,1)) 
```














