---
title: "Simulation_Methods_Reproducible"
author: "Jennifer Blanc"
date: "1/3/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
library(data.table)
library(tidyverse)
library(latex2exp)
library(ggpubr)
library(dplyr)
```

## Introduction 

Here I will reccord the details for the simulations presented in the main text and the supplement. I will also include the code to make the the figures.

## 4PopSplit 

### Msprime simulation 

Generate diplois individuals according to the demographic model below 

* T2 = 2200  
* T1 (deeper split) = 4400 
* Ne (ancestral = A = B = C = D)  = 10000
* sample size = 10000 diploids per population  
* chromosome length = 100000    
* 200 chromosomes  
* $\mu$ = 1e-08  
* $\rho$ = 1e-08  

### Create genotypes panels and filter SNPs  

1. Split individuals into 2 configurations  
  a. C1: GWAS Panel = A and C, Test Panel = B and D (overlapping structure)  
  b. C2: GWAS Panel = A and B, Test Panel = C and D (no overlapping structure)  
2. Randomly downsample 1,000 individuals from each test panel population. Finale sample sizes: GWAS panel n=20,000 and Test panel n=2,000.  
3. Restrict snps to sites segregating at greater than 5% frequency within each panel.   

See [link](Simulate_Genotypes.html) for more simluations details.  

### Draw phenotypes 

We simulated both heritable $h^2 = 0.3$ and non-heritable $h^2=0$ phenotypes with a ranger of environmental confounders $\Delta_{AB}$. Additionally, for the heritable simulations, we also conducted simulations where there was a true signal of polygenic selection by flipping the signs of causal effect sizes depending on allele frequency difference in the test panel.  

See [link](Simulate_Phenotypes.html) for more simluations details.  

### Project Test Vector  

1. Set test vector as a mean centered $n \times 1$ vector with each entry as an indicator of the test panel individual's popultion ID.  
2. Project test vector in GWAS panel, resulting in an $m \times 1$ (see [link](Calculate_TGWAS.html) for more detail)

### Run GWASs  

1. Use plink to run GWAS with and without including projected test vector (see [link](Run_GWAS.html) for more detail)

### Calculate PGS and Qx 

**Ascertained Sites**

1. Pick the SNP with lowest p-value per chromosome to include in PGS (200 SNPs total)  
2. Compute PGS in test panel individuals 
3. Calculate $V_a$ 
4. Calculate the amount of variance explained by the test vector ($\lambda_T$)  
5. Calculate chi-squared p-vale
5. Calculate empirical null p-value by permuting the signs of the SNP effect sizes 1,000 times and computing the proportion of permutations greater than the true $Q_x$ value.  

**Causal sites** 

For heritable traits we can also repeat the above steps with the acual causal sites as opposed to choosing the site with the lowest p-value. 

See [link](Polygenic_Selection_Test.html) for more detail.  


```{r,eval = TRUE, echo=FALSE, warning=FALSE}
plot_data <- function(data, color_name, eb_width = 0.005) {
  
  pl <- ggplot(data = data, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =2.5)+   xlab(TeX("$\\Delta_{AB}$")) + ylab("False Postive Rate") + theme_classic(base_size = 12) + scale_color_manual(values = c(color_name)) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(-0.1,1) + labs(colour="") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme( axis.title = element_text(size = 12),legend.position = "none") + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width = eb_width)
  
  return(pl)
}

#plot_data <- function(data, color_name) {
  
#  pl <- ggplot(data = data, aes(x = diff, y = fp_strat, color = type)) + geom_point(size =3)+   xlab(TeX("$\\Delta_{AB}$")) + ylab("False Postive Rate") + theme_classic(base_size = 20) + scale_color_manual(values = c(color_name)) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(-0.1,1) + labs(colour="") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme( axis.title = element_text(size = 20),legend.position = "none") + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width = 0.0025)
  
#  return(pl)
#}
```

### Figures 

#### Figure 1  

* $h^2 = 0$  
* Ascertained SNPs (minimum p-value per chromosome) 

```{r,eval = TRUE, echo=FALSE, warning=FALSE}
df <- fread("../code/Make_Plots/A_4PopSplit_h20.txt.gz")

# C1 - no correction
nc <- df %>% filter(type == "nc") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")  
nc$diff <- as.numeric(nc$diff)
pl1 <- plot_data(nc, "goldenrod2") + ggtitle("C1 - No correction") + theme(plot.title = element_text(hjust = 0.5,size = 12))

# C1 - correction 
nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl2 <- plot_data(nc, "navy") + ggtitle("C1 - TGWAS") + theme(plot.title = element_text(hjust = 0.5,size = 12))

# C2 - No correction 
nc <- df %>% filter(type == "nc") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl3 <- plot_data(nc, "goldenrod2")+ ggtitle("C2 - No correction") + theme(plot.title = element_text(hjust = 0.5, size = 12))

# C2 - correction 
nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl4 <- plot_data(nc, "navy") + ggtitle("C2 - TGWAS") + theme(plot.title = element_text(hjust = 0.5, size = 12))

p <- ggarrange(pl1, pl3, pl2, pl4)
p


#ggsave("~/Desktop/C2-C.png",pl4, height = 5, width = 5.5)
```

We can also compare the above results using $T^{GWAS}$ to a GWAS done using population ID as covariate.  

```{r,eval = TRUE, echo=FALSE, warning=FALSE, fig.height=2.5, fig.width=7}
# C1 - correction Pop ID 
nc <- df %>% filter(type == "nc-ID") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl5 <- plot_data(nc, "darkgreen") + ggtitle("C1 - Pop ID") + theme(plot.title = element_text(hjust = 0.5, size = 12))

# C2 - correction 
nc <- df %>% filter(type == "nc-ID") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl6 <- plot_data(nc, "darkgreen") + ggtitle("C2 - Pop ID") + theme(plot.title = element_text(hjust = 0.5, size = 12))

p2 <- ggarrange(pl5, pl6)
p2
```

#### Heritable Simulations 

* $h^2 = 0.3$ 
* Asceratined SNPs  

```{r,eval = TRUE, echo=FALSE, warning=FALSE, fig.height=7, fig.width=7}
df <- fread("../code/Make_Plots/A_4PopSplit_h23.txt.gz")

# Get true fp rates 
df_true_C1 <- df %>% filter(type == "True" & case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_") 
df_true_C1$diff <- as.numeric(df_true_C1$diff)
df_true_C2 <- df %>% filter(type == "True" & case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")  
df_true_C2$diff <- as.numeric(df_true_C2$diff)

# C1 - no correction
nc <- df %>% filter(type == "nc") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")  
nc$diff <- as.numeric(nc$diff)
pl1 <- plot_data(nc, "goldenrod2", eb_width = 0.1) + ggtitle("C1 - No correction") + theme(plot.title = element_text(hjust = 0.5,size = 12)) + geom_point(data = df_true_C1, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C1 - correction 
nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl2 <- plot_data(nc, "navy", eb_width = 0.1) + ggtitle("C1 - TGWAS") + theme(plot.title = element_text(hjust = 0.5,size = 12)) + geom_point(data = df_true_C1, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C2 - No correction 
nc <- df %>% filter(type == "nc") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl3 <- plot_data(nc, "goldenrod2", eb_width = 0.1) + ggtitle("C2 - No correction") + theme(plot.title = element_text(hjust = 0.5, size = 12)) + geom_point(data = df_true_C2, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C2 - correction 
nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl4 <- plot_data(nc, "navy",eb_width = 0.1) + ggtitle("C2 - TGWAS") + theme(plot.title = element_text(hjust = 0.5, size = 12)) + geom_point(data = df_true_C2, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C1 - correction Pop ID 
nc <- df %>% filter(type == "nc-ID") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl5 <- plot_data(nc, "darkgreen",eb_width = 0.1) + ggtitle("C1 - Pop ID") + theme(plot.title = element_text(hjust = 0.5, size = 12)) + geom_point(data = df_true_C1, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C2 - correction 
nc <- df %>% filter(type == "nc-ID") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl6 <- plot_data(nc, "darkgreen",eb_width = 0.1) + ggtitle("C2 - Pop ID") + theme(plot.title = element_text(hjust = 0.5, size = 12)) + geom_point(data = df_true_C2, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

p <- ggarrange(pl1, pl3, pl2, pl4, pl5, pl6, nrow = 3, ncol = 2)
p
```

* $h^2 = 0.3$ 
* Causal SNPs  

```{r,eval = TRUE, echo=FALSE, warning=FALSE,fig.height=7, fig.width=7}
# C1 - no correction
nc <- df %>% filter(type == "c") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")  
nc$diff <- as.numeric(nc$diff)
pl1 <- plot_data(nc, "goldenrod2",eb_width = 0.1)+ ggtitle("C1 - No correction") + theme(plot.title = element_text(hjust = 0.5,size = 12)) + geom_point(data = df_true_C1, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C1 - correction 
nc <- df %>% filter(type == "c-Tm") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl2 <- plot_data(nc, "navy",eb_width = 0.1) + ggtitle("C1 - TGWAS") + theme(plot.title = element_text(hjust = 0.5,size = 12)) + geom_point(data = df_true_C1, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C2 - No correction 
nc <- df %>% filter(type == "c") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl3 <- plot_data(nc, "goldenrod2",eb_width = 0.1)+ ggtitle("C2 - No correction") + theme(plot.title = element_text(hjust = 0.5, size = 12)) + geom_point(data = df_true_C2, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)


# C2 - correction 
nc <- df %>% filter(type == "c-Tm") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl4 <- plot_data(nc, "navy",eb_width = 0.1) + ggtitle("C2 - TGWAS") + theme(plot.title = element_text(hjust = 0.5, size = 12)) + geom_point(data = df_true_C2, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C1 - correction Pop ID 
nc <- df %>% filter(type == "c-ID") %>% filter(case == "C1") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl5 <- plot_data(nc, "darkgreen",eb_width = 0.1) + ggtitle("C1 - Pop ID") + theme(plot.title = element_text(hjust = 0.5, size = 12))+ geom_point(data = df_true_C1, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

# C2 - correction 
nc <- df %>% filter(type == "c-ID") %>% filter(case == "C2") %>% group_by(envs, type, case) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")
nc$diff <- as.numeric(nc$diff)
pl6 <- plot_data(nc, "darkgreen",eb_width = 0.1) + ggtitle("C2 - Pop ID") + theme(plot.title = element_text(hjust = 0.5, size = 12)) + geom_point(data = df_true_C2, aes(x = diff, y = fp_strat), col = "red", shape = 8, size = 2.5)

p <- ggarrange(pl1, pl3, pl2, pl4, pl5, pl6, nrow = 3, ncol = 2)
p
```

#### True Signal Simulations  

For these simulations we simulate true signals of polygenic selection on the trait by drawing the sign of the effect sizes according to the allele frequency difference in the test panel. All examples are in the case of overlapping structure (C1).  

##### Flip probabilities  

The x-axis here is probability that the signs of the allele frequency contrasts in the test panel, $p_C - p_D$), and the GWAS panel, $p_A - p_B$, have the same sign. Each postive rate is calculated from 100 separate evolutionary replicates.   


```{r,eval = TRUE, echo=FALSE, warning=FALSE, fig.height=3, fig.width=8}
df <- fread("../code/Make_Plots/A_4PopSplit_h23_TS.txt.gz")
ts <- fread("../code/Make_Plots/A_4PopSplit_h23_TS_magnitude.txt.gz")
df <- full_join(ts, df)

# Ascertained sites
data_nc <- df %>% group_by(envs, ts, type) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()),avg_mag = mean(ts_magnitude)) %>% separate(ts, c("ts", "prob"), "-") %>% separate(envs, c("env", "diff"), "_") %>% filter(type %in% c("nc", "nc-Tm", "nc-ID", "True")) %>% mutate(prob = as.numeric(prob)) %>% mutate(direction = case_when(diff == "-0.1" ~ "Opposite Direction", diff == "0.0" ~ "No Environmental Stratification", diff == "0.1" ~ "Same Direction")) %>% mutate(PRS = case_when(type == "nc" ~ "Uncorrected", type == "nc-Tm" ~  "TGWAS", type == "nc-ID" ~ "Population ID", type == "True" ~ "True Effect Size"))
d2_nc <- data_nc %>% filter(type %in% c("nc", "nc-Tm", "nc-ID"))

labs <- c("Uncorrected","Population ID", expression(T^{GWAS}), "True Effect Size")
#ggplot(data = data_nc, aes(x = prob, y = fp_strat, color = type, shape = type, size = type)) + geom_point() + facet_grid(~direction) + theme_classic(base_size = 12) + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 12)) + ylim(-0.1,1.1)+scale_colour_manual(name = "",labels = labs,values = c("goldenrod2", "darkgreen", "navy", "red")) + scale_shape_manual(name = "",labels = labs,values = c(16, 16, 16, 8)) + scale_size_manual(name = "",labels = labs,values = c(2, 2, 2, 3)) + geom_errorbar(data = d2_nc, aes(ymin = lowerci, ymax = upperci), width = 0.01, size =0.5)  +  scale_y_continuous(name = "Positive Rate", breaks = seq(0, 1, 0.2)) + scale_x_continuous(name = "True Signal", breaks = seq(0.5, 0.62, 0.03))
```


```{r,eval = TRUE, echo=FALSE, warning=FALSE, fig.height=3, fig.width=8}
# Causal sites
data_c <- df %>% group_by(envs, ts, type) %>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()),avg_mag = mean(ts_magnitude)) %>% separate(ts, c("ts", "prob"), "-") %>% separate(envs, c("env", "diff"), "_") %>% filter(type %in% c("c", "c-Tm", "c-ID", "True")) %>% mutate(prob = as.numeric(prob)) %>% mutate(direction = case_when(diff == "-0.1" ~ "Opposite Direction", diff == "0.0" ~ "No Environmental Stratification", diff == "0.1" ~ "Same Direction")) %>% mutate(PRS = case_when(type == "c" ~ "Uncorrected", type == "c-Tm" ~  "TGWAS", type == "c-ID" ~ "Population ID", type == "True" ~ "True Effect Size"))
d2_c <- data_c %>% filter(type %in% c("c", "c-Tm", "c-ID"))

labs <- c("Uncorrected","Population ID", expression(T^{GWAS}), "True Effect Size")
#ggplot(data = data_c, aes(x = prob, y = fp_strat, color = type, shape = type, size = type)) + geom_point() + facet_grid(~direction) + theme_classic(base_size = 12) + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 12)) + ylim(-0.1,1.1) + scale_colour_manual(name = "",labels = labs,values = c("goldenrod2", "darkgreen","navy", "red")) + scale_shape_manual(name = "",labels = labs,values = c(16, 16, 16, 8)) + scale_size_manual(name = "",labels = labs,values = c(2, 2, 2, 3)) + geom_errorbar(data = d2_c, aes(ymin = lowerci, ymax = upperci), width = 0.01, size =0.5)  +  scale_y_continuous(name = "Positive Rate", breaks = seq(0, 1, 0.2)) + scale_x_continuous(name = "True Signal", breaks = seq(0.5, 0.62, 0.03))
```

```{r,eval = TRUE, echo=FALSE, warning=FALSE, fig.height=5, fig.width=8}
data_c$SNP <- "Causal"
data_nc$SNP <- "Ascertained"

data <- rbind(data_c, data_nc)
data$PRS <- as.factor(data$PRS)
data$PRS <- factor(data$PRS, levels = c("Uncorrected", "TGWAS", "Population ID" , "True Effect Size"))

d2_c$SNP <- "Causal"
d2_nc$SNP <- "Ascertained"
d2 <- rbind(d2_c, d2_nc)

labs <- c("Uncorrected",expression(T^{GWAS}),"Population ID", "True Effect Size")
ggplot(data = data, aes(x = prob, y = fp_strat, color = PRS, shape = PRS, size = PRS)) + geom_point() + facet_grid(SNP~direction) + theme_classic(base_size = 12) + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 12)) + ylim(-0.1,1.1)+
  scale_colour_manual(name = "",labels = labs, values = c("goldenrod2", "navy", "darkgreen", "red")) + 
  scale_shape_manual(name = "",labels = labs, values = c(16, 16, 16, 8)) + 
  scale_size_manual(name = "",labels = labs,values = c(2, 2, 2, 3)) + 
  scale_y_continuous(name = "Positive Rate", breaks = seq(0, 1, 0.2)) + scale_x_continuous(name = "True Signal", breaks = seq(0.5, 0.62, 0.03)) + geom_errorbar(data = d2, aes(ymin = lowerci, ymax = upperci), width = 0.01, size =0.5) 
```

##### True signal magnitude 

The x-axis here is the average true difference in genetic values between the test panel populations ($\Delta_{CD}$).The average is taken over the 100 evolutionary replicates. 

```{r,eval = TRUE, echo=FALSE, warning=FALSE, fig.height=5, fig.width=8}
pl <- ggplot(data = data, aes(x = avg_mag, y = fp_strat, color = PRS, shape = PRS, size = PRS)) + geom_point() + facet_grid(SNP~direction) + theme_classic(base_size = 12) + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 12)) +
  scale_colour_manual(name = "",labels = labs, values = c("goldenrod2", "navy", "darkgreen", "red")) + 
  scale_shape_manual(name = "",labels = labs, values = c(16, 16, 16, 8)) + 
  scale_size_manual(name = "",labels = labs,values = c(2, 2, 2, 3)) + 
  scale_y_continuous(name = "Positive Rate", breaks = seq(0, 1, 0.2)) + scale_x_continuous(name = TeX("$\\bar{\\Delta}_{CD}$"), breaks = seq(0, 0.25, 0.05)) + geom_errorbar(data = d2, aes(ymin = lowerci, ymax = upperci), width = 0.01, size =0.5) 

pl
ggsave("~/Desktop/TS.png", pl,width = 10, height = 6)
```

#### Varying the number of SNPs 

To investigate the number of SNPs needed to accurately compute $\vec{T}^{GWAS}$, we re-ran the 4 population simulations in the overlapping structure configuration with environmental stratification equal to 0.03 while changing the number of SNPs used to compute $\vec{T}^{GWAS}$. While the exact number of total SNPs varied across simulations, there were ~22,300 total snps per simulation.        

```{r}
df <- fread("../code/Make_Plots/A_4PopSplit_snps.txt.gz")

nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C1") %>% group_by(envs, type, case, snps)%>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")  %>% separate(snps, c("snp", "num_snps"), "-")  
nc$diff <- as.numeric(nc$diff)
nc$num_snps <- as.numeric(nc$num_snps)

pl <- ggplot(data=nc, aes(x=num_snps, y=fp_strat))  + scale_x_log10(breaks = c(10,30,100,300,1000, 3000)) + geom_point(size =3, color = "turquoise4")+ xlab("Number of SNPs") + ylab("False Postive Rate") + theme_classic(base_size = 20)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(-0.1,1) + labs(colour="") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme( axis.title = element_text(size = 20),legend.position = "none") + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width = 0.1, color = "turquoise4") 
pl
#ggsave("~/Desktop/SNPs.png", pl, width = 6, height = 4)
```




## All SNPs

```{r}
df <- fread("../code/Make_Plots/A_4PopSplit_shift_GXT_allSNPs.txt.gz")
df_True <- fread("../code/Make_Plots/A_4PopSplit_shift23_TS_GXT.txt.gz")
df_True <- df_True %>% filter(type == "True", ts %in% c("p-0.50", "p-0.62"))
df <- rbind(df, df_True)

data_nc <- df %>% group_by(envs, ts, type)%>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(ts, c("ts", "prob"), "-") %>% separate(envs, c("env", "diff"), "_") 

+ scale_colour_manual(name = "", values = c("darkgreen", "navy", "darkgreen", "red"))


labs <- c("Uncorrected","Population ID", expression(T^{GWAS}), "True Effect Size")
ggplot(data = data_nc, aes(x = prob, y = fp_strat, color = type)) + geom_point() + facet_grid(~diff) + theme_classic(base_size = 12)+ theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 12)) + ylim(-0.1,1.1) + scale_colour_manual(name = "", values = c("darkgreen", "navy", "red", "goldenrod2")) + geom_errorbar(data = data_nc, aes(ymin = lowerci, ymax = upperci), width = 0.01, size =0.5) 


```

```{r}
df <- fread("../code/Make_Plots/A_4PopSplit_shift_GXT_allSNPs.txt.gz")
df_True <- fread("../code/Make_Plots/A_4PopSplit_shift23_TS_GXT.txt.gz")
df_True <- df_True %>% filter(type == "True", ts %in% c("p-0.50", "p-0.62")) %>% mutate("signal" = type) %>% select(-type)

df2 <- left_join(df, df_True, by = c("rep", "case", "h2", "ts", "envs"))
df2 <- df2 %>% mutate(bias = Qx.x - Qx.y)

df_bias <- df2 %>% group_by(envs, ts, type)%>% summarise(avg_bias = mean(bias)) %>% separate(ts, c("ts", "prob"), "-") %>% separate(envs, c("env", "diff"), "_") 

ggplot(data = data_nc, aes(x = prob, y = fp_strat, color = type)) + geom_point() + facet_grid(~diff) + theme_classic(base_size = 12)+ theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title = element_text(size = 12)) + ylim(-0.1,1.1) + scale_colour_manual(name = "", values = c("darkgreen", "navy", "red", "goldenrod2")) + geom_errorbar(data = data_nc, aes(ymin = lowerci, ymax = upperci), width = 0.01, size =0.5) 
```



## Simple Grid 

### Msprime simulation 

Generate diploid individuals using a stepping stone model on a 6x6 grid with the parameters below: 

Parameters: 

* 200 chromosomes  
* sample size = 80 diploids  
* Ne = 1000 
* $\mu$ = 1e-07  
* $\rho$ = 1e-07 
* m = 0.01  
* chromosome length = 10000 
* tmove = -9 (perpetual structure)  

The simulation was run 100 times total.  

### Create genotypes panels and filter SNPs

1. Sample 20 individuals per deme for test panels (36 * 20 individuals total) 
2. Remaining individuals (60 per deme) become the GWAS panel  
3. Restrict snps to sites segregating at greater than 5% frequency within each panel  

### Draw phenotype 

Notes:  

* All simulations have $h^2 = 0$  
* ENV is the magnitude of the phenotype shift 

1. Sample random phenotype $N(0,1)$ for all GWAS indviduals 
2. Scale variance within deme to match full size GWAS variance (see __ for more details)
3. Simulate stratification:  
  1. Lat/Lat: Shift average phenotype per Latitude (if env is the total shift, shift 1 lat by env/5, 2 lat by 2*(env/5), etc.) and use Latitude as the test vector 
  2. stratD/Lat: Shift the average phenotype on the diagonals of the grid going lower left to upper right and use latitude as the test vector 
  
### Project Test Vector  

1. Set test vector as a mean centered $n \times 1$ vector with each entry as the test panel individuals latitude/longitude/popID  
2. Project test vector in GWAS panel, resulting in an $m \times 1$ (see ___ for more detail)

### Run GWASs  

1. Use plink to run GWAS with and without including projected test vector 

### Calculate PGS and Qx 

1. Pick the SNP with lowest p-value per chromosome to include in PGS (200 SNPs total)  
2. Compute PGS in test panel individuals 
3. Calculate $V_a$ 
4. Calculate the amount of variance explained by the test vector ($\lambda_T$; note that we used the mean-centered test vector to compute this)  
5. Calculate empirical null p-value by permuting the signs of the SNP effect sizes 1,000 times and computing the proportion of permutations greater than the true $Q_x$ value.  


### Figures

```{r,eval = TRUE, echo=FALSE, warning=FALSE}
# Function to plot the false positive rates 
make_fp_plot <- function(df, pheno_type, env_type, test_type, longitude=F) {
  
  if (test_type == "LAT") {
    test_name <- "Latitude"
  }
  if (test_type == "PS") {
    test_name = "Single Deme"
  }
  
  if (longitude == F) {
    dat <- df %>% filter(test == test_type & pheno == pheno_type & envs == env_type) %>% group_by(type) %>% summarise(fp_strat = sum(P.EN <= 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% filter(type %in% c("nc", "nc-Tm", "nc-ID")) %>% mutate(test = test_name) %>% mutate(type = factor(type, levels = c("nc", "nc-Tm", "nc-ID")))
    
  pl <-ggplot(data = dat, aes(x = type, y = fp_strat, fill = type)) + geom_col()+ geom_point(fill = "black", size = 3)+ xlab("") + ylab("False Postive Rate") + theme_classic(base_size = 14) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.text.x = element_text(size = 12, color = "black",  angle = 35, hjust=1), legend.position = "none") + ylim(-0.01,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width =0.25) + scale_fill_manual(values = c("goldenrod2","navy", "darkgreen")) + scale_x_discrete(labels = unname(TeX(c("No covariate", "Including $T^{GWAS}", "Including confounder")))) + facet_wrap(~test)
  } else {
    
    dat_lat <- df %>% filter(test == test_type & pheno == pheno_type & envs == env_type) %>% group_by(type) %>% summarise(fp_strat = sum(P.EN <= 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% filter(type %in% c("nc", "nc-Tm", "nc-ID")) %>% mutate(test = test_name) %>% mutate(type = factor(type, levels = c("nc", "nc-Tm", "nc-ID")))
    
    dat_long <- df %>% filter(test == test_type & pheno == pheno_type & envs == env_type) %>% group_by(type) %>% summarise(fp_strat = sum(P.EN_Long <= 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% filter(type %in% c("nc", "nc-Tm", "nc-ID")) %>% mutate(test = "Longitude") %>% mutate(type = factor(type, levels = c("nc", "nc-Tm", "nc-ID")))
    
    dat <- full_join(dat_lat, dat_long)

    pl <-ggplot(data = dat, aes(x = type, y = fp_strat, fill = type)) + geom_col()+ geom_point(fill = "black", size = 3)+ xlab("") + ylab("False Postive Rate") + theme_classic(base_size = 14) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.text.x = element_text(size = 12, color = "black",  angle = 35, hjust=1), legend.position = "none") + ylim(-0.01,1) + labs(colour="GWAS") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width =0.25) + scale_fill_manual(values = c("goldenrod2","navy", "darkgreen")) + scale_x_discrete(labels = unname(TeX(c("No covariate", "Including $T^{GWAS}", "Including confounder")))) + facet_wrap(~test)    
    
  }
  
  return(pl)
}

# Function to plot PGS on grid
make_pgs_plot <- function(df, pheno_type, env_type, test_type, ll, up) {
  
  dat <- df %>% filter(pheno == pheno_type & envs == env_type & test == test_type) %>% mutate(nc_stan = (nc - mean(nc)) / sd(nc), nc_Tm_stan = (nc_Tm - mean(nc_Tm)) / sd(nc_Tm)) %>% ungroup() %>% group_by(POP) %>% summarise(avg_PRS = mean(nc_stan), avg_PRS_Tm = mean(nc_Tm_stan),LAT = mean(LAT), LONG = mean(LONG))
  
  dat <- dat[,2:5] %>%  gather(variable, value, -c(LAT, LONG))
  
  plu <- dat %>% filter(variable == "avg_PRS") %>% ggplot(aes(x = LONG, y = LAT, fill = value)) + geom_tile() +  theme_classic(base_size = 12) +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "brown4", high = "royalblue4",limits = c(ll, up), name = "PGS") + xlab("Longitude") + ylab("Latitude") + scale_y_continuous(breaks=seq(0,6,1)) + scale_x_continuous(breaks=seq(0,6,1)) + theme(axis.title = element_blank())
  
  plc <- dat %>% filter(variable == "avg_PRS_Tm") %>% ggplot(aes(x = LONG, y = LAT, fill = value)) + geom_tile() +  theme_classic(base_size = 12) +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_fill_gradient2(low = "brown4", high = "royalblue4",limits = c(ll, up), name = "PGS") + xlab("Longitude") + ylab("Latitude") + scale_y_continuous(breaks=seq(0,6,1)) + scale_x_continuous(breaks=seq(0,6,1)) + theme(axis.title = element_blank())
  
  pl <- ggarrange(plu, plc, common.legend = T, legend="right")
  
  return(pl)
}

# Function to plot phenotypes 
make_phenotype_plot <- function(df, pheno_type, env_type) {
  
  dat <- df %>% filter(pheno == pheno_type & envs == env_type) %>% mutate(pheno_strat = (pheno_strat - mean(pheno_strat)) / sd(pheno_strat)) %>%  group_by(POP) %>% summarise(avg_pheno = mean(pheno_strat),LAT = mean(LAT), LONG = mean(LONG))
  
  dat <- dat[,2:4] %>%  gather(variable, value, -c(LAT, LONG))
  
  pl <- ggplot(dat, aes(x = LONG, y = LAT, fill = value)) + geom_tile() + theme_classic(base_size = 12) +  theme(panel.border = element_rect(colour = "black", fill=NA, size=1), legend.position = "none") + scale_fill_gradient2(low = "palegoldenrod", high = "purple4") + xlab("Longitude") + ylab("Latitude") + scale_y_continuous(breaks=seq(0,6,1)) + scale_x_continuous(breaks=seq(0,6,1)) + theme(axis.title = element_blank())
  
  return(pl)
}

make_gwas_weights_plot <- function(df, test_type) {
  
  dat <- df %>% filter(test == test_type) %>% group_by(PC) %>% summarise(relative_avg_weight = mean(Relative_PC_weights))
  
  pl <- ggplot(data = dat, aes(x=PC, y=relative_avg_weight)) + geom_col(color = "purple4") + theme_classic(base_size = 12) + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), legend.position = "none") + xlab("Principal Component") + ylab("Avg. Relative Weight")  
  
  return(pl)
  
}


fp <- fread("../code/Make_Plots/SimpleGrid_fp_C.txt.gz")
pgs <- fread("../code/Make_Plots/SimpleGrid_pgs_C.txt.gz")
phenos <- fread("../code/Make_Plots/SimpleGrid_phenos_C.txt.gz")
gwas_weights <- fread("../code/Make_Plots/SimpleGrid_gw_C.txt.gz")
```

#### Lat/Lat 

* Simulated phenotype according to latitudinal gradient 
* Test vector is latitude 

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
pl_fp <- make_fp_plot(fp, pheno_type = "LAT", env_type = "env-0.07", test_type = "LAT", longitude = F)
pl_pgs <- make_pgs_plot(pgs, pheno_type = "LAT", env_type = "env-0.07", test_type = "LAT", -0.25, 0.25)
pl_phenotype <- make_phenotype_plot(phenos, pheno_type = "LAT", env_type = "env-0.07")
pl_gw <- make_gwas_weights_plot(gwas_weights, "LAT")

ggarrange(pl_phenotype, pl_pgs, pl_fp, pl_gw, nrow = 1, widths = c(0.2, 0.4, 0.2, 0.2))
#sv <- ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.23, 0.5, 0.2))
#ggsave("~/Desktop/LAT.png", sv, height = 2.5, width = 11)
ggsave("~/Desktop/LAT.png", pl_gw, height = 3, width = 3)
```

#### stratD/Lat 

* Simulated phenotype according to gradient along the diagonal going from lower left to upper right  
* Project latitude as the test vector 
* For each simulation do 2 polygenic adaptation tests one with latitude as the test vector and one with longitude  

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
pl_fp <- make_fp_plot(fp, pheno_type = "DIAG", env_type = "env-0.5", test_type = "LAT", longitude = T)
pl_pgs <- make_pgs_plot(pgs, pheno_type = "DIAG", env_type = "env-0.5", test_type = "LAT", -1, 1)
pl_phenotype <- make_phenotype_plot(phenos, pheno_type = "DIAG", env_type = "env-0.5")
pl_gw <- make_gwas_weights_plot(gwas_weights, "LAT")

ggarrange(pl_phenotype, pl_pgs, pl_fp, pl_gw, nrow = 1, widths = c(0.2, 0.4, 0.2, 0.2))
#sv <- ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.23, 0.5, 0.2))
#ggsave("~/Desktop/DLAT.png", sv, height = 2.5, width = 11)
ggsave("~/Desktop/DLAT.png", pl_fp, height = 3, width = 3)
```


#### singleDeme/Lat 

* Simulated phenotype difference in a single deme
* Project latitude as the test vector 
* For each simulation do 1 polygenic adaptation test with latitude as the test vector 

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
pl_fp <- make_fp_plot(fp, pheno_type = "PS", env_type = "env-0.5", test_type = "LAT", longitude = F)
pl_pgs <- make_pgs_plot(pgs, pheno_type = "PS", env_type = "env-0.5", test_type = "LAT", -0.4, 0.4)
pl_phenotype <- make_phenotype_plot(phenos, pheno_type = "PS", env_type = "env-0.5")
pl_gw <- make_gwas_weights_plot(gwas_weights, "LAT")

ggarrange(pl_phenotype, pl_pgs, pl_fp, pl_gw, nrow = 1, widths = c(0.2, 0.4, 0.2, 0.2))
#sv <- ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.23, 0.5, 0.2))
#ggsave("~/Desktop/SLAT.png", sv, height = 2.5, width = 11)
ggsave("~/Desktop/SLAT.png", pl_fp, height = 3, width = 3)
```


#### singleDeme/singleDeme 

* Simulated phenotype difference in a single deme
* Project single deme vs everyone else as test vector 
* For each simulation do 1 polygenic adaptation test with single deme vs everybody as the test vector 

```{r, eval = TRUE, echo=FALSE, warning=FALSE,fig.width=10, fig.height=3}
pl_fp <- make_fp_plot(fp, pheno_type = "PS", env_type = "env-0.5", test_type = "PS", longitude = F)
pl_pgs <- make_pgs_plot(pgs, pheno_type = "PS", env_type = "env-0.5", test_type = "PS", -0.5, 0.5)
pl_phenotype <- make_phenotype_plot(phenos, pheno_type = "PS", env_type = "env-0.5")
pl_gw <- make_gwas_weights_plot(gwas_weights, "PS")

ggarrange(pl_phenotype, pl_pgs, pl_fp, pl_gw, nrow = 1, widths = c(0.2, 0.4, 0.2, 0.2))
#sv <- ggarrange(pl_phenotype, pl_pgs, pl_fp, nrow = 1, widths = c(0.23, 0.5, 0.2))
#ggsave("~/Desktop/DD.png", sv, height = 2.5, width = 11)
ggsave("~/Desktop/DD.png", pl_gw, height = 3, width = 3)
```


```{r, echo=FALSE, eval=FALSE}
C2GWAS <- fread("../code/Make_Plots/4PopSplit_Fst_C2GWAS.txt")
mean(C2GWAS$V3)
C2Test <- fread("../code/Make_Plots/4PopSplit_Fst_C2Test.txt")
mean(C2Test$V3)
```



```{r}
df <- fread("../code/Make_Plots/A_4PopSplit_snps.txt.gz")

nc <- df %>% filter(type == "nc-Tm") %>% filter(case == "C1") %>% group_by(envs, type, case, snps)%>% summarise(fp_strat = sum(`P.EN` < 0.05)/ n(), lowerci = fp_strat - 1.96 * sqrt((fp_strat * (1-fp_strat))/ n()), upperci = fp_strat + 1.96 * sqrt((fp_strat * (1-fp_strat))/ n())) %>% separate(envs, c("env", "diff"), "_")  %>% separate(snps, c("snp", "num_snps"), "-")  
nc$diff <- as.numeric(nc$diff)
nc$num_snps <- as.numeric(nc$num_snps)

pl <- ggplot(data=nc, aes(x=num_snps, y=fp_strat))  + scale_x_log10(breaks = c(10,30,100,300,1000, 3000)) + geom_point(size =3, color = "turquoise4")+ xlab("Number of SNPs") + ylab("False Postive Rate") + theme_classic(base_size = 20)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + ylim(-0.1,1) + labs(colour="") + theme(legend.text.align = 0)+ geom_hline(yintercept = 0.05, color = "red") + theme( axis.title = element_text(size = 20),legend.position = "none") + geom_errorbar(aes(ymin = lowerci, ymax = upperci), width = 0.1, color = "turquoise4") 
pl
ggsave("~/Desktop/SNPs.png", pl, width = 6, height = 4)
```


```{r}
r_bins <- fread("../code/Make_Plots/A_4PopSplit_rbins.txt.gz")
colnames(r_bins)[4:5] <- c("r_mean", "r_sd")
B_bins <- fread("../code/Make_Plots/A_4PopSplit_Bbins.txt.gz")

df <- left_join(B_bins, r_bins)
df <- df %>% filter(case == "C1", envs %in% c("env_0.0", "env_0.01", "env_0.02", "env_0.03")) 
labs <- list("env_0.0" = TeX("$\\Delta_{AB}$ = 0"),
          "env_0.01" = TeX("$\\Delta_{AB}$ = 0.01"),
          "env_0.02" = TeX("$\\Delta_{AB}$ = 0.02"),
          "env_0.03" = TeX("$\\Delta_{AB}$ = 0.03"))
hospital_labeller <- function(variable,value){
  return(labs[value])
}

las <- c("Population ID",expression(T^{GWAS}), "Uncorrected")
pl <- ggplot(data = df, aes(x = r_mean, y = avg, color= type))+ geom_point() +geom_smooth(method = "lm", se = FALSE) + facet_wrap(~envs,labeller=hospital_labeller) + ylab("Mean estimated effect size") + xlab("Mean r per decile") + theme_classic(base_size = 10)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_colour_manual(name = "Correction Type",labels = las,values = c("darkgreen", "navy","goldenrod2"))

#ggsave("~/Desktop/r_plot.png", pl, width = 7, height = 5)
```

Ascertained causal sites 
```{r}
r_bins <- fread("../code/Make_Plots/A_4PopSplit_rbins.txt.gz")
colnames(r_bins)[4:5] <- c("r_mean", "r_sd")
B_bins <- fread("../code/Make_Plots/A_4PopSplit_Bbins_causal.txt.gz")

df <- left_join(B_bins, r_bins)
df <- df %>% filter(case == "C1", envs %in% c("env_0.0", "env_-0.1", "env_0.1")) %>% filter(ts == "p-0.62")
labs <- list("env_0.0" = TeX("$\\Delta_{AB}$ = 0"),
          "env_-0.1" = TeX("$\\Delta_{AB}$ = -0.1"),
          "env_0.1" = TeX("$\\Delta_{AB}$ = 0.1"))
hospital_labeller <- function(variable,value){
  return(labs[value])
}

las <- c("Population ID",expression(T^{GWAS}),"True Effect Size", "Uncorrected")
pl <- ggplot(data = df, aes(x = r_mean, y = avg, color= type))+ geom_point(alpha = 0.2, shape = 1) +geom_smooth(method = "lm", se = FALSE) + facet_wrap(~envs,labeller=hospital_labeller) + ylab("Mean estimated effect size") + xlab("Mean r per decile") + theme_classic(base_size = 10)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_colour_manual(name = "Correction Type",labels = las,values = c("darkgreen", "navy","red", "goldenrod2"))

ggsave("~/Desktop/r_plot_causal.png", pl, width = 7, height = 5)
```

All sites 
```{r}
r_bins <- fread("../code/Make_Plots/A_4PopSplit_rbins.txt.gz")
colnames(r_bins)[4:5] <- c("r_mean", "r_sd")
B_bins <- fread("../code/Make_Plots/A_4PopSplit_ts_Bbins.txt.gz")
tmp <- fread("../code/Make_Plots/A_4PopSplit_Bbins_causal.txt.gz")%>% filter(type == "True")
B_bins <- rbind(B_bins, tmp) 

df <- left_join(B_bins, r_bins)
df <- df %>% filter(case == "C1", envs %in% c("env_0.0", "env_-0.1", "env_0.1")) %>% filter(ts == "p-0.62")
labs <- list("env_0.0" = TeX("$\\Delta_{AB}$ = 0"),
          "env_-0.1" = TeX("$\\Delta_{AB}$ = -0.1"),
          "env_0.1" = TeX("$\\Delta_{AB}$ = 0.1"))
hospital_labeller <- function(variable,value){
  return(labs[value])
}

las <- c("Population ID",expression(T^{GWAS}),"True Effect Size", "Uncorrected")
pl <- ggplot(data = df, aes(x = r_mean, y = avg, color= type))+ geom_point(alpha = 0.2, shape = 1) +geom_smooth(method = "lm", se = FALSE) + facet_wrap(~envs,labeller=hospital_labeller) + ylab("Mean estimated effect size") + xlab("Mean r per decile") + theme_classic(base_size = 10)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) + scale_colour_manual(name = "Correction Type",labels = las,values = c("darkgreen", "navy","red", "goldenrod2"))

ggsave("~/Desktop/r_plot_all.png", pl, width = 7, height = 5)
```









```{r}
Va <- 1
compute_fpr <- function(N, Ft) {
  
  t1 <- qnorm(0.025, mean = 0, sd = sqrt(Va * (1/(N-1))))
  t2 <- qnorm(0.975, mean = 0, sd = sqrt(Va * (1/(N-1))))

  p1 <- pnorm(t1, mean = 0, sd=sqrt(Va * ((1/(N-1)) +  Ft)))
  p2 <- 1 - pnorm(t2, mean = 0, sd=sqrt(Va * ((1/(N-1)) +  Ft)))

  fpr <- p1 + p2
  return(fpr)
}

Fts <- c(0.01, 0.001, 0.0001,0.00001, 0)
col<-terrain.colors(length(Fts))
col <- c("#00A600", "#63C600" ,"#E6E600", "#EAB64E", "#EEB99F" )
#col <- c("#00A600" ,"#E6E600","#EAB64E", "#EEB99F", "#F2F2F2")
for (i in 1:length(Fts)) {
  curve(compute_fpr(x, Ft = Fts[i]), from = 10, to = 1000000,add=i!=1, col=col[i], log = "x", xlab = "Test panel sample size", ylab = "False positive rate", lwd=2)
}
abline(v = 300000, col="red", lwd=3, lty=2)
legend("topleft", inset = 0.05,legend = Fts, col=col, lty=1, title=TeX("$f_{TX}$"))
```



```{r}
df <- fread("../code/Make_Plots/A_4PopSplit_nc_Bias.txt.gz")
tmp <- df %>% filter(case == "C1", envs %in% c("env_0.1")) %>% filter(ts == "p-0.62", type ==  "c-uncorrected") 
df_plot <- df %>% group_by(case, envs, nc, type, ts) %>% summarise(avg_bias = mean(Bias), avg_q = mean(q))
df_plot <- df_plot %>% filter(envs == "env_0.0") %>% separate(ts, c("tmp", "ts"), "-") %>% filter(type %in% c("c-Tm", "true"))
#%>% filter(type %in% c("c-uncorrected", "c-Tm", "c-ID", "true"))
df_plot$ts <- as.numeric(df_plot$ts)

las <- c(expression(T^{GWAS}),"True Effect Size") 
pl1 <- ggplot(data = df_plot,  aes(x =  ts, y = avg_q, fill = type)) + geom_bar(stat='identity',position='dodge') + facet_grid(~nc) + ylab("Avg. q") + xlab(" ") + scale_fill_manual(values = c("navy", "red"), labels = las, name = "" ) + theme_classic(base_size = 10)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.text.x = element_blank())+ scale_x_continuous(breaks = c(0.5, 0.53, 0.56, 0.59, 0.62))
pl2 <- ggplot(data = df_plot,  aes(x =  ts, y = avg_bias, fill = type)) + geom_bar(stat='identity',position='dodge') + facet_grid(~nc) + ylab("Avg. Bias") + xlab("Strength of Signal") + scale_fill_manual(values = c("navy", "red"), labels = las, name = "" ) + theme_classic(base_size = 10)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.text.x = element_text(angle = 45,  hjust = 1))+ scale_x_continuous(breaks = c(0.5, 0.53, 0.56, 0.59, 0.62))
f <- ggarrange(pl1, pl2, nrow = 2, common.legend = TRUE, legend="bottom")
ggsave("~/Desktop/pl1.png",f)

df_plot <- df %>% group_by(case, envs, nc, type, ts) %>% summarise(avg_bias = mean(Bias), avg_q = mean(q))
df_plot <- df_plot %>% filter(envs == "env_0.0") %>% separate(ts, c("tmp", "ts"), "-") %>% filter(type %in% c("c-uncorrected", "c-Tm", "c-ID", "true"))
df_plot$ts <- as.numeric(df_plot$ts)

las <- c("Population ID",expression(T^{GWAS}), "Uncorrected", "True Effect Size")
pl1 <- ggplot(data = df_plot,  aes(x =  ts, y = avg_q, fill = type)) + geom_bar(stat='identity',position='dodge') + facet_wrap(~nc, scales = "free_y", ncol = 4) + ylab("Avg. q") + xlab(" ")  + theme_classic(base_size = 10)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.text.x = element_blank())+ scale_x_continuous(breaks = c(0.5, 0.53, 0.56, 0.59, 0.62)) +  scale_fill_manual(name = "Correction Type",labels = las,values = c("darkgreen", "navy", "goldenrod2", "red"))
pl2 <- ggplot(data = df_plot,  aes(x =  ts, y = avg_bias, fill = type)) + geom_bar(stat='identity',position='dodge') + facet_wrap(~nc,scales = "free_y", ncol = 4) + ylab("Avg. Bias") + xlab("Strength of Signal") + theme_classic(base_size = 10)  + theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.text.x = element_text(angle = 45,  hjust = 1))+ scale_x_continuous(breaks = c(0.5, 0.53, 0.56, 0.59, 0.62)) +  scale_fill_manusal(name = "Correction Type",labels = las,values = c("darkgreen", "navy", "goldenrod2", "red"))
f <- ggarrange(pl1, pl2, nrow = 2, common.legend = TRUE, legend="bottom")
f
ggsave("~/Desktop/pl2.png",f)
#ggplot(data = df_plot,  aes(x =  ts, y = avg_bias, col = type)) + geom_point()
```









